---
title: "A multi-omics molecular landscape of 30 tissues in aging female rhesus macaques (Macaca mulatta)"
author: 
    - "Code author: Gong-Hua Li \n"
    - "Project correspondings: Xing-Hua Pan \\& Qing-Peng Kong\n"
    - "Code contact: Gong-Hua Li (ligonghua@mail.kiz.ac.cn)"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: no
  pdf_document: 
    toc: yes
    latex_engine: xelatex
---

# 1. Setwd and env

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  collapse = TRUE)
## please setup your root dir that it should have ./data sub dir
rootdir <- "/home/ligh/pubproject/MCMT/"
knitr::opts_knit$set(root.dir = rootdir)
setwd(rootdir)

# mk subdirs
if (!dir.exists("./results"))
  dir.create("./results")
tmpdirs <- c("data_overall", "data_statu", 
  "trajactory", "trajactory_combine",
  "differTypes", "ratio", "lm_metabolites",
  "meta_lm_GO", "compare_ms_mcc_human", "experiments")
for (x in tmpdirs) {
  tmpx <- paste0("./results/", x)
  if (!dir.exists(tmpx)) {
    dir.create(tmpx)
  }
}
```

```{r}
library(ggplot2)
library(ggbreak)
library(reshape2)
library(Hmisc)
library(pracma)
library(ggrepel)
library(grid)
library(gridExtra)
library(rlist)
library(ggsci)
library(scales)
library(data.table)
library(RColorBrewer)
library(readxl)
library(circlize)
library(M3C)
library(edgeR)
library(MetaDE)
library(limma)
library(Mfuzz)
library(EnhancedVolcano)
library(pheatmap)
library(org.Hs.eg.db)
library(clusterProfiler)
library(fgsea)
library(msigdbr)
library(enrichplot)
library(ComplexHeatmap)
source("./Subfunctions_Macaca_30tissue_aging.R")
set.seed(2025520)
```

# 2. Load data & QC

## 2.1 protein and metbolites

```{r}
# whole body data
load("./data/pro.whole.fdr0.01_from_NOVO_remap_solid_tissues.Rdata")
load("./data/met_whole_from_novo_solid_tissues.Rdata")

# tissue data
load("./data/pro.tissues_solid_tissues.Rdata")
load("./data/met.tissues_solid_tissues.Rdata")


load("./data/met.header.all.hmdb_curated.Rdata")
idx <- !is.na(met.header.all.hmdb$hmdbid_highconfidence)
met.header.all.hmdb.v <- met.header.all.hmdb[idx, ]

# gene info
# mouse
mousegeneInfo <- file2frame("./data/mouse_geneInfo_GRCm39v113.txt")
mousegeneInfo <- mousegeneInfo[!duplicated(mousegeneInfo$Symbol) &
  !is.na(mousegeneInfo$Symbol), ]
rownames(mousegeneInfo) <- mousegeneInfo$Symbol

# human
geneInfo.v38 <- file2frame("./data/geneInfo_encodev38.tab")
geneInfo.v38 <- geneInfo.v38[!duplicated(geneInfo.v38$Symbol), ]
rownames(geneInfo.v38) <- geneInfo.v38$Symbol

# macaca
macacageneInfo <- file2frame("./data/Macaca_mulatta.Mmul_10.99.ensemble_symbol_biotype.txt")
macacageneInfo <- macacageneInfo[!duplicated(macacageneInfo$Symbol) &
  !is.na(macacageneInfo$Symbol), ]
rownames(macacageneInfo) <- macacageneInfo$Symbol
```

## 2.2 set color

```{r}
alltissues <- names(pro.tissues)

tissue.systems <- c("Integumentary", "Endocrine", "Brain", "Respiratory", "Digestive",
  "Cardiovascular", "Cardiovascular", "Brain", "Digestive", "Endocrine",
  "Cardiovascular", "Muscle", "Reproductive", "Digestive", "Brain",
  "Immune", "Renal", "Endocrine", "Digestive", "Reproductive",
  "Digestive", "Brain", "Muscle", "Immune", "Integumentary", "Endocrine",
  "Brain", "Immune", "Cardiovascular", "Reproductive")
names(tissue.systems) <- alltissues
tissue.color <- pal_npg()(10)
names(tissue.color) <- levels(factor(tissue.systems))

tissue.systems
tissue.color
```

```{r}
mypal <- pal_npg()(10)
mypal
show_col(mypal)
alltissues <- names(pro.tissues)
```

## 2.3 get mRNA

```{r}
# load raw RNA counts data and calculate log2 TMM normalized CPM and
#  perform batch and RIN adjust
load("./data/rnaData.Rdata")
outpath <- "./results/data_statu/"
rnaData <- updateRNA(rnaData, headers, tissue.systems, macacageneInfo, outpath)

mrna.tissues.noadj <- rnaData$mrna.tissues.org
mrna.tissues <- rnaData$mrna.tissues
mrna.tissues.info <- rnaData$mrna.tissues.info
mrna.whole <- rnaData$mrna.whole
mrna.whole.info <- rnaData$mrna.whole.info
for (i in 1:length(mrna.tissues)) {
  colnames(mrna.tissues[[i]]) <- substr(colnames(mrna.tissues[[i]]), 1, 6)
  colnames(mrna.tissues.noadj[[i]]) <- substr(colnames(mrna.tissues.noadj[[i]]), 1, 6)
}
for (i in 1:length(mrna.tissues.info)) {
  rownames(mrna.tissues.info[[i]]) <- substr(rownames(mrna.tissues.info[[i]]), 1, 6)
}


# get stats batch and RIN adjust
tissuenames <- names(pro.tissues)
tmpaa <- mRNA_batch_RIN_analysis(rnaData, tissuenames, outpath)
```

## 2.4. quality control

```{r}
# protein remove outliners
tmp <- pro.whole
pp <- prcomp(t(tmp), cor = F)
outlinerids <- c()
thetissues <- unique(pro.whole.info$tissue_en)
for (i in 1:length(thetissues)) {
  idx <- pro.whole.info$tissue_en == thetissues[i]
  tmpinfo <- pro.whole.info[idx, ]
  outx <- is.outliner(pp$x[idx, 1])
  outy <- is.outliner(pp$x[idx, 2])
  if (sum(outx | outy) > 0) {
    outlinerids <- c(outlinerids, rownames(tmpinfo)[outx | outy])
  }
}
outlinerids
vid <- !is.element(colnames(pro.whole), outlinerids)
pro.whole <- pro.whole[, vid]
pro.whole.info <- pro.whole.info[vid, ]
```

```{r}
# met remove outliners
tmp <- met.whole
pp <- prcomp(t(tmp), cor = F)
outlinerids <- c()
thetissues <- unique(met.whole.info$tissue_en)
for (i in 1:length(thetissues)) {
  idx <- met.whole.info$tissue_en == thetissues[i]
  tmpinfo <- met.whole.info[idx, ]
  outx <- is.outliner(pp$x[idx, 1])
  outy <- is.outliner(pp$x[idx, 2])
  if (sum(outx | outy) > 0) {
    outlinerids <- c(outlinerids, rownames(tmpinfo)[outx | outy])
  }
}
outlinerids
vid <- !is.element(colnames(met.whole), outlinerids)
met.whole <- met.whole[, vid]
met.whole.info <- met.whole.info[vid, ]
```

```{r}
# mrna remove outliners
tmp <- mrna.whole
pp <- prcomp(t(tmp), cor = F)
outlinerids <- c()
thetissues <- unique(mrna.whole.info$tissue_en)
for (i in 1:length(thetissues)) {
  idx <- mrna.whole.info$tissue_en == thetissues[i]
  tmpinfo <- mrna.whole.info[idx, ]
  outx <- is.outliner(pp$x[idx, 1])
  outy <- is.outliner(pp$x[idx, 2])
  if (sum(outx | outy) > 0) {
    outlinerids <- c(outlinerids, rownames(tmpinfo)[outx | outy])
  }
}
outlinerids
vid <- !is.element(colnames(mrna.whole), outlinerids)
mrna.whole <- mrna.whole[, vid]
mrna.whole.info <- mrna.whole.info[vid, ]
```

```{r}
# match proteome, trancriptome and metabolism
alltissues <- names(pro.tissues)
pro.tissues.v <- pro.tissues
for (i in 1:length(alltissues)) {
  pro.tissues.v[[i]] <- delete_dup_genes_forprotein(pro.tissues[[i]],
    gtf.gene2symbol = pro.tissues.header[[i]])
}

mrna.tissues <- mrna.tissues[names(pro.tissues)]
mrna.tissues.noadj <- mrna.tissues.noadj[names(pro.tissues)]
mrna.tissues.info <- mrna.tissues.info[names(pro.tissues)]

met.tissues <- met.tissues[names(pro.tissues)]
met.tissues.info <- met.tissues.info[names(pro.tissues)]
met.tissues.header <- met.tissues.header[names(pro.tissues)]
```

# 3. Figure 1: overall data distribution & DE

## 3.1 Figure 1A flowchart

## 3.2 Figure 1B

```{r}
num_omics <- data.frame(num_mrna = rep(0, length(alltissues)), stringsAsFactors = F,
  tissues = alltissues,
  tissue_systems = tissue.systems,
  num_protein = rep(0, length(alltissues)),
  num_met = rep(0, length(alltissues))
)
for (i in 1:length(alltissues)) {
  num_omics$num_mrna[i] <- nrow(mrna.tissues[[alltissues[i]]])
  num_omics$num_protein[i] <- nrow(pro.tissues[[alltissues[i]]])
  num_omics$num_met[i] <- nrow(met.tissues[[alltissues[i]]])
}
rownames(num_omics) <- alltissues
```

```{r}
idx <- sort.int(num_omics$num_protein, decreasing = F, index.return = T)$ix
num_omics.v <- num_omics[idx, ]
idx <- sort.int(num_omics.v$tissue_systems, decreasing = F, index.return = T)$ix
num_omics.v <- num_omics.v[idx, ]
num_omics.v


writetxt(num_omics.v[, c(2, 3, 1, 4, 5)],
  "./results/data_overall/Figure_1B_number_omics_data.txt")
```

```{r,fig.width=12, fig.height=9,out.width="100%", fig.align="center",dpi = 144}
idx <- sort.int(num_omics$num_protein, decreasing = F, index.return = T)$ix
num_omics.v <- num_omics[idx, ]
idx <- sort.int(num_omics.v$tissue_systems, decreasing = F, index.return = T)$ix
num_omics.v <- num_omics.v[idx, ]

pomics <- list()
pomics[[1]] <- ggplot(num_omics.v, aes(x = factor(tissues, level = tissues),
  y = num_met,
  color = tissue_systems, fill = tissue_systems)) +
  geom_bar(stat = "identity") +
  theme_classic() + lghplot.addtheme(legend.position = "none", hjust = 1, size = 20) +
  scale_color_npg() + scale_fill_npg() +
  theme(axis.text.x = element_blank(), axis.title.x = element_blank()) +
  ylab("Number of metabolites")


pomics[[2]] <- ggplot(num_omics.v, aes(x = factor(tissues, level = tissues),
  y = num_protein,
  color = tissue_systems, fill = tissue_systems)) +
  geom_bar(stat = "identity") +
  theme_classic() + lghplot.addtheme(legend.position = "none", hjust = 1, size = 20) +
  scale_color_npg() + scale_fill_npg() +
  ylab("Number of proteins") + xlab("")

pdf(file = "./results/data_overall/Figure1B_number_of_omicsV__promet.pdf",
  width = 12, height = 9)
grid.arrange(arrangeGrob(grobs = pomics, ncol = 1,
  top = textGrob("Identified proteins and metabolites",
    gp = gpar(fontsize = 22, fontface = "bold")),
  heights = c(3.2, 5.8)))
dev.off()
mean(num_omics.v$num_mrna)
mean(num_omics.v$num_protein)
mean(num_omics.v$num_met)

# just for plot visualization
grid.arrange(arrangeGrob(grobs = pomics, ncol = 1,
  top = textGrob("Identified proteins and metabolites",
    gp = gpar(fontsize = 22, fontface = "bold")),
  heights = c(3.2, 5.8)))
```

## 3.3 Figure 1C_1E

### 3.3.1 Figure 1C mRNA

```{r, fig.height = 4, fig.width = 6, out.width="80%", fig.align="center",dpi = 144}
### for mRNA
mrna.whole.std <- standardise_matrix(mrna.whole)

p <- tsne(mrna.whole.std, labels = tissue.systems[mrna.whole.info$tissues],
  legendtextsize = 10, dotsize = 2, seed = 2025520)
p <- p + theme_classic() + lghplot.addtheme(legend.position = "none") +
  scale_color_npg() + xlab("tsne 1") + ylab("tsne 2") +
  theme(axis.line = element_line(size = 1.0)) + ggtitle("Transcriptome")
pdf(file = "./results/data_overall/Figure1C_tSNE_mrna_using_standardised.pdf",
  height = 4, width = 4)
print(p)
dev.off()


p <- tsne(mrna.whole.std, labels = tissue.systems[mrna.whole.info$tissues],
  legendtextsize = 10, dotsize = 2, seed = 2025520)
p <- p + theme_classic() + lghplot.addtheme(legend.position = "right") +
  scale_color_npg() + xlab("tsne 1") + ylab("tsne 2") +
  theme(axis.line = element_line(size = 1.0)) + ggtitle("Transcriptome")
pdf(file = "./results/data_overall/Figure1C_tSNE_mrna_using_standardised_with_legend.pdf",
  height = 4, width = 6)
print(p)
dev.off()

# just for plot visualization
print(p)
```

### 3.3.2 Figure 1D protein

```{r, fig.height = 4, fig.width = 4, out.width="60%", fig.align="center",dpi = 144}

ida <- rowSums(pro.whole < 0.1) < 0.2 * ncol(pro.whole)
sum(ida)
pro.whole.cons <- pro.whole[ida, ]
pro.whole.std <- standardise_matrix(pro.whole.cons)

p <- tsne(pro.whole.std, labels = tissue.systems[pro.whole.info$tissue_en],
  legendtextsize = 10, dotsize = 2, seed = 2025520)
p <- p + theme_classic() + lghplot.addtheme(legend.position = "none") +
  theme(axis.line = element_line(size = 1.0)) +
  scale_color_npg() + xlab("tsne 1") + ylab("tsne 2") + ggtitle("Proteome")
pdf(file = "./results/data_overall/Figure1D_tSNE_protein_using_standardised.pdf",
  height = 4, width = 4)
print(p)
dev.off()

# just for plot visualization
print(p)
```

### 3.3.3 Figure 1E metabolism

```{r, fig.height = 4, fig.width = 4, out.width="60%", fig.align="center",dpi = 144}
met.whole.std <- standardise_matrix(met.whole)
p <- tsne(met.whole.std, labels = tissue.systems[met.whole.info$Tissues],
  legendtextsize = 10, dotsize = 2, seed = 2025520)
p <- p + theme_classic() + lghplot.addtheme(legend.position = "none") +
  scale_color_npg() + xlab("tsne 1") + ylab("tsne 2") +
  theme(axis.line = element_line(size = 1.0)) + ggtitle("Metabolome")
pdf(file = "./results/data_overall/Figure1E_tSNE_met_using_std.pdf", height = 4, width = 4)
print(p)
dev.off()

# just for output html protein tsne
print(p)
```

# 4. Aging related molecules

## 4.1 one_line_DE

### 4.1.1 protein

```{r, fig.height = 4, fig.width = 4, out.width="50%", fig.align="center",dpi = 144}
DEpro.tissues.lm <- get_tissue_DEgenes_lm(pro.tissues.v, pro.tissues.info,
  tissue.systems = tissue.systems)
DEpro.tissues.MetaDE <- metaGene(pro.tissues.v, pro.tissues.info,meta.method = "REM")
DEpro.tissues.MetaDE.FEM <- metaGene(pro.tissues.v, pro.tissues.info,meta.method = "FEM")

#overlap

plot_save_DEgenes(DEpro.tissues.lm,
  pdffile = "./results/data_overall/Figure_S2_DEpro_each_tissueV1_lm.pdf",
  excelfile = "./results/data_overall/reduce_Data_S2_DE_tissue_Aging_pro_lm.xlsx"
)

## the following is for manuscript revision compare MetaDE and linear regression model
DEpro_compare_limma_MetaDE <- compare_limma_metaDE(DEpro.tissues.lm,
  DEpro.tissues.MetaDE, "Protein",
  "./results/data_overall/Figure_SX_compare_limma_metaDE_pro.pdf")

#ggven
DEpro_compare_limma_MetaDE_ggven <- compare_limma_metaDE_ggven(DEpro.tissues.lm,
  DEpro.tissues.MetaDE, DEpro.tissues.MetaDE.FEM,"Protein",
  "./results/data_overall/Figure_SX_compare_limma_metaDE_pro_ggven.pdf")

# just for plot visualization
plot(DEpro_compare_limma_MetaDE_ggven)

```

### 4.1.2 mRNA

```{r, fig.height = 4, fig.width = 4, out.width="50%", fig.align="center",dpi = 144}
DEmrna.tissues.lm <- get_tissue_DEgenes_lm(mrna.tissues, mrna.tissues.info,
  tissue.systems = tissue.systems)
DEmrna.tissues.MetaDE  <- metaGene(mrna.tissues, mrna.tissues.info,meta.method = "REM")
DEmrna.tissues.MetaDE.FEM <- metaGene(mrna.tissues, mrna.tissues.info,meta.method = "FEM")

## the following is for manuscript revision compare MetaDE and linear regression model
DEmrna_compare_limma_MetaDE <- compare_limma_metaDE(DEmrna.tissues.lm,
  DEmrna.tissues.MetaDE, "mRNA",
  "./results/data_overall/Figure_SX_compare_limma_metaDE_mrna.pdf")
plot_save_DEgenes(DEmrna.tissues.lm,
  pdffile = "./results/data_overall/Figure_S1_DEmrna_each_tissueV1_lm.pdf",
  excelfile = "./results/data_overall/reduce_Data_S1_DE_tissue_Aging_mrna_lm.xlsx"
)
#ggven
DEmrna_compare_limma_MetaDE_ggven <- compare_limma_metaDE_ggven(DEmrna.tissues.lm,
  DEmrna.tissues.MetaDE, DEmrna.tissues.MetaDE.FEM,"mRNA",
  "./results/data_overall/Figure_SX_compare_limma_metaDE_mrna_ggven.pdf")

# just for plot visualization
plot(DEmrna_compare_limma_MetaDE_ggven)

```

### 4.1.3 metabolites

```{r, fig.height = 4, fig.width = 4, out.width="50%", fig.align="center",dpi = 144}
DEmet.tissues.lm <- get_tissue_DEgenes_lm(met.tissues, met.tissues.info,
  tissue.systems = tissue.systems)
DEmet.tissues.MetaDE <- metaGene(met.tissues, met.tissues.info,meta.method = "REM")
DEmet.tissues.MetaDE.FEM <- metaGene(met.tissues, met.tissues.info,meta.method = "FEM")

DEmet_compare_limma_MetaDE <- compare_limma_metaDE(DEmet.tissues.lm,
  DEmet.tissues.MetaDE, "Metabolites",
  "./results/data_overall/Figure_SX_compare_limma_metaDE_mets.pdf")
plot_save_DEgenes(DEmet.tissues.lm,
  pdffile = "./results/data_overall/Figure_S3_DEmet_each_tissueV1_lm.pdf",
  excelfile = "./results/data_overall/reduce_Data_S3_DE_tissue_Aging_met_lm.xlsx"
)
#ggven
DEmet_compare_limma_MetaDE_ggven <- compare_limma_metaDE_ggven(DEmet.tissues.lm,
  DEmet.tissues.MetaDE, DEmet.tissues.MetaDE.FEM, "Metabolites",
  "./results/data_overall/Figure_SX_compare_limma_metaDE_mets_ggven.pdf")

# just for plot visualization
plot(DEmet_compare_limma_MetaDE_ggven)

```

## 4.2 plot Figure 1F

```{r,fig.width=12, fig.height=5.1, out.width= "100%", fig.align="center",dpi = 144}
plotAgingNum <- list()
Aging_pro_updown.v <- obtain_updown_mols(DEpro.tissues.lm,
  betacutoff = 0.008,
  tissue.systems = tissue.systems)
Aging_pro_updown.v <-
  Aging_pro_updown.v[order(Aging_pro_updown.v$num.all, decreasing = F), ]
Aging_pro_updown.v <-
  Aging_pro_updown.v[order(Aging_pro_updown.v$tissue_systems, decreasing = F), ]
tissueindex <- rownames(Aging_pro_updown.v)

# metabolites
Aging_met_updown.v <- obtain_updown_mols(DEmet.tissues.lm,
  betacutoff = 0.008,
  tissue.systems = tissue.systems)
Aging_met_updown.v <- Aging_met_updown.v[tissueindex, ]

p1 <- ggplot(Aging_met_updown.v, aes(x = factor(tissues, level = tissues), y = num.up,
  color = tissue_systems, fill = tissue_systems)) +
  geom_bar(stat = "identity", alpha = 0.8) + theme(axis.line = element_line(size = 1.0))

plotAgingNum[[1]] <- p1 + geom_bar(stat = "identity", aes(y = num.down), alpha = 0.6) +
  geom_hline(yintercept = 0) +
  theme_classic() + lghplot.addtheme(legend.position = "none", hjust = 1, size = 11) +
  theme(axis.text.y = element_text(size = 8, face = "bold", color = "black")) +
  theme(axis.text.x = element_text(size = 1, face = "bold", color = "black")) +
  scale_color_npg(alpha = 0.9) +
  scale_fill_npg(alpha = 0.9) + theme(axis.ticks.x = element_blank(),
    axis.text.x = element_blank(), axis.line.x = element_blank()) +
  scale_y_continuous(labels = fmt_dcimals(2)) + ggtitle("Metabolites") +
  ylab("") + xlab("")

# protein
# Aging_pro_updown.v = DEpro.tissues.lm$Aging_updown[tissueindex,]

p1 <- ggplot(Aging_pro_updown.v, aes(x = factor(tissues, level = tissues), y = num.up,
  color = tissue_systems, fill = tissue_systems)) +
  geom_bar(stat = "identity", alpha = 0.8) + theme(axis.line = element_line(size = 1.0))

plotAgingNum[[2]] <- p1 + geom_bar(stat = "identity", aes(y = num.down), alpha = 0.6) +
  geom_hline(yintercept = 0) +
  theme_classic() + lghplot.addtheme(legend.position = "none", hjust = 1, size = 11) +
  theme(axis.text.x = element_text(size = 1, face = "bold", color = "black")) +
  theme(axis.text.y = element_text(size = 8, face = "bold", color = "black")) +
  scale_color_npg(alpha = 0.9) +
  scale_fill_npg(alpha = 0.9) + theme(axis.ticks.x = element_blank(),
    axis.text.x = element_blank(), axis.line.x = element_blank()) +
  scale_y_continuous(labels = fmt_dcimals(2)) +
  ggtitle("Proteins") + xlab("") + ylab("")

# mrna
Aging_mrna_updown.v <- obtain_updown_mols(DEmrna.tissues.lm,
  betacutoff = 0.008,
  tissue.systems = tissue.systems)
Aging_mrna_updown.v <- Aging_mrna_updown.v[tissueindex, ]

p1 <- ggplot(Aging_mrna_updown.v, aes(x = factor(tissues, level = tissues), y = num.up,
  color = tissue_systems, fill = tissue_systems)) +
  geom_bar(stat = "identity", alpha = 0.8) + theme(axis.line = element_line(size = 1.0))

plotAgingNum[[3]] <- p1 + geom_bar(stat = "identity", aes(y = num.down), alpha = 0.6) +
  geom_hline(yintercept = 0) +
  theme_classic() + lghplot.addtheme(legend.position = "none", hjust = 1, size = 11) +
  theme(axis.text.x = element_text(size = 11, face = "bold", color = "black")) +
  theme(axis.text.y = element_text(size = 8, face = "bold", color = "black")) +
  scale_color_npg(alpha = 0.9) +
  scale_fill_npg(alpha = 0.9) + scale_y_continuous(labels = fmt_dcimals(2)) +
  ggtitle("mRNAs") + xlab("") + ylab("")

pdf(file = "./results/data_overall/Figure1F_number_of_Aging_moleculars_prometv3_npg.pdf",
  width = 12, height = 5.1)
grid.arrange(arrangeGrob(grobs = lapply(plotAgingNum[1:3], function(p) {
  p + theme(plot.margin = margin(t = 0, b = -3, unit = "mm"))
}),
left = textGrob("Proportion of changed molecules",
  gp = gpar(fontsize = 14, fontface = "bold"),
  rot = 90, hjust = 0.3),
ncol = 1, heights = c(1.2, 1.2, 2.7)))
dev.off()

colMeans(Aging_mrna_updown.v[, 1:3])
colMeans(Aging_pro_updown.v[, 1:3])
colMeans(Aging_met_updown.v[, 1:3])
Aging_pro_updown.v

# just for output html
grid.arrange(arrangeGrob(grobs = lapply(plotAgingNum[1:3], function(p) {
  p + theme(plot.margin = margin(t = 0, b = -3, unit = "mm"))
}),
left = textGrob("Proportion of changed molecules",
  gp = gpar(fontsize = 14, fontface = "bold"),
  rot = 90, hjust = 0.3),
ncol = 1, heights = c(1.2, 1.2, 2.7)))
```

# 5 Figure 2 common aging protein and mrna

## 5.1 meta common DEs

```{r}
# protein
tmp <- DEpro.tissues.lm$MetaLimma
tmp <- tmp[order(tmp$MetaFDR), ]
tmp[, 2:(ncol(tmp))] <- signif(tmp[, 2:(ncol(tmp))], 4)
openxlsx::write.xlsx(tmp,
  file = "./results/meta_lm_GO/Common_DEpro_from_metapro_reducesize.xlsx",
  keepNA = T, na.string = "NA")
CommMetapro <- ggvenn_two_meta_methods(DEpro.tissues.lm,
  "./results/meta_lm_GO/ggvenn_common_two_methods_pro.pdf")



# mrna
tmp <- DEmrna.tissues.lm$MetaLimma
tmp <- tmp[order(tmp$MetaFDR), ]
tmp[, 2:(ncol(tmp))] <- signif(tmp[, 2:(ncol(tmp))], 4)
openxlsx::write.xlsx(tmp,
  file = "./results/meta_lm_GO/Common_DEpro_from_metamrna_reducesize.xlsx",
  keepNA = T, na.string = "NA")
CommMetamrna <- ggvenn_two_meta_methods(DEmrna.tissues.lm,
  "./results/meta_lm_GO/ggvenn_common_two_methods_mrna.pdf")


tmp <- DEmet.tissues.lm$MetaLimma
tmp <- tmp[order(tmp$MetaFDR), ]
tmp[, 2:(ncol(tmp))] <- signif(tmp[, 2:(ncol(tmp))], 4)
openxlsx::write.xlsx(tmp,
  file = "./results/meta_lm_GO/Common_DEpro_from_metamet_reducesize.xlsx",
  keepNA = T, na.string = "NA")
CommMetamet <- ggvenn_two_meta_methods(DEmet.tissues.lm,
  "./results/meta_lm_GO/ggvenn_common_two_methods_met.pdf")
```

```{r,fig.width=8, fig.height=8, out.width= "60%", fig.align="center",dpi = 144}
# meta DE for REM common files
tmpcomm <- list()

tmpcomm <-
  list(macaca_uppro_meta = DEpro.tissues.MetaDE$ID[DEpro.tissues.MetaDE$MetaFDR < 0.05 &
    DEpro.tissues.MetaDE$Metalog2FC > 0.2 &
    !DEpro.tissues.MetaDE$manyNA],
  macaca_downpro_meta = DEpro.tissues.MetaDE$ID[DEpro.tissues.MetaDE$MetaFDR < 0.05 &
    DEpro.tissues.MetaDE$Metalog2FC < -0.2 &
    !DEpro.tissues.MetaDE$manyNA],

  macaca_upmrna_meta = DEmrna.tissues.MetaDE$ID[DEmrna.tissues.MetaDE$MetaFDR < 0.05 &
    DEmrna.tissues.MetaDE$Metalog2FC > 0.2 &
    !DEmrna.tissues.MetaDE$manyNA],
  macaca_downmrna_meta = DEmrna.tissues.MetaDE$ID[DEmrna.tissues.MetaDE$MetaFDR < 0.05 &
    DEmrna.tissues.MetaDE$Metalog2FC < -0.2 &
    !DEmrna.tissues.MetaDE$manyNA],
  macaca_upmet_meta = DEmet.tissues.MetaDE$ID[DEmet.tissues.MetaDE$MetaFDR < 0.05 &
    DEmet.tissues.MetaDE$Metalog2FC > 0.2 &
    !DEmet.tissues.MetaDE$manyNA],
  macaca_downmet_meta = DEmet.tissues.MetaDE$ID[DEmet.tissues.MetaDE$MetaFDR < 0.05 &
    DEmet.tissues.MetaDE$Metalog2FC < -0.2 &
    !DEmet.tissues.MetaDE$manyNA])

pdf("./results/meta_lm_GO/Figure_X_vennon_mrna_pro_useMetaDE_REM.pdf", width = 8)
pp <- ggvenn::ggvenn(data = list(Up_pro = tmpcomm$macaca_uppro_meta,
  Up_mrna = tmpcomm$macaca_upmrna_meta,
  Down_mrna = tmpcomm$macaca_downmrna_meta,
  Down_pro = tmpcomm$macaca_downpro_meta),
fill_color = c("darkorchid1", "yellow", "green", "cornflowerblue"),
show_percentage = F, stroke_size = 0.5, stroke_alpha = 0.6, text_size = 9
)
print(pp)
dev.off()

# just for plot visualization
print(pp)
```

```{r}
commonMols.filter <- list()

Metapro <- DEpro.tissues.lm$MetaLimma
Metamrna <- DEmrna.tissues.lm$MetaLimma
Metamet <- DEmet.tissues.lm$MetaLimma


beta_cutoff <- 0.008 # abs(sort(Metamrna$MetaBeta)[nrow(Metamrna)*0.1]), 0.2/25

commonMols.filter$FDR005 <-
  list(macaca_uppro_meta = Metapro$ID[Metapro$MetaFDR < 0.05 &
    Metapro$MetaBeta > beta_cutoff & !Metapro$manyNA],
  macaca_downpro_meta = Metapro$ID[Metapro$MetaFDR < 0.05 &
    Metapro$MetaBeta < -beta_cutoff & !Metapro$manyNA],
  macaca_upmrna_meta = Metamrna$ID[Metamrna$MetaFDR < 0.05 &
    Metamrna$MetaBeta > beta_cutoff & !Metamrna$manyNA],
  macaca_downmrna_meta = Metamrna$ID[Metamrna$MetaFDR < 0.05 &
    Metamrna$MetaBeta < -beta_cutoff & !Metamrna$manyNA],
  macaca_upmet_meta = Metamet$ID[Metamet$MetaFDR < 0.05 &
    Metamet$MetaBeta > beta_cutoff & !Metamet$manyNA],
  macaca_downmet_meta = Metamet$ID[Metamet$MetaFDR < 0.05 &
    Metamet$MetaBeta < -beta_cutoff & !Metamet$manyNA]

  )
commonMols.filter$FDR001 <-
  list(macaca_uppro_meta = Metapro$ID[Metapro$MetaFDR < 0.01 &
    Metapro$MetaBeta > beta_cutoff & !Metapro$manyNA],
  macaca_downpro_meta = Metapro$ID[Metapro$MetaFDR < 0.01 &
    Metapro$MetaBeta < -beta_cutoff & !Metapro$manyNA],
  macaca_upmrna_meta = Metamrna$ID[Metamrna$MetaFDR < 0.01 &
    Metamrna$MetaBeta > beta_cutoff & !Metamrna$manyNA],
  macaca_downmrna_meta = Metamrna$ID[Metamrna$MetaFDR < 0.01 &
    Metamrna$MetaBeta < -beta_cutoff & !Metamrna$manyNA],
  macaca_upmet_meta = Metamet$ID[Metamet$MetaFDR < 0.01 &
    Metamet$MetaBeta > beta_cutoff & !Metamet$manyNA],
  macaca_downmet_meta = Metamet$ID[Metamet$MetaFDR < 0.01 &
    Metamet$MetaBeta < -beta_cutoff & !Metamet$manyNA]
  )

length(commonMols.filter$FDR005$macaca_upmrna_meta)
length(commonMols.filter$FDR005$macaca_downmrna_meta)
length(commonMols.filter$FDR005$macaca_uppro_meta)
length(commonMols.filter$FDR005$macaca_downpro_meta)
length(commonMols.filter$FDR005$macaca_upmet_meta)
length(commonMols.filter$FDR005$macaca_downmet_meta)

# write for metascape comparison
# write geneset for metascape analysis
upgenesets <- c(paste(commonMols.filter$FDR005$macaca_uppro_meta, collapse = ","),
  paste(commonMols.filter$FDR005$macaca_upmrna_meta, collapse = ",")
)
names(upgenesets) <- c("Protein", "mRNA")
upgenesets <- as.matrix(upgenesets, 2, 1)
writetxt_forGPMM(upgenesets,
  filename = "./results/meta_lm_GO/up_genes_mrna_protein_for_metascape.txt")

downgenesets <- c(paste(commonMols.filter$FDR005$macaca_downpro_meta, collapse = ","),
  paste(commonMols.filter$FDR005$macaca_downmrna_meta, collapse = ",")
)
names(downgenesets) <- c("Protein", "mRNA")
downgenesets <- as.matrix(downgenesets, 2, 1)
writetxt_forGPMM(downgenesets,
  filename = "./results/meta_lm_GO/down_genes_mrna_protein_for_metascape.txt")
```

## 5.2 Figure 2A common mrna

```{r,fig.width=7, fig.height=5.7, out.width="60%", fig.align="center",dpi = 144}
tmpdata <- Metamrna
tmpdata$Pvalue <- tmpdata$MetaFDR
tmpdata$log2FC <- tmpdata$MetaBeta
tmpdata$log2FC[tmpdata$Pvalue == 1] <- 0
pdf("./results/meta_lm_GO/Figure_2A_vocano_common_mRNA.pdf", height = 5.7)
pp <- plot_DEflux(tmpdata[tmpdata$manyNA == FALSE, ],
  xlab = "LM beta (across 30 tissues)", FCcutoff = beta_cutoff,
  num.showlab = 5, alpha = 0.3, fixpointsize = 2, labSize = 6,
  ylab = bquote(~ -Log[10] ~ italic(FDR))) + xlim(c(-0.08, 0.08)) +
  ggtitle("Common changed transcripts across 30 tissues") +
  theme(axis.line = element_line(size = 1.0),
    axis.title.x = element_text(size = 18),
    axis.title.y = element_text(size = 18),
    axis.text.x = element_text(size = 16),
    axis.text.y = element_text(size = 16)
  )
print(pp)
dev.off()

# just for plot visualization
print(pp)
```

## 5.3 Figure 2B common protein

```{r,fig.width=7, fig.height=5.7, out.width="60%", fig.align="center",dpi = 144}
tmpdata <- Metapro
tmpdata$Pvalue <- tmpdata$MetaFDR
tmpdata$log2FC <- tmpdata$MetaBeta
tmpdata$log2FC[tmpdata$Pvalue == 1] <- 0
tmplim <- max(tmpdata$MetaBeta, na.rm = T)
pdf("./results/meta_lm_GO/Figure_2B_vocano_common_pro.pdf", height = 5.7)
pp <- plot_DEflux(tmpdata[tmpdata$manyNA == FALSE, ],
  xlab = "LM beta (across 30 tissues)",
  FCcutoff  = beta_cutoff,
  num.showlab = 5, alpha = 0.3, fixpointsize = 2, labSize = 6,
  ylab = bquote(~ -Log[10] ~ italic(FDR))) + xlim(c(-0.15, 0.15)) +
  ggtitle("Common changed transcripts across 30 tissues") +
  theme(axis.line = element_line(size = 1.0),
    axis.title.x = element_text(size = 18),
    axis.title.y = element_text(size = 18),
    axis.text.x = element_text(size = 16),
    axis.text.y = element_text(size = 16)
  )
print(pp)
dev.off()
# just for plot visualization
print(pp)
```

## 5.4 Figure 2C vennon mrna pro

```{r,fig.width=8, fig.height=7, out.width="60%", fig.align="center",dpi = 144}
pdf("./results/meta_lm_GO/Figure_2C_vennon_mrna_pro.pdf", width = 8)
pp <- ggvenn::ggvenn(data = list(Up_pro = commonMols.filter$FDR005$macaca_uppro_meta,
  Up_mrna = commonMols.filter$FDR005$macaca_upmrna_meta,
  Down_mrna = commonMols.filter$FDR005$macaca_downmrna_meta,
  Down_pro = commonMols.filter$FDR005$macaca_downpro_meta),
fill_color = c("darkorchid1", "yellow", "green", "cornflowerblue"),
show_percentage = F, stroke_size = 0.5, stroke_alpha = 0.6, text_size = 9
)
print(pp)
dev.off()

pdf("./results/meta_lm_GO/Figure_SX_vennon_mrna_pro_fdr0.01.pdf", width = 8)
ggvenn::ggvenn(data = list(Up_pro = commonMols.filter$FDR001$macaca_uppro_meta,
  Up_mrna = commonMols.filter$FDR001$macaca_upmrna_meta,
  Down_mrna = commonMols.filter$FDR001$macaca_downmrna_meta,
  Down_pro = commonMols.filter$FDR001$macaca_downpro_meta),
fill_color = c("darkorchid1", "yellow", "green", "cornflowerblue"),
show_percentage = F, stroke_size = 0.5, stroke_alpha = 0.6, text_size = 9
)
dev.off()

# just for plot visualization
print(pp)
```

## 5.5 Figure 2D heatmap mrna

```{r,fig.width=4.3, fig.height=4, out.width="60%", fig.align="center",dpi = 144}
# mrna all
mrnafc.common <- Metamrna[Metamrna$MetaFDR < 0.05 &
  abs(Metamrna$MetaBeta) > beta_cutoff & !Metamrna$manyNA,
substr(colnames(Metamrna), 1, 5) == "beta_"]
colnames(mrnafc.common) <- gsub("beta_", "", colnames(mrnafc.common))

mrnapval.common <- Metamrna[Metamrna$MetaFDR < 0.05 &
  abs(Metamrna$MetaBeta) > beta_cutoff & !Metamrna$manyNA,
substr(colnames(Metamrna), 1, 7) == "Pvalue_"]
colnames(mrnapval.common) <- gsub("Pvalue_", "", colnames(mrnapval.common))

mrnafc.common[is.na(mrnafc.common)] <- 0
mrnapval.common[is.na(mrnapval.common)] <- 1

enbrks <- 0.1 * c(-1, -0.8, -0.4, -0.2, -0.1, -0.05, 0.05, 0.1, 0.2, 0.4, 0.8, 1)

cname <- colnames(mrnafc.common)


tclass <- data.frame(System = tissue.systems[cname], row.names = cname)
mrnafc.common.v <- mrnafc.common
mrnafc.common.v[mrnafc.common.v > 0.1] <- 0.1
mrnafc.common.v[mrnafc.common.v < -0.1] <- -0.1

ann_colors <- list(
  System = tissue.color)

heatmap_mrna <- pheatmap::pheatmap(mrnafc.common.v, scale = "none",
  cluster_rows = T, show_rownames = F,
  cluster_cols = T, show_colnames = F,
  fontsize_row = 12, fontsize_col = 11,
  annotation_col = tclass,
  annotation_colors = ann_colors,
  breaks = enbrks,
  angle_col = 45, clustering_method = "ward.D2",
  treeheight_row = 20, treeheight_col = 20, legend = T,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  # file ='./results/meta_lm_GO/Figure_2D_mrna_heatmap.pdf',
  height = 4, width = 4.3)
pdf("./results/meta_lm_GO/Figure_2D_mrna_heatmap.pdf", height = 4, width = 4.3)
grid::grid.draw(heatmap_mrna$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(heatmap_mrna$gtable)
```

## 5.6 Figure 2E heatmap protein

```{r,fig.width=4.3, fig.height=4, out.width="60%", fig.align="center",dpi = 144}
profc.common <- Metapro[Metapro$MetaFDR < 0.05 &
  abs(Metapro$MetaBeta) > beta_cutoff & !Metapro$manyNA,
substr(colnames(Metapro), 1, 5) == "beta_"]
colnames(profc.common) <- gsub("beta_", "", colnames(profc.common))

propval.common <- Metapro[Metapro$MetaFDR < 0.05 &
  abs(Metapro$MetaBeta) > beta_cutoff & !Metapro$manyNA,
substr(colnames(Metapro), 1, 7) == "Pvalue_"]
colnames(propval.common) <- gsub("Pvalue_", "", colnames(propval.common))

enbrks <- 0.1 * c(-1, -0.8, -0.4, -0.2, -0.1, -0.05, 0.05, 0.1, 0.2, 0.4, 0.8, 1)

cname <- colnames(profc.common)


tclass <- data.frame(System = tissue.systems[cname], row.names = cname)
profc.common.v <- profc.common
profc.common.v[profc.common.v > 0.1] <- 0.1
profc.common.v[profc.common.v < -0.1] <- -0.1


heapmap_protein <- pheatmap::pheatmap(profc.common.v, scale = "none",
  show_rownames = F, cluster_rows = T,
  cluster_cols = T, show_colnames = F,
  fontsize_row = 8, fontsize_col = 11,
  annotation_col = tclass,
  annotation_colors = ann_colors,
  breaks = enbrks,
  clustering_method = "ward.D2",
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  # file ='./results/meta_lm_GO/Figure_2E_protein_heatmap.pdf',
  height = 4, width = 4.3)

pdf("./results/meta_lm_GO/Figure_2E_protein_heatmap.pdf", height = 4, width = 4.3)
grid::grid.draw(heapmap_protein$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(heapmap_protein$gtable)
```

## 5.7 Figure 2F mrna & pro heatmap

### 5.7.1 FDR 0.001

```{r,fig.width=9.5, fig.height=7, out.width= "80%", fig.align="center",dpi = 144}
common_up <- intersect(commonMols.filter$FDR001$macaca_uppro_meta,
  commonMols.filter$FDR001$macaca_upmrna_meta)
common_down <- intersect(commonMols.filter$FDR001$macaca_downmrna_meta,
  commonMols.filter$FDR001$macaca_downpro_meta)
common_all <- c(common_up, common_down)

# there is too many genes and we just show the top ranked genes in main figure
idx <- Metamrna[common_all, ]$MetaFDR < 1e-3 & Metapro[common_all, ]$MetaFDR < 1e-3
common_all <- common_all[idx]

# mRNA
tissue.systems.v <- sort(tissue.systems[names(pro.tissues)])

mrnafc.common <- Metamrna[common_all,
  substr(colnames(Metamrna), 1, 5) == "beta_"]
colnames(mrnafc.common) <- gsub("beta_", "", colnames(mrnafc.common))

mrnapval.common <- Metamrna[common_all,
  substr(colnames(Metamrna), 1, 7) == "Pvalue_"]
colnames(mrnapval.common) <- gsub("Pvalue_", "", colnames(mrnapval.common))

mrnafc.common <- mrnafc.common[, names(tissue.systems.v)]

mrnapval.common <- mrnapval.common[rownames(mrnafc.common), colnames(mrnafc.common)]

#
display_matrix <- matrix(" ", nrow(mrnapval.common), ncol(mrnapval.common))
display_matrix[mrnapval.common < 0.1] <- "."
display_matrix[mrnapval.common >= 0.01 & mrnapval.common < 0.05] <- "*"
display_matrix[mrnapval.common >= 0.001 & mrnapval.common < 0.01] <- "**"
display_matrix[mrnapval.common < 0.001] <- "***"

enbrks <- 0.1 * c(-1, -0.8, -0.4, -0.2, -0.1, -0.05, 0.05, 0.1, 0.2, 0.4, 0.8, 1)

cname <- colnames(mrnafc.common)

tclass <- data.frame(System = tissue.systems[cname], row.names = cname)
mrnafc.common.v <- mrnafc.common
mrnafc.common.v[mrnafc.common.v > 0.1] <- 0.1 # this is just to show make more constrast
mrnafc.common.v[mrnafc.common.v < -0.1] <- -0.1 # this is just to show make more constrast

ann_colors <- list(
  System = tissue.color)

heatmap_mrna <- pheatmap::pheatmap(mrnafc.common.v, scale = "none",
  cluster_rows = T, cluster_cols = F,
  annotation_colors = ann_colors,
  fontsize_row = 11, fontsize_col = 11,
  annotation_col = tclass,
  breaks = enbrks,
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  display_numbers = display_matrix,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  # file ='./results/meta_lm_GO/Figure_2F_mrna_nocluster_col_FDR0.001.pdf',
  height = 7, width = 9.5)

mrna_order <- rownames(mrnafc.common.v)[heatmap_mrna$tree_row$order]

pdf("./results/meta_lm_GO/Figure_2F_mrna_nocluster_col_FDR0.001.pdf",
  height = 7, width = 9.5)
grid::grid.draw(heatmap_mrna$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(heatmap_mrna$gtable)
```

```{r,fig.width=9.2, fig.height=7, out.width= "80%", fig.align="center",dpi = 144}
# protein
profc.common <- Metapro[mrna_order,
  substr(colnames(Metapro), 1, 5) == "beta_"]
colnames(profc.common) <- gsub("beta_", "", colnames(profc.common))

propval.common <- Metapro[mrna_order,
  substr(colnames(Metapro), 1, 7) == "Pvalue_"]
colnames(propval.common) <- gsub("Pvalue_", "", colnames(propval.common))


profc.common <- profc.common[mrna_order, names(tissue.systems.v)]
propval.common <- propval.common[rownames(profc.common), colnames(profc.common)]

#
display_matrix <- matrix(" ", nrow(propval.common), ncol(propval.common))
display_matrix[propval.common < 0.1] <- "."
display_matrix[propval.common >= 0.01 & propval.common < 0.05] <- "*"
display_matrix[propval.common >= 0.001 & propval.common < 0.01] <- "**"
display_matrix[propval.common < 0.001] <- "***"

cname <- colnames(profc.common)


tclass <- data.frame(System = tissue.systems[cname], stringsAsFactors = F, row.names = cname)
profc.common.v <- profc.common
profc.common.v[profc.common.v > 0.1] <- 0.1
profc.common.v[profc.common.v < -0.1] <- -0.1


heapmap_protein <- pheatmap::pheatmap(profc.common.v,
  scale = "none", cluster_rows = F, cluster_cols = F,
  fontsize_row = 11, fontsize_col = 11, legend_labels = "beta",
  annotation_col = tclass,
  annotation_colors = ann_colors,
  breaks = enbrks,
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  display_numbers = display_matrix,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  # file ='./results/meta_lm_GO/Figure_2F_proteins_notclusterrow_col_FDR0.001.pdf',
  height = 7, width = 9.2)

pdf("./results/meta_lm_GO/Figure_2F_proteins_notclusterrow_col_FDR0.001.pdf",
  height = 7, width = 9.2)
grid::grid.draw(heapmap_protein$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(heapmap_protein$gtable)
```

### 5.7.2 FDR 0.01

```{r}
common_up <- intersect(commonMols.filter$FDR001$macaca_uppro_meta,
  commonMols.filter$FDR001$macaca_upmrna_meta)
common_down <- intersect(commonMols.filter$FDR001$macaca_downmrna_meta,
  commonMols.filter$FDR001$macaca_downpro_meta)
common_all <- c(common_up, common_down)


# mRNA
tissue.systems.v <- sort(tissue.systems[names(pro.tissues)])

mrnafc.common <- Metamrna[common_all,
  substr(colnames(Metamrna), 1, 5) == "beta_"]
colnames(mrnafc.common) <- gsub("beta_", "", colnames(mrnafc.common))

mrnapval.common <- Metamrna[common_all,
  substr(colnames(Metamrna), 1, 7) == "Pvalue_"]
colnames(mrnapval.common) <- gsub("Pvalue_", "", colnames(mrnapval.common))

mrnafc.common <- mrnafc.common[, names(tissue.systems.v)]

mrnapval.common <- mrnapval.common[rownames(mrnafc.common), colnames(mrnafc.common)]

#
display_matrix <- matrix(" ", nrow(mrnapval.common), ncol(mrnapval.common))
display_matrix[mrnapval.common < 0.1] <- "."
display_matrix[mrnapval.common >= 0.01 & mrnapval.common < 0.05] <- "*"
display_matrix[mrnapval.common >= 0.001 & mrnapval.common < 0.01] <- "**"
display_matrix[mrnapval.common < 0.001] <- "***"

enbrks <- 0.1 * c(-1, -0.8, -0.4, -0.2, -0.1, -0.05, 0.05, 0.1, 0.2, 0.4, 0.8, 1)

cname <- colnames(mrnafc.common)

tclass <- data.frame(System = tissue.systems[cname], row.names = cname)
mrnafc.common.v <- mrnafc.common
mrnafc.common.v[mrnafc.common.v > 0.1] <- 0.1
mrnafc.common.v[mrnafc.common.v < -0.1] <- -0.1

ann_colors <- list(
  System = tissue.color)

heatmap_mrna <- pheatmap::pheatmap(mrnafc.common.v, scale = "none",
  cluster_rows = T, cluster_cols = F, annotation_colors = ann_colors,
  fontsize_row = 11, fontsize_col = 11,
  annotation_col = tclass,
  breaks = enbrks,
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  display_numbers = display_matrix,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  file = "./results/meta_lm_GO/Figure_2F_mrna_nocluster_col_FDR0.01.pdf",
  height = 12, width = 9.5)

mrna_order <- rownames(mrnafc.common.v)[heatmap_mrna$tree_row$order]
```

```{r}
# protein
profc.common <- Metapro[mrna_order,
  substr(colnames(Metapro), 1, 5) == "beta_"]
colnames(profc.common) <- gsub("beta_", "", colnames(profc.common))

propval.common <- Metapro[mrna_order,
  substr(colnames(Metapro), 1, 7) == "Pvalue_"]
colnames(propval.common) <- gsub("Pvalue_", "", colnames(propval.common))


profc.common <- profc.common[mrna_order, names(tissue.systems.v)]
propval.common <- propval.common[rownames(profc.common), colnames(profc.common)]

#
display_matrix <- matrix(" ", nrow(propval.common), ncol(propval.common))
display_matrix[propval.common < 0.1] <- "."
display_matrix[propval.common >= 0.01 & propval.common < 0.05] <- "*"
display_matrix[propval.common >= 0.001 & propval.common < 0.01] <- "**"
display_matrix[propval.common < 0.001] <- "***"

cname <- colnames(profc.common)


tclass <- data.frame(System = tissue.systems[cname],
  stringsAsFactors = F, row.names = cname)
profc.common.v <- profc.common
profc.common.v[profc.common.v > 0.1] <- 0.1
profc.common.v[profc.common.v < -0.1] <- -0.1


heapmap_protein <- pheatmap::pheatmap(profc.common.v,
  scale = "none", cluster_rows = F, cluster_cols = F,
  fontsize_row = 11, fontsize_col = 11, legend_labels = "beta",
  annotation_col = tclass,
  annotation_colors = ann_colors,
  breaks = enbrks,
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  display_numbers = display_matrix,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  file = "./results/meta_lm_GO/Figure_2F_proteins_notclusterrow_col_FDR0.01.pdf",
  height = 12, width = 9.2)
```

### 5.7.3 FDR 0.05

```{r}
common_up <- intersect(commonMols.filter$FDR005$macaca_uppro_meta,
  commonMols.filter$FDR005$macaca_upmrna_meta)
common_down <- intersect(commonMols.filter$FDR005$macaca_downmrna_meta,
  commonMols.filter$FDR005$macaca_downpro_meta)
common_all <- c(common_up, common_down)


# mRNA
tissue.systems.v <- sort(tissue.systems[names(pro.tissues)])

mrnafc.common <- Metamrna[common_all,
  substr(colnames(Metamrna), 1, 5) == "beta_"]
colnames(mrnafc.common) <- gsub("beta_", "", colnames(mrnafc.common))

mrnapval.common <- Metamrna[common_all,
  substr(colnames(Metamrna), 1, 7) == "Pvalue_"]
colnames(mrnapval.common) <- gsub("Pvalue_", "", colnames(mrnapval.common))

mrnafc.common <- mrnafc.common[, names(tissue.systems.v)]

mrnapval.common <- mrnapval.common[rownames(mrnafc.common), colnames(mrnafc.common)]

#
display_matrix <- matrix(" ", nrow(mrnapval.common), ncol(mrnapval.common))
display_matrix[mrnapval.common < 0.1] <- "."
display_matrix[mrnapval.common >= 0.01 & mrnapval.common < 0.05] <- "*"
display_matrix[mrnapval.common >= 0.001 & mrnapval.common < 0.01] <- "**"
display_matrix[mrnapval.common < 0.001] <- "***"

enbrks <- 0.1 * c(-1, -0.8, -0.4, -0.2, -0.1, -0.05, 0.05, 0.1, 0.2, 0.4, 0.8, 1)

cname <- colnames(mrnafc.common)

tclass <- data.frame(System = tissue.systems[cname], row.names = cname)
mrnafc.common.v <- mrnafc.common
mrnafc.common.v[mrnafc.common.v > 0.1] <- 0.1
mrnafc.common.v[mrnafc.common.v < -0.1] <- -0.1

ann_colors <- list(
  System = tissue.color)

heatmap_mrna <- pheatmap::pheatmap(mrnafc.common.v, scale = "none",
  cluster_rows = T, cluster_cols = F,
  annotation_colors = ann_colors,
  fontsize_row = 9, fontsize_col = 11,
  annotation_col = tclass,
  breaks = enbrks,
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  display_numbers = display_matrix,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  file = "./results/meta_lm_GO/Figure_2F_mrna_nocluster_col_FDR0.05.pdf",
  height = 16, width = 9.5)

mrna_order <- rownames(mrnafc.common.v)[heatmap_mrna$tree_row$order]
```

```{r}
# protein
profc.common <- Metapro[mrna_order,
  substr(colnames(Metapro), 1, 5) == "beta_"]
colnames(profc.common) <- gsub("beta_", "", colnames(profc.common))

propval.common <- Metapro[mrna_order,
  substr(colnames(Metapro), 1, 7) == "Pvalue_"]
colnames(propval.common) <- gsub("Pvalue_", "", colnames(propval.common))


profc.common <- profc.common[mrna_order, names(tissue.systems.v)]
propval.common <- propval.common[rownames(profc.common), colnames(profc.common)]

#
display_matrix <- matrix(" ", nrow(propval.common), ncol(propval.common))
display_matrix[propval.common < 0.1] <- "."
display_matrix[propval.common >= 0.01 & propval.common < 0.05] <- "*"
display_matrix[propval.common >= 0.001 & propval.common < 0.01] <- "**"
display_matrix[propval.common < 0.001] <- "***"

cname <- colnames(profc.common)


tclass <- data.frame(System = tissue.systems[cname],
  stringsAsFactors = F, row.names = cname)
profc.common.v <- profc.common
profc.common.v[profc.common.v > 0.1] <- 0.1
profc.common.v[profc.common.v < -0.1] <- -0.1


heapmap_protein <- pheatmap::pheatmap(profc.common.v,
  scale = "none", cluster_rows = F, cluster_cols = F,
  fontsize_row = 9, fontsize_col = 11, legend_labels = "beta",
  annotation_col = tclass,
  annotation_colors = ann_colors,
  breaks = enbrks,
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  display_numbers = display_matrix,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  file = "./results/meta_lm_GO/Figure_2F_proteins_notclusterrow_col_FDR0.05.pdf",
  height = 16, width = 9.2)
```

# 6 Figure 3 common aging metabolites

## 6.1 Figure 3A vocano mets

```{r,fig.width=7, fig.height=5.7, out.width="60%", fig.align="center",dpi = 144}
# metabolites

tmpdata <- Metamet
tmpdata$Pvalue <- tmpdata$MetaFDR
tmpdata$log2FC <- tmpdata$MetaBeta
tmpdata$log2FC[tmpdata$Pvalue == 1] <- 0

sortname <- short_met_names(tmpdata, outype = "vector")

rownames(tmpdata) <- sortname
tmpdata$ID <- sortname



pdf("./results/lm_metabolites/Figure_3A_vocano_common_metabolites.pdf", height = 5.7)
pp <- plot_DEflux(tmpdata[tmpdata$manyNA == FALSE, ],
  FCcutoff = beta_cutoff, xlab = "LM beta (30 tissues)",
  num.showlab = 5, alpha = 0.3,
  fixpointsize = 3, labSize = 6,
  ylab = bquote(~ -Log[10] ~ italic(FDR))) + xlim(c(-0.2, 0.2)) +
  ggtitle("Common changed metabolites across 30 tissues") +
  theme(axis.line = element_line(size = 1.0),
    axis.title.x = element_text(size = 18),
    axis.title.y = element_text(size = 18),
    axis.text.x = element_text(size = 16),
    axis.text.y = element_text(size = 16)
  )
print(pp)
dev.off()

# just for plot visualization
print(pp)
```

## 6.2 Figure 3B heatmap mets

```{r,fig.width=13, fig.height=10, out.width="80%", fig.align="center",dpi = 144}
# met0.01
commonmet <- c(commonMols.filter$FDR001$macaca_upmet_meta,
  commonMols.filter$FDR001$macaca_downmet_meta)
tissue.systems.v <- sort(tissue.systems[names(pro.tissues)])

metfc.common <- Metamet[is.element(Metamet$ID, commonmet) & Metamet$MetaFDR < 0.01,
  substr(colnames(Metamet), 1, 5) == "beta_"]
colnames(metfc.common) <- gsub("beta_", "", colnames(metfc.common))

metpval.common <- Metamet[is.element(Metamet$ID, commonmet) & Metamet$MetaFDR < 0.01,
  substr(colnames(Metamet), 1, 7) == "Pvalue_"]
colnames(metpval.common) <- gsub("Pvalue_", "", colnames(metpval.common))

metfc.common <- metfc.common[, names(tissue.systems.v)]

metpval.common <- metpval.common[rownames(metfc.common), colnames(metpval.common)]

#
display_matrix <- matrix(" ", nrow(metpval.common), ncol(metpval.common))
display_matrix[metpval.common < 0.1] <- "."
display_matrix[metpval.common >= 0.01 & metpval.common < 0.05] <- "*"
display_matrix[metpval.common >= 0.001 & metpval.common < 0.01] <- "**"
display_matrix[metpval.common < 0.001] <- "***"

enbrks <- 0.1 * c(-1, -0.8, -0.4, -0.2, -0.05, -0.02, 0.02, 0.05, 0.2, 0.4, 0.8, 1)
cname <- colnames(metfc.common)


tclass <- data.frame(System = tissue.systems[cname], row.names = cname)
metfc.common.v <- metfc.common
metfc.common.v[metfc.common.v > 0.1] <- 0.1
metfc.common.v[metfc.common.v < -0.1] <- -0.1

sortname <- short_met_names(metfc.common.v, outype = "vector")
rownames(metfc.common.v) <- sortname


heatmap_tmpmet <- pheatmap::pheatmap(metfc.common.v, scale = "none",
  cluster_rows = T, cluster_cols = F,
  fontsize_row = 9, fontsize_col = 10,
  annotation_col = tclass,
  annotation_colors = ann_colors,
  breaks = enbrks,
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  display_numbers = display_matrix,

  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  # file ='./results/lm_metabolites/Figure_3B_beta_met_nocluster_col_FDR001.pdf',
  height = 10, width = 13)

pdf("./results/lm_metabolites/Figure_3B_beta_met_nocluster_col_FDR001.pdf",
  height = 10, width = 13)
grid::grid.draw(heatmap_tmpmet$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(heatmap_tmpmet$gtable)
```

```{r}
# met0.05
commonmet <- c(commonMols.filter$FDR005$macaca_upmet_meta,
  commonMols.filter$FDR005$macaca_downmet_meta)
tissue.systems.v <- sort(tissue.systems[names(pro.tissues)])

metfc.common <- Metamet[is.element(Metamet$ID, commonmet) & Metamet$MetaFDR < 0.05,
  substr(colnames(Metamet), 1, 5) == "beta_"]
colnames(metfc.common) <- gsub("beta_", "", colnames(metfc.common))

metpval.common <- Metamet[is.element(Metamet$ID, commonmet) & Metamet$MetaFDR < 0.05,
  substr(colnames(Metamet), 1, 7) == "Pvalue_"]
colnames(metpval.common) <- gsub("Pvalue_", "", colnames(metpval.common))

metfc.common <- metfc.common[, names(tissue.systems.v)]

metpval.common <- metpval.common[rownames(metfc.common), colnames(metpval.common)]

#
display_matrix <- matrix(" ", nrow(metpval.common), ncol(metpval.common))
display_matrix[metpval.common < 0.1] <- "."
display_matrix[metpval.common >= 0.01 & metpval.common < 0.05] <- "*"
display_matrix[metpval.common >= 0.001 & metpval.common < 0.01] <- "**"
display_matrix[metpval.common < 0.001] <- "***"

enbrks <- 0.1 * c(-1, -0.8, -0.4, -0.2, -0.05, -0.02, 0.02, 0.05, 0.2, 0.4, 0.8, 1)
cname <- colnames(metfc.common)

tclass <- data.frame(System = tissue.systems[cname], row.names = cname)
metfc.common.v <- metfc.common
metfc.common.v[metfc.common.v > 0.1] <- 0.1
metfc.common.v[metfc.common.v < -0.1] <- -0.1

sortname <- short_met_names(metfc.common.v, outype = "vector")
rownames(metfc.common.v) <- sortname

pheatmap::pheatmap(metfc.common.v, scale = "none", cluster_rows = T, cluster_cols = F,
  fontsize_row = 10, fontsize_col = 12,
  annotation_col = tclass,
  annotation_colors = ann_colors,
  breaks = enbrks,
  angle_col = 45,
  treeheight_row = 20, treeheight_col = 20, legend = T,
  display_numbers = display_matrix,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11),
  file = "./results/lm_metabolites/Figure_3B_beta_met_nocluster_col_FDR005.pdf",
  height = 16, width = 13)
```

## 6.3 Figure 3C TS related protein

```{r,fig.width=7, fig.height=5.7, out.width="60%", fig.align="center",dpi = 144}
teron <- list()
for (i in 1:length(met.tissues)) {
  teron[[i]] <- met.tissues[[i]]["Testosterone sulfate", ]
}
names(teron) <- names(met.tissues)

# get_correlation protein
corTeron2pro <- get_corr(teron, pro.tissues.v)
corTeron2pro.Pvalue <- str2num(list_element_select(corTeron2pro, rownames(Metapro), "Pvalue"))
corTeron2pro.cor <- str2num(list_element_select(corTeron2pro, rownames(Metapro), "cor"))
corTeron2pro.Pvalue[is.na(corTeron2pro.Pvalue)] <- 1

x <- list(p = corTeron2pro.Pvalue)
xx.pro <- MetaDE.pvalue(x = x, meta.method = "AW")
tmpdata <- data.frame(ID = rownames(corTeron2pro.Pvalue),
  log2FC = rowMeans(corTeron2pro.cor, na.rm = T),
  Pvalue = as.vector(xx.pro$meta.analysis$FDR))
rownames(tmpdata) <- tmpdata$ID

# tmpdata$Pvalue[tmpdata$Pvalue < 1e-30] = 1e-30
pdf("./results/lm_metabolites/Figure_3C_Tesosteron_vs_pro_cor1.pdf", height = 5.7)
pp <- plot_DEflux(tmpdata, alpha = 0.5, num.showlab = 8,
  xlab = bquote("Correlation(mean 30 tissues)"), labSize = 5,
  ylab = bquote(~ -Log[10] ~ italic("FDR(meta analysis)")),
  title = "Testosterone sulfate related proteins") + xlim(c(-1.01, 1.01)) +
  theme(axis.line = element_line(size = 1.0),
    axis.title.x = element_text(size = 18),
    axis.title.y = element_text(size = 18),
    axis.text.x = element_text(size = 16),
    axis.text.y = element_text(size = 16))
print(pp)
dev.off()
sum(tmpdata$Pvalue < 0.05)
sum(tmpdata$Pvalue < 0.05 & tmpdata$log2FC > 0)

# just for plot visualization
print(pp)
```

## 6.4 Figure 3D-3E

```{r}
## in cytoscape and metascape enrichment
```

# 7. Figure 4 mouse,macaca,human

## 7.1 get data

### 7.1.1 mouse

```{r}

msdata <- loadRData("./data/Mus_agingNature_reads_list.Rdata")
sum(sapply(msdata$Mus_agingNature_pdata_list, nrow))
msdata$cpm <- list()
msdata$cpm.clin <- list()
for (i in 1:length(msdata$Mus_agingNature_reads_list)) {
  counts <- msdata$Mus_agingNature_reads_list[[i]]
  rownames(counts) <- toupper(rownames(counts))
  clin <- msdata$Mus_agingNature_pdata_list[[i]]
  rownames(clin) <- clin$`Sample name`
  # consider coding genes
  tmptype <- mousegeneInfo[rownames(counts), ]$Type
  idx <- tmptype == "protein_coding" |
    substr(tmptype, nchar(tmptype) - 4, nchar(tmptype)) == "_gene"
  idx[is.na(idx)] <- FALSE
  counts <- counts[idx, ]
  clin <- clin[colnames(counts), ]
  cpm <- get_CPM_counts(counts)
  cpm.nonormal <- get_CPM_nonormal(counts)
  # filtering
  idx <- rowMeans(cpm.nonormal) > 1 & rowSums(counts < 5) < ncol(counts) * 0.2
  cpm <- cpm[idx, ]

  clin$age <- as.numeric(clin$`characteristics: age`)
  clin$type <- clin$age # just for data hase the type field
  idx <- clin$age >= 3 & clin$`characteristics: sex` == "f"
  cpm <- cpm[, idx]
  clin <- clin[idx, ]
  clin$type <- clin$age

  # del outliers
  pcamrna <- prcomp(t(cpm), cor = F)
  vid <- !is.outliner(pcamrna$x[, 1])

  msdata$cpm[[i]] <- cpm[, vid]
  msdata$cpm.clin[[i]] <- as.data.frame(clin[vid, ])
}
names(msdata$cpm) <- names(msdata$Mus_agingNature_reads_list)
names(msdata$cpm.clin) <- names(msdata$Mus_agingNature_reads_list)

mean(sapply(msdata$cpm, nrow))

MetageneMouse <- get_tissue_DEgenes_lm(msdata$cpm, msdata$cpm.clin,
  tissue.systems = NULL)$MetaLimma

writetxt_forGPMM(MetageneMouse,
  "./results/compare_ms_mcc_human/metagenemouse_usingCPM_LM_Female.txt")
```

### 7.1.2 human

```{r}
load("./data/GETX.cpm_21_Female_tissues.RData")
mean(sapply(GTEX.cpm.v,nrow))

MetageneGETX <- get_tissue_DEgenes_lm(GTEX.cpm.v, GTEX.clin.v,
  tissue.systems = NULL)$MetaLimma
writetxt_forGPMM(MetageneGETX,
  "./results/compare_ms_mcc_human/metageneGTEX_usingCPM_Female_LM.txt")
```

## 7.2 common mRNAs

```{r}
betacut.mouse <- 0.007 # abs(sort(MetageneMouse$MetaBeta)[nrow(MetageneMouse)*0.1])
betacut.macaca <- 0.008 # abs(sort(Metamrna$MetaBeta)[nrow(Metamrna)*0.1])  or 0.2/25years
betacut.human <- 0.004 # abs(sort(MetageneGETX$MetaBeta)[nrow(MetageneGETX)*0.1]) #0.2/50years

commonMols.filter$FDR005$mouse_upmrna <-
  MetageneMouse$ID[MetageneMouse$MetaFDR < 0.05 & MetageneMouse$manyNA == FALSE &
    MetageneMouse$MetaBeta > betacut.mouse]

commonMols.filter$FDR005$mouse_downmrna <-
  MetageneMouse$ID[MetageneMouse$MetaFDR < 0.05 & MetageneMouse$manyNA == FALSE &
    MetageneMouse$MetaBeta < -betacut.mouse]

commonMols.filter$FDR005$human_upmrna <-
  MetageneGETX$ID[MetageneGETX$MetaFDR < 0.05 & MetageneGETX$manyNA == FALSE &
    MetageneGETX$MetaBeta > betacut.human]

commonMols.filter$FDR005$human_downmrna <-
  MetageneGETX$ID[MetageneGETX$MetaFDR < 0.05 & MetageneGETX$manyNA == FALSE &
    MetageneGETX$MetaBeta < -betacut.human]

sum(MetageneGETX$MetaFDR < 0.05 & MetageneGETX$manyNA == FALSE &
  MetageneGETX$MetaBeta > betacut.human)
sum(MetageneGETX$MetaFDR < 0.05 & MetageneGETX$manyNA == FALSE &
  MetageneGETX$MetaBeta < -betacut.human)

sum(Metamrna$MetaFDR < 0.05 & Metamrna$manyNA == FALSE &
  Metamrna$MetaBeta > betacut.macaca)
sum(Metamrna$MetaFDR < 0.05 & Metamrna$manyNA == FALSE &
  Metamrna$MetaBeta < -betacut.macaca)

sum(MetageneMouse$MetaFDR < 0.05 & MetageneMouse$manyNA == FALSE &
  MetageneMouse$MetaBeta > betacut.mouse)
sum(MetageneMouse$MetaFDR < 0.05 & MetageneMouse$manyNA == FALSE &
  MetageneMouse$MetaBeta < -betacut.mouse)

# write geneset for metascape analysis
upgenesets <- c(paste(MetageneMouse$ID[MetageneMouse$MetaFDR < 0.05 &
  MetageneMouse$manyNA == FALSE &
  MetageneMouse$MetaBeta > betacut.mouse],
collapse = ","),
paste(Metamrna$ID[Metamrna$MetaFDR < 0.05 &
  Metamrna$manyNA == FALSE &
  Metamrna$MetaBeta > betacut.macaca],
collapse = ","),
paste(MetageneGETX$ID[MetageneGETX$MetaFDR < 0.05 &
  MetageneGETX$manyNA == FALSE &
  MetageneGETX$MetaBeta > betacut.human],
collapse = ","))
names(upgenesets) <- c("Mouse", "Macaca", "Human")
upgenesets <- as.matrix(upgenesets, 3, 1)
writetxt_forGPMM(upgenesets,
  filename = "./results/compare_ms_mcc_human/up_genes_for_metascape.txt")


downgenesets <- c(paste(MetageneMouse$ID[MetageneMouse$MetaFDR < 0.05 &
  MetageneMouse$manyNA == FALSE &
  MetageneMouse$MetaBeta < -betacut.mouse],
collapse = ","),
paste(Metamrna$ID[Metamrna$MetaFDR < 0.05 &
  Metamrna$manyNA == FALSE &
  Metamrna$MetaBeta < -betacut.macaca],
collapse = ","),
paste(MetageneGETX$ID[MetageneGETX$MetaFDR < 0.05 &
  MetageneGETX$manyNA == FALSE &
  MetageneGETX$MetaBeta < -betacut.human],
collapse = ","))
names(downgenesets) <- c("Mouse", "Macaca", "Human")
downgenesets <- as.matrix(downgenesets, 3, 1)
writetxt_forGPMM(downgenesets,
  filename = "./results/compare_ms_mcc_human/down_genes_for_metascape.txt")
```

## 7.2 plot volcano_mrna_mouse macaca human

```{r,fig.width=7, fig.height=5.7, out.width="60%", fig.align="center",dpi = 144}

xxcomm <- intersect(intersect(MetageneMouse$ID, Metamrna$ID), MetageneGETX$ID)

range <- max(abs(MetageneMouse$MetaBeta), na.rm = T) + 0.02
pdf("./results/compare_ms_mcc_human/Figure_S_mRNA_mouse_vocano.pdf", height = 5.7)
p <- volcono_plot_species(MetageneMouse, "Aging related mRNA in mouse",
  FCcutoff = betacut.mouse) + xlim(c(-range, range))
print(p)
dev.off()
# just for htmloutput
print(p)

range <- max(abs(Metamrna$MetaBeta), na.rm = T) + 0.02
pdf("./results/compare_ms_mcc_human/Figure_S_mRNA_macaca_vocano.pdf", height = 5.7)
p <- volcono_plot_species(Metamrna, "Aging related mRNA in macaca",
  FCcutoff = betacut.macaca) + xlim(c(-range, range))
print(p)
dev.off()
# just for htmloutput
print(p)

range <- max(abs(MetageneGETX$MetaBeta), na.rm = T) + 0.02
pdf("./results/compare_ms_mcc_human/Figure_S_mRNA_human_vocano.pdf", height = 5.7)
p <- volcono_plot_species(MetageneGETX, "Aging related mRNA in human",
  FCcutoff = betacut.human) + xlim(c(-range, range))
print(p)
dev.off()
# just for htmloutput
print(p)
```

```{r,fig.width=8, fig.height=7, out.width="60%", fig.align="center",dpi = 144}
# ggvenn
p <- ggvenn::ggvenn(data = list(mrna.up.human = commonMols.filter$FDR005$human_upmrna,
  mrna.up.macaca = commonMols.filter$FDR005$macaca_upmrna_meta,
  mrna.up.mouse = commonMols.filter$FDR005$mouse_upmrna),
fill_color = c("#E64B35FF", "#3C5488FF", "green"),
show_percentage = F, stroke_size = 0.5,
stroke_alpha = 0.6, text_size = 9)
pdf("./results/compare_ms_mcc_human/ggvenn_cross_species_up_mrna.pdf", width = 8)
print(p)
dev.off()
# just for plot visualization
print(p)

p <- ggvenn::ggvenn(data = list(mrna.down.human = commonMols.filter$FDR005$human_downmrna,
  mrna.down.macaca = commonMols.filter$FDR005$macaca_downmrna_meta,
  mrna.down.mouse = commonMols.filter$FDR005$mouse_downmrna),
fill_color = c("#E64B35FF", "#3C5488FF", "green"),
show_percentage = F, stroke_size = 0.5,
stroke_alpha = 0.6, text_size = 9)
pdf("./results/compare_ms_mcc_human/ggvenn_cross_species_down_mrna.pdf", width = 8)
print(p)
dev.off()
# just for plot visualization
print(p)


intersect(intersect(commonMols.filter$FDR005$mouse_upmrna,
  commonMols.filter$FDR005$macaca_upmrna_meta),
commonMols.filter$FDR005$human_upmrna)

intersect(intersect(commonMols.filter$FDR005$mouse_downmrna,
  commonMols.filter$FDR005$macaca_downmrna_meta),
commonMols.filter$FDR005$human_downmrna)
```


## 7.4 enrichment mRNA across 3 species

```{r}
# slid run: this will take minutes
# Rscript compare_C2_GO_human_mouse_macaque.R
```

```{r,fig.width=8, fig.height=7, out.width="60%", fig.align="center",dpi = 144}
outpath <- "./results/compare_ms_mcc_human/"
p <- ggvenn::ggvenn(data = list(mrna.up.human = commonMols.filter$FDR005$human_upmrna,
  mrna.up.macaca = commonMols.filter$FDR005$macaca_upmrna_meta,
  mrna.up.mouse = commonMols.filter$FDR005$mouse_upmrna),
fill_color = c("#E64B35FF", "#3C5488FF", "green"),
show_percentage = F, stroke_size = 0.5,
stroke_alpha = 0.6, text_size = 9)
pdf(paste0(outpath, "/ggvenn_cross_species_up_mrna.pdf"), width = 8)
print(p)
dev.off()

# just for plot visualization
print(p)

p <- ggvenn::ggvenn(data = list(mrna.down.human = commonMols.filter$FDR005$human_downmrna,
  mrna.down.macaca = commonMols.filter$FDR005$macaca_downmrna_meta,
  mrna.down.mouse = commonMols.filter$FDR005$mouse_downmrna),
fill_color = c("#E64B35FF", "#3C5488FF", "green"),
show_percentage = F, stroke_size = 0.5,
stroke_alpha = 0.6, text_size = 9)
pdf(paste0(outpath, "/ggvenn_cross_species_down_mrna.pdf"), width = 8)
print(p)
dev.off()
# just for plot visualization
print(p)
```

```{r,fig.width=10.5, fig.height=7.2, out.width="90%", fig.align="center",dpi = 144}
c2_gene_sets <- msigdbr(species = "Homo sapiens", category = "C2")
m_t2g <- c2_gene_sets %>% dplyr::select(gs_name, gene_symbol)
C2_compare <- compare_GO_C2(commonMols.filter, m_t2g,
  outfile = "./results/compare_ms_mcc_human/log2ratio_commpare_C2.pdf")

# just for plot visualization
grid.arrange(arrangeGrob(grobs = c(C2_compare$plots.up, C2_compare$plots.down),
  ncol = 3, heights = c(3.5, 3.5),
  top = textGrob("log2(odds ratio) in C2 enrichment comparision",
    gp = gpar(fontface = "bold",  fontsize = 16))
))
```

## 7.5 GSEA comparison

```{r}
library(msigdbr)
c2_gene_sets <- msigdbr(species = "Homo sapiens", category = "C2")
gene_sets_c2 <- split(c2_gene_sets$gene_symbol, c2_gene_sets$gs_name)


vids <- intersect(intersect(MetageneMouse$ID, Metamrna$ID), MetageneGETX$ID)
tmpmouse <- MetageneMouse[vids, ]$MetaBeta
names(tmpmouse) <- vids
tmpmouse <- sort(tmpmouse, decreasing = TRUE)

tmpmacaca <- Metamrna[vids, ]$MetaBeta
names(tmpmacaca) <- vids
tmpmacaca <- sort(tmpmacaca, decreasing = TRUE)

tmphuman <- MetageneGETX[vids, ]$MetaBeta
names(tmphuman) <- vids
tmphuman <- sort(tmphuman, decreasing = TRUE)


gsea_mouse <- fgsea(pathways = gene_sets_c2, stats = tmpmouse)
gsea_macaca <- fgsea(pathways = gene_sets_c2, stats = tmpmacaca)
gsea_human <- fgsea(pathways = gene_sets_c2, stats = tmphuman)
rownames(gsea_mouse) <- gsea_mouse$pathway
rownames(gsea_macaca) <- gsea_macaca$pathway
rownames(gsea_human) <- gsea_human$pathway

writetxt(gsea_mouse[, 1:7], "./results/compare_ms_mcc_human/gsea_mouse.txt")
writetxt(gsea_macaca[, 1:7], "./results/compare_ms_mcc_human/gsea_macaca.txt")
writetxt(gsea_human[, 1:7], "./results/compare_ms_mcc_human/gsea_human.txt")
```

```{r,fig.width=3.2, fig.height=3.2, out.width="50%", fig.align="center",dpi = 144}

combined_data <- data.frame(
  pathway = gsea_mouse$pathway,  # both datasets have the same pathways
  NES_mouse = gsea_mouse$NES,
  NES_human = gsea_human$NES,
  NES_macaca = gsea_macaca$NES,
  Sigtype = (gsea_human$padj < 0.05) + 0,
  Sigtypemcc = (gsea_macaca$padj < 0.05) + 0
)

# Create density point plot
pdf("./results/compare_ms_mcc_human/NES_commpare_human_vs_mouse.pdf",
  width = 3.2, height = 3.2)
quadrant_percentages <- c("Q1" = sum(combined_data$NES_human > 0 &
  combined_data$NES_mouse > 0, na.rm = T),
"Q2" = sum(combined_data$NES_human < 0 &
  combined_data$NES_mouse > 0, na.rm = T),
"Q3" = sum(combined_data$NES_human < 0 &
  combined_data$NES_mouse < 0, na.rm = T),
"Q4" = sum(combined_data$NES_human > 0 &
  combined_data$NES_mouse < 0, na.rm = T))
quadrant_percentages <- round(quadrant_percentages / sum(quadrant_percentages) * 100, 2)

rho <- cor.test(combined_data$NES_mouse, combined_data$NES_human,
  method = "spearman")$estimate
p <- ggplot(combined_data, aes(x = NES_human, y = NES_mouse)) +
  geom_point(aes(color = Sigtype), size = 0.2 +
    combined_data$Sigtype * 0.2,
  alpha = 0.2 + combined_data$Sigtype * 0.2) +
  geom_density_2d(aes(color = ..level..),
    size = 0.3, alpha = 0.1) +  # Density
  labs(
    title = "NES human vs mouse",
    x = "NES (Human)",
    y = "NES (Mouse)"
  ) +
  theme_classic() +
  theme(legend.position = "none") + xlim(c(-3.3, 3.3)) + ylim(c(-3.3, 3.3)) +
  # Annotate each quadrant with the percentage of points
  annotate("text", x = 2.5, y = 2.7,
    label = paste0("Q1: ", quadrant_percentages["Q1"], "%"),
    size = 4, color = "black") +
  annotate("text", x = -2.5, y = 2.7,
    label = paste0("Q2: ", quadrant_percentages["Q2"], "%"),
    size = 4, color = "black") +
  annotate("text", x = -2.5, y = -2.7,
    label = paste0("Q3: ", quadrant_percentages["Q3"], "%"),
    size = 4, color = "black") +
  annotate("text", x = 2.5, y = -2.7,
    label = paste0("Q4: ", quadrant_percentages["Q4"], "%"),
    size = 4, color = "black") +
  annotate("text", x = 0, y = 3, label = paste("Rho =", round(rho, 2)),
    size = 4, color = "black")
print(p)
dev.off()
print(p)

# Create density point plot
pdf("./results/compare_ms_mcc_human/NES_commpare_human_vs_macaca.pdf",
  width = 3.2, height = 3.2)
quadrant_percentages <- c("Q1" = sum(combined_data$NES_human > 0 &
  combined_data$NES_macaca > 0, na.rm = T),
"Q2" = sum(combined_data$NES_human < 0 &
  combined_data$NES_macaca > 0, na.rm = T),
"Q3" = sum(combined_data$NES_human < 0 &
  combined_data$NES_macaca < 0, na.rm = T),
"Q4" = sum(combined_data$NES_human > 0 &
  combined_data$NES_macaca < 0, na.rm = T))
quadrant_percentages <- round(quadrant_percentages / sum(quadrant_percentages) * 100, 2)

rho <- cor.test(combined_data$NES_macaca, combined_data$NES_human,
  method = "spearman")$estimate
p <- ggplot(combined_data, aes(x = NES_human, y = NES_macaca)) +
  geom_point(aes(color = Sigtype), size = 0.2 +
    combined_data$Sigtype * 0.2,
  alpha = 0.2 + combined_data$Sigtype * 0.2) +
  geom_density_2d(aes(color = ..level..), size = 0.3, alpha = 0.1) +
  labs(
    title = "NES human vs macaca",
    x = "NES (Human)",
    y = "NES (Macaca)"
  ) +
  theme_classic() +
  theme(legend.position = "none") + xlim(c(-3.3, 3.3)) + ylim(c(-3.3, 3.3)) +
  # Annotate each quadrant with the percentage of points
  annotate("text", x = 2.5, y = 2.7,
    label = paste0("Q1: ", quadrant_percentages["Q1"], "%"),
    size = 4, color = "black") +
  annotate("text", x = -2.5, y = 2.7,
    label = paste0("Q2: ", quadrant_percentages["Q2"], "%"),
    size = 4, color = "black") +
  annotate("text", x = -2.5, y = -2.7,
    label = paste0("Q3: ", quadrant_percentages["Q3"], "%"),
    size = 4, color = "black") +
  annotate("text", x = 2.5, y = -2.7,
    label = paste0("Q4: ", quadrant_percentages["Q4"], "%"),
    size = 4, color = "black") +
  annotate("text", x = 0, y = 3,
    label = paste("Rho =", round(rho, 2)),
    size = 4, color = "black")
print(p)
dev.off()
print(p)

# Create density point plot
pdf("./results/compare_ms_mcc_human/NES_commpare_macaca_vs_mouse.pdf",
  width = 3.2, height = 3.2)
quadrant_percentages <- c("Q1" = sum(combined_data$NES_macaca > 0 &
  combined_data$NES_mouse > 0, na.rm = T),
"Q2" = sum(combined_data$NES_macaca < 0 &
  combined_data$NES_mouse > 0, na.rm = T),
"Q3" = sum(combined_data$NES_macaca < 0 &
  combined_data$NES_mouse < 0, na.rm = T),
"Q4" = sum(combined_data$NES_macaca > 0 &
  combined_data$NES_mouse < 0, na.rm = T))
quadrant_percentages <- round(quadrant_percentages / sum(quadrant_percentages) * 100, 2)

rho <- cor.test(combined_data$NES_mouse, combined_data$NES_macaca,
  method = "spearman")$estimate
p <- ggplot(combined_data, aes(x = NES_macaca, y = NES_mouse)) +
  geom_point(aes(color = Sigtypemcc),
    size = 0.2 + combined_data$Sigtypemcc * 0.2,
    alpha = 0.2 + combined_data$Sigtypemcc * 0.2) +
  geom_density_2d(aes(color = ..level..), size = 0.3, alpha = 0.1) +  # Density contours
  labs(
    title = "NES macaca vs mouse",
    x = "NES (Macaca)",
    y = "NES (Mouse)"
  ) +
  theme_classic() +
  theme(legend.position = "none") + xlim(c(-3.3, 3.3)) + ylim(c(-3.3, 3.3)) +
  # Annotate each quadrant with the percentage of points
  annotate("text", x = 2.5, y = 2.7,
    label = paste0("Q1: ", quadrant_percentages["Q1"], "%"),
    size = 4, color = "black") +
  annotate("text", x = -2.5, y = 2.7,
    label = paste0("Q2: ", quadrant_percentages["Q2"], "%"),
    size = 4, color = "black") +
  annotate("text", x = -2.5, y = -2.7,
    label = paste0("Q3: ", quadrant_percentages["Q3"], "%"),
    size = 4, color = "black") +
  annotate("text", x = 2.5, y = -2.7,
    label = paste0("Q4: ", quadrant_percentages["Q4"], "%"),
    size = 4, color = "black") +
  annotate("text", x = 0, y = 3,
    label = paste("Rho =", round(rho, 2)),
    size = 4, color = "black")
print(p)
dev.off()
print(p)
```

## 7.6 protein_ms_macaca

```{r}
xx <- readxl::excel_sheets("./data/mouse_proteome_aging.xlsx")
mousepro <- list()
for (i in 1:length(xx)) {
  mousepro[[i]] <- as.data.frame(readxl::read_excel("./data/mouse_proteome_aging.xlsx",
    sheet = xx[i]))
}
names(mousepro) <- xx
Metapromouse <- mousepro$`Cross-tissue effects`
idx <- sort.int(Metapromouse$age_qval, decreasing = F, index.return = T)$ix
Metapromouse <- Metapromouse[idx, ]
Metapromouse$symbol <- toupper(Metapromouse$symbol)
Metapromouse <- Metapromouse[!duplicated(Metapromouse$symbol), ]

commonMols.filter$FDR005$mouse_uppro_meta <-
  unique(toupper(Metapromouse$symbol[Metapromouse$age_qval < 0.05 &
    Metapromouse$age_effect > 0]))
commonMols.filter$FDR005$mouse_downpro_meta <-
  unique(toupper(Metapromouse$symbol[Metapromouse$age_qval < 0.05 &
    Metapromouse$age_effect < 0]))


# write geneset for metascape analysis
upgenesets <- c(paste(commonMols.filter$FDR005$mouse_uppro_meta, collapse = ","),
  paste(commonMols.filter$FDR005$macaca_uppro_meta, collapse = ","))
names(upgenesets) <- c("Mouse", "Macaca")
upgenesets <- as.matrix(upgenesets, 2, 1)
writetxt_forGPMM(upgenesets,
  filename = "./results/compare_ms_mcc_human/up_pro_for_metascape.txt")


downgenesets <- c(paste(commonMols.filter$FDR005$mouse_downpro_meta, collapse = ","),
  paste(commonMols.filter$FDR005$macaca_downpro_meta, collapse = ","))
names(downgenesets) <- c("Mouse", "Macaca")
downgenesets <- as.matrix(downgenesets, 2, 1)
writetxt_forGPMM(downgenesets,
  filename = "./results/compare_ms_mcc_human/down_pro_for_metascape.txt")
```

```{r}
tmp <- Metapromouse # mousepro$`Cross-tissue effects`
tmp$log2FC <- tmp$age_effect
tmp$Pvalue <- tmp$age_qval
tmp$ID <- toupper(tmp$symbol)
idx <- sort.int(tmp$age_qval, decreasing = F, index.return = T)$ix
tmp <- tmp[idx, ]
vid <- !duplicated(tmp$ID)
tmp <- tmp[vid, ]

pdf("./results/compare_ms_mcc_human/FigureS_volcano_pro_mouse.pdf")
p <- plot_DEflux(tmp, num.showlab = 5, title = "Aging related proteins in mouse",
  xlab = c("Aging effect (12 tissues)"),
  fixpointsize = 3, labSize = 6,
  ylab = bquote(~ -Log[10] ~ italic("FDR")))
print(p)
dev.off()
```

```{r,fig.width=8, fig.height=7, out.width="60%", fig.align="center",dpi = 144}
p <- ggvenn::ggvenn(data = list(pro.up.macaca = commonMols.filter$FDR005$macaca_uppro_meta,
  pro.down.macaca = commonMols.filter$FDR005$macaca_downpro_meta,
  pro.up.mouse = commonMols.filter$FDR005$mouse_uppro_meta,
  pro.down.mouse = commonMols.filter$FDR005$mouse_downpro_meta),
fill_color = c("darkorchid1", "yellow", "green", "cornflowerblue"),
show_percentage = F, stroke_size = 0.5,
stroke_alpha = 0.6, text_size = 9)
pdf("./results/compare_ms_mcc_human/ggvenn_pro_macaca_mouse.pdf", width = 8)
print(p)
dev.off()
intersect(commonMols.filter$FDR005$macaca_uppro_meta,commonMols.filter$FDR005$mouse_uppro_meta)
intersect(commonMols.filter$FDR005$macaca_downpro_meta,commonMols.filter$FDR005$mouse_downpro_meta)

print(p)
```

# 8 Tissue-specific aging mols

## 8.1 protein

```{r,fig.width=12, fig.height=8, out.width="90%", fig.align="center",dpi = 144}
tissue_aging_markers <- list()
tissue_aging_markers$pro <- list()
tmpout <- get_tissue_specific_aging_markers(Metapro,
  tissue.systems, tissue.color,
  beta_cutoff = 0.04, topnum = 1000,
  outfile =
    "./results/meta_lm_GO/heatmap_tissue_tissue_aging_markers_pro_up_full.pdf",
  direction = "UP")
tissue_aging_markers$pro$up <- tmpout$makers

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(tmpout$heatmap))
grid.draw(heatmap_grob)
```

```{r,fig.width=12, fig.height=8, out.width="90%", fig.align="center",dpi = 144}
tmpout <- get_tissue_specific_aging_markers(Metapro,
  tissue.systems, tissue.color,
  beta_cutoff = 0.04, topnum = 1000,
  outfile =
    "./results/meta_lm_GO/heatmap_tissue_tissue_aging_markers_pro_down_full.pdf",
  direction = "down")

tissue_aging_markers$pro$down <- tmpout$makers

# write
tissue_markers.pro <- write_tissue_specific_aging_markers(tissue_aging_markers,
                                                          Metapro,'pro',
      "./results/meta_lm_GO/Tissue_specific_aging_markers_protein.xlsx")

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(tmpout$heatmap))
grid.draw(heatmap_grob)
```

## 8.2 mrna

```{r,fig.width=12, fig.height=8, out.width="90%", fig.align="center",dpi = 144}
tissue_aging_markers$mrna <- list()
tmpout <- get_tissue_specific_aging_markers(Metamrna,
  tissue.systems, tissue.color,
  beta_cutoff = 0.04, topnum = 1000,
  outfile =
    "./results/meta_lm_GO/heatmap_tissue_tissue_aging_markers_mrna_up_full.pdf",
  direction = "UP")

tissue_aging_markers$mrna$up <- tmpout$makers

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(tmpout$heatmap))
grid.draw(heatmap_grob)
```

```{r,fig.width=12, fig.height=8, out.width="90%", fig.align="center",dpi = 144}
tmpout <- get_tissue_specific_aging_markers(Metamrna,
  tissue.systems, tissue.color,
  beta_cutoff = 0.04, topnum = 1000,
  outfile =
    "./results/meta_lm_GO/heatmap_tissue_tissue_aging_markers_mrna_down_full.pdf",
  direction = "down")

tissue_aging_markers$mrna$down <- tmpout$makers

tissue_markers.mrna <- write_tissue_specific_aging_markers(tissue_aging_markers,
                                                          Metamrna,'mrna',
      "./results/meta_lm_GO/Tissue_specific_aging_markers_mRNA.xlsx")

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(tmpout$heatmap))
grid.draw(heatmap_grob)
```


## 8.3 mets

```{r,fig.width=12, fig.height=8, out.width="90%", fig.align="center",dpi = 144}
tissue_aging_markers$mets <- list()
tmpout <- get_tissue_specific_aging_markers(Metamet,
  tissue.systems, tissue.color,
  beta_cutoff = 0.04, topnum = 1000,
  outfile =
    "./results/meta_lm_GO/heatmap_tissue_tissue_aging_markers_mets_up_full.pdf",
  direction = "UP")

tissue_aging_markers$mets$up <- tmpout$makers

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(tmpout$heatmap))
grid.draw(heatmap_grob)
```

```{r,fig.width=12, fig.height=8, out.width="90%", fig.align="center",dpi = 144}
tmpout <- get_tissue_specific_aging_markers(Metamet,
  tissue.systems, tissue.color,
  beta_cutoff = 0.04, topnum = 1000,
  outfile =
    "./results/meta_lm_GO/heatmap_tissue_tissue_aging_markers_mets_down_full.pdf",
  direction = "down")

tissue_aging_markers$mets$down <- tmpout$makers

#write
tissue_markers.mets <- write_tissue_specific_aging_markers(tissue_aging_markers,
                                                          Metamet,'mets',
      "./results/meta_lm_GO/Tissue_specific_aging_markers_metabolite.xlsx")

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(tmpout$heatmap))
grid.draw(heatmap_grob)
```

```{r}
my_func <- function(x) {
  y <- c()
  for (i in 1:length(x)) {
    y <- c(y, x[[i]])
  }
  y <- unique(y)
  return(y)
}
length(my_func(tissue_aging_markers$pro$up))
length(my_func(tissue_aging_markers$pro$down))
length(my_func(tissue_aging_markers$mrna$up))
length(my_func(tissue_aging_markers$mrna$down))
length(my_func(tissue_aging_markers$mets$up))
length(my_func(tissue_aging_markers$mets$down))

sum(sapply(tissue_aging_markers$pro$up, length))
sum(sapply(tissue_aging_markers$pro$down, length))
sum(sapply(tissue_aging_markers$mrna$up, length))
sum(sapply(tissue_aging_markers$mrna$down, length))
sum(sapply(tissue_aging_markers$mets$up, length))
sum(sapply(tissue_aging_markers$mets$down, length))

```

# 9 Figure 5 trajactory and different types

## 9.1 Figure 5A

```{r}
set.seed(2025520)
tissues <- names(pro.tissues.v)
promet.tissues <- list()
promet.tissues.info <- list()
promet.tissues.Z <- list()
for (i in 1:length(tissues)) {
  tt <- tissues[i]
  thispro <- pro.tissues.v[[tt]]
  thispro <- thispro[rowSums(is.na(thispro)) < 1 / 3 * ncol(thispro), ]
  thismet <- met.tissues[[tt]]
  rownames(thismet) <- paste0("met_", rownames(thismet))
  thispromet <- rbind2(thispro, thismet)
  thisinfo <- pro.tissues.info[[tt]]
  thisinfo <- thisinfo[colnames(thispromet), ]
  thispromet.median <- t(aggregate(t(thispromet), by = list(thisinfo$stage),
    FUN = median, na.rm = T))
  mstd <- standardise_matrix(thispromet.median)
  promet.tissues.Z[[i]] <- mstd
  promet.tissues[[i]] <- thispromet
  promet.tissues.info[[i]] <- thisinfo
}
names(promet.tissues) <- tissues
names(promet.tissues.Z) <- tissues
names(promet.tissues.info) <- tissues
```

```{r}
promet.tissues.Z.frame.t <- promet.tissues.Z
for (i in 1:length(promet.tissues.Z)) {
  promet.tissues.Z.frame.t[[i]] <- as.data.frame(t(promet.tissues.Z[[i]]))
}
names(promet.tissues.Z.frame.t) <- names(pro.tissues.v)
promet.whole.Z <- t(as.matrix(rbindlist(promet.tissues.Z.frame.t, fill = T)))
colnames(promet.whole.Z) <- paste0(rep(tissues, each = 4), "_", rep(1:4, times = 30))
promet.whole.Z.info <- data.frame(tissue = rep(tissues, each = 4), stringsAsFactors = F,
  stage = rep(1:4, times = 30))
```

```{r}
promet.whole.Z.v <- promet.whole.Z[rowSums(is.na(promet.whole.Z)) < 0.5 * 120, ]
promet.whole.Z.mean <- t(aggregate(t(promet.whole.Z.v),
  by = list(promet.whole.Z.info$stage), FUN = mean, na.rm = T))
promet.whole.Z.mean  <- promet.whole.Z.mean[-1, ]
promet.whole.Z.mean.v <- promet.whole.Z.mean[rowSums(is.na(promet.whole.Z.mean))  == 0, ]
promet.whole.Z.mean.eset <- new("ExpressionSet", exprs = promet.whole.Z.mean.v)
dim(promet.whole.Z.mean.v)
sum(substr(rownames(promet.whole.Z.mean.v), 1, 4) == "met_")
```

```{r,fig.width=8, fig.height=4, out.width="100%", fig.align="center",dpi = 144}
# run mfuzz promet whole
mfuzz.promet.whole <- mfuzz(promet.whole.Z.mean.eset, c = 8, m = 1.5)
# this load is to reproduce our results in Figure 5A "mfuzz.promet.whole"
# if you do not use this load, it is ok, the only diff is the order of clusters
# e.g. you may have the cluster 1 is corresponding to our cluster 5,
#      but the trajectories are the same
load("./data/mfuzz.promet.whole.repoduce.Rdata")

pdf("./results/trajactory/Figure5A.pdf", width = 8, height = 4)
mfuzz.plot(promet.whole.Z.mean.eset, mfuzz.promet.whole, mfrow = c(2, 4),
  time.labels = c("Juvenile", "Young_adult", "Middle_aged",
    "Elderly"), new.window = F)
dev.off()

## just for plot visualization
mfuzz.plot(promet.whole.Z.mean.eset, mfuzz.promet.whole, mfrow = c(2, 4),
  time.labels = c("Juvenile", "Young_adult", "Middle_aged",
    "Elderly"), new.window = F)
```

## 9.2 Figure 5B

```{r,fig.width=11, fig.height=7, out.width="100%", fig.align="center",dpi = 144}
tissue_trj_promet <- get_trj_data(promet.tissues, promet.tissues.info,
  promet.tissues.Z, mfuzz.promet.whole,
  tissue.systems, tissue.color,
  outfile =
    "./results/trajactory/Figure5B_tissue_trajectory_prometv1.pdf",
  center = 1)
tissue_trajectory_matrix <- tissue_trj_promet$tissue_trajectory_matrix
tissue_names <- names(promet.tissues)


# just for plot visualization
grid.arrange(arrangeGrob(grobs = tissue_trj_promet$plot, ncol = 4, heights = c(4, 4),
  top = textGrob("Protein & metabolites aging trajectory in 30 tissues",
    gp = gpar(fontface = "bold",  fontsize = 20)),
  left = textGrob("Z values", gp = gpar(fontface = "bold",  fontsize = 18), rot = 90)))
```

## 9.3 promet heatmap clustering

### 9.3.1 using consistent clustering K-means

```{r}
consclass <- runclustering(tissue_trajectory_matrix,
  title = "./results/trajactory/consclass_tissue_trajectory_matrix")


tissueclass <- data.frame(cons3 = consclass[[3]]$consensusClass,
  cons4 = consclass[[4]]$consensusClass,
  type_by_cons = rep("Undefined", length(pro.tissues.v)),
  row.names = names(consclass[[3]]$consensusClass))
tissueclass$cons3[tissueclass$cons3 == 3] <- 2
tissueclass$cons4[tissueclass$cons4 == 3] <- 1
tissueclass$cons4[tissueclass$cons4 == 4] <- 2
tissueclass$type_by_cons[tissueclass$cons3 == 2 & tissueclass$cons4 == 2] <- "Type I"
tissueclass$type_by_cons[tissueclass$cons3 == 1 & tissueclass$cons4 == 1] <- "Type II"
```

### 9.3.2 using hc trajactory

```{r,fig.width=10, fig.height=12, out.width="85%", fig.align="center",dpi = 144}
enbrks <- 1.5 * c(-1, -0.8, -0.6, -0.4, -0.2, -0.1, -0.05, -0.02,
  0.02, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1)

col_fun <- colorRamp2(
  breaks = enbrks,
  colors = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(length(enbrks))
)

row_groups <- c(rep(paste0("C", 1:8), each = 4))

col_dend <- hclust(dist(t(tissue_trajectory_matrix),
  method = "euclidean"), method = "complete")

col_clusters <- cutree(col_dend, k = 2)

pdf("./results/trajactory/FigureXC_heatmap_based_trajectory_promot_small_border1.pdf",
  width = 10, height = 12)

pp <- Heatmap(
  tissue_trajectory_matrix,
  name = "Z-value",
  col = col_fun,
  cluster_rows = FALSE,
  cluster_columns = col_dend,
  column_split = 2,
  row_split = row_groups,

  column_title = "Trajectories of protein & metabolites",
  row_names_gp = gpar(fontsize = 12),
  column_names_gp = gpar(fontsize = 12),

  column_names_rot = 45,
  border_gp = gpar(col = "gray50", lwd = 1),
  column_dend_height = unit(10, "mm"),

)
print(pp)
dev.off()

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(pp))
grid.draw(heatmap_grob)
```

```{r}
enbrks <- 1.5 * c(-1, -0.8, -0.6, -0.4, -0.2, -0.1, -0.05, -0.02, 0.02,
  0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1)
dent <- pheatmap::pheatmap(tissue_trajectory_matrix,
  scale = "none", cluster_rows = F, cutree_cols = 2, cutree_rows = 8,
  main = "Heatmap of molecular aging trajectory in 30 tissues",
  height = 7, width = 6, angle_col = 45,
  fontsize_row = 8, fontsize_col = 8, breaks = enbrks,
  treeheight_row = 20, treeheight_col = 20,
  file = "./results/trajactory/FigureSX_pheatmap_based_trajectory.pdf",
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(15))

dendcol <- as.dendrogram(dent$tree_col)

labelColors <- c("#3C5488FF", "#E64B35FF")

clusMember <- cutree(dent$tree_col, 2)

# plot(clusDendro)
class1 <- which(clusMember == 2)
class2 <- which(clusMember == 1)
tissueType <- data.frame(tissue = c(tissue_names[class1], tissue_names[class2]),
  stringsAsFactors = F,
  type_by_hc = c(rep("Type I", length(class1)), rep("Type II", length(class2))))
rownames(tissueType) <- tissueType$tissue
tissueType <- cbind(tissueType, tissueclass[rownames(tissueType), ])
```

### 9.3.3 using MAA hc

```{r,fig.width=7, fig.height=5, out.width="60%", fig.align="center",dpi = 144}
promet.norm.eset.stand.matrix <- as.matrix(promet.whole.Z.mean.eset)
variability <- rep(0, 8)
for (j in 1:8) {
  tgene <- names(mfuzz.promet.whole$cluster)[mfuzz.promet.whole$cluster == j]
  tgene <- intersect(tgene, rownames(promet.norm.eset.stand.matrix))
  bx <- t(t(promet.norm.eset.stand.matrix[tgene, ]) - mfuzz.promet.whole$centers[j, ])
  variability[j] <- mean(sqrt(rowSums(bx^2) / (ncol(bx) - 1)))
}
tmpdata <- data.frame(amplitude = as.vector(mfuzz.promet.whole$centers[, 4] -
  mfuzz.promet.whole$centers[, 1]),
stringsAsFactors = F,
variability = variability,
class = paste0("Cluster ", 1:8))
pdf(file = "./results/trajactory/FigureSX_cluster_amplitude_variability_8_promet.pdf",
  width = 7, height = 5)
pp <- ggplot(tmpdata, aes(variability, amplitude)) + geom_point(size = 4) +
  lghplot.addtheme(size = 18) +
  geom_text_repel(aes(label = class), size = 7) +
  theme(axis.line = element_line(size = 1.0)) +
  ggtitle("Figure SX cluster_amplitude_variability")
print(pp)
dev.off()

# just for plot visualization
print(pp)
```

```{r,fig.width=16, fig.height=9, out.width="100%", fig.align="center",dpi = 144}
tissues <- names(promet.tissues.Z)
clusterdist <- matrix(0, length(tissues), 8)
clusteramplitude <- matrix(0, length(tissues), 8)
clusteramplitude_xx <- matrix(0, length(tissues), 8)
for (i in 1:length(tissues)) {
  mstd <-   promet.tissues.Z[[i]]
  for (j in 1:8) {
    tgene <- names(mfuzz.promet.whole$cluster)[mfuzz.promet.whole$cluster == j]
    tgene <- intersect(tgene, rownames(mstd))
    bx <- t(t(mstd[tgene, ]) - mfuzz.promet.whole$centers[j, ])
    clusterdist[i, j] <- mean(sqrt(rowSums(bx^2) / (ncol(bx) - 1)), na.rm = T)
    if (length(tgene) < 2) {
      clusteramplitude[i, j] <- NA
      next
    }
    tamp <- colMeans(mstd[tgene, ], na.rm = T)
    clusteramplitude[i, j] <- abs(tamp[4] - tamp[1])
    clusteramplitude_xx[i, j] <- tamp[4] - tamp[1]
  }
}
rownames(clusteramplitude) <- tissues
rownames(clusteramplitude_xx) <- tissues
rownames(clusterdist) <- tissues

px <- list()
for (j in 1:8) {
  tmpdata <- data.frame(amplitude = clusteramplitude_xx[, j],
    stringsAsFactors = F,
    variability = clusterdist[, j],
    class = tissues,
    tissue.systems = tissue.systems)
  px[[j]] <- ggplot(tmpdata, aes(variability, amplitude, color = tissue.systems)) +
    geom_point(size = 6) + lghplot.addtheme(size = 14) +
    scale_color_npg() + scale_fill_aaas() +
    geom_text_repel(aes(label = class), size = 3, box.padding = 0.5) +
    theme(axis.line = element_line(size = 1.5)) + ggtitle(paste0("Cluster ", j)) +
    xlab("") + ylab("")
}

pdf(file = "./results/trajactory/FigureXS_cluster8_tissues_amplitude_variability.pdf",
  height = 9, width = 16)
grid.arrange(arrangeGrob(grobs = px, ncol = 4, margin = c(1, 1, 1, 1),
  top = textGrob("Molecular alteration amplitude of each cluster in 30 tissues",
    gp = gpar(fontface = "bold",  fontsize = 24)),
  bottom = textGrob("Molecular alteration variability",
    gp = gpar(fontface = "bold",  fontsize = 24)),
  left = textGrob("Molecular alteration amplitude(MAA)",
    gp = gpar(fontface = "bold",  fontsize = 24),
    rot = 90)))
dev.off()

# just for plot visualization
grid.arrange(arrangeGrob(grobs = px, ncol = 4, margin = c(1, 1, 1, 1),
  top = textGrob("Molecular alteration amplitude of each cluster in 30 tissues",
    gp = gpar(fontface = "bold",  fontsize = 24)),
  bottom = textGrob("Molecular alteration variability",
    gp = gpar(fontface = "bold",  fontsize = 24)),
  left = textGrob("Molecular alteration amplitude(MAA)",
    gp = gpar(fontface = "bold",  fontsize = 24),
    rot = 90)))
```

```{r,fig.width=3.5, fig.height=4, out.width="60%", fig.align="center",dpi = 144}
tmpaa <- tissue_trj_promet$MAA
colnames(tmpaa) <- paste0("C", 1:8)
dent.MAA <- pheatmap::pheatmap(tmpaa, scale = "none", height = 4, width = 3.5,
  main  = "Aging types in solid tissues using MAA",
  fontsize_row = 7, fontsize_col = 7,
  treeheight_row = 20, treeheight_col = 20,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(15)
)


dendrow <- as.dendrogram(dent.MAA$tree_row)



clusMember <- cutree(dent.MAA$tree_row, 2)


class1 <- which(clusMember == 1)
class2 <- which(clusMember == 2)


tissueType.MAA <- data.frame(tissue = c(tissue_names[class1], tissue_names[class2]),
  stringsAsFactors = F,
  type_by_hc = c(rep("Type II", length(class1)),
    rep("Type I", length(class2))))
rownames(tissueType.MAA) <- tissueType.MAA$tissue

tissueType$type_by_MAA <- tissueType.MAA[rownames(tissueType), ]$type_by_hc

pdf("./results/trajactory/FigureX_promet_heatmap_tissue_aging_byMAA.pdf",
  height = 4, width = 3.5)
grid::grid.draw(dent.MAA$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(dent.MAA$gtable)
```

### 9.3.4 define type by three methods

```{r}
thistype <- rep("Undefined", nrow(tissueType))
thistype[tissueType$type_by_MAA == "Type I" &
  tissueType$type_by_cons == "Type I" &
  tissueType$type_by_hc == "Type I"] <- "Type I"
thistype[tissueType$type_by_MAA == "Type II" &
  tissueType$type_by_cons == "Type II" &
  tissueType$type_by_hc == "Type II"] <- "Type II"
tissueType$type <- thistype
tissueType$class <- thistype
tissueType$color <- tissueType$type
tissueType$color[tissueType$color == "Type I"] <- "#FF9289"
tissueType$color[tissueType$color == "Type II"] <- "#00C7FF"
tissueType$color[tissueType$color == "Undefined"] <- "#DAB700"
tissueType_promet <- tissueType

writetxt_forGPMM(tissueType_promet, "./results/trajactory/tissueType_promet.txt")
```

```{r,fig.width=7, fig.height=7, out.width="60%", fig.align="center",dpi = 144}
# pca
tissue_pca <- prcomp(t(tissue_trajectory_matrix), cor = F)
perc_tissue_pca <- 100 * summary(tissue_pca)$importance
tissue_names <- names(promet.tissues)
pdf("./results/trajactory/FigureSX_promet_PCA.pdf", width = 7, height = 7)
tmpType <- tissueType[tissue_names, ]$type
pp <- ggplot(, aes(tissue_pca$x[, 1], tissue_pca$x[, 2], color = tmpType)) +
  geom_point(size = 5, alpha = 0.6) +
  theme_classic() + lghplot.addtheme(legend.position = "top") +
  # stat_ellipse(lwd=1,level = 0.95) +
  geom_text_repel(aes(label = tissue_names), size = 5,
    box.padding = 0.5, face = "bold") +
  xlab(paste("PC1(", as.character(round(perc_tissue_pca[2, 1], 1)), "%)", sep = "")) +
  ylab(paste("PC2(", as.character(round(perc_tissue_pca[2, 2], 1)), "%)", sep = "")) +
  scale_color_manual(values = c("#FF6347", "#1E90FF", "#DAB700")) +
  theme(legend.text = element_text(size = 12, face = "bold")) +
  labs(title = "Pro & met trajectory PCA")
print(pp)
dev.off()

# just for plot visualization
print(pp)
```

## 9.4 using mrna for for combine

### 9.4.1 mrna mfuzz trajactory

```{r}
tissues <- names(mrna.tissues)
mrna.tissues.Z <- list()
for (i in 1:length(tissues)) {
  tt <- tissues[i]
  thismrna <- mrna.tissues[[tt]]
  thisinfo <- mrna.tissues.info[[tt]]
  thismrna.median <- t(aggregate(t(thismrna), by = list(thisinfo$stage),
    FUN = median, na.rm = T))
  mstd <- standardise_matrix(thismrna.median)
  mrna.tissues.Z[[i]] <- mstd
}
names(mrna.tissues.Z) <- tissues
```

```{r}
mrna.tissues.Z.frame.t <- mrna.tissues.Z
for (i in 1:length(mrna.tissues.Z)) {
  mrna.tissues.Z.frame.t[[i]] <- as.data.frame(t(mrna.tissues.Z[[i]]))
}
names(mrna.tissues.Z.frame.t) <- names(mrna.tissues.Z)
mrna.whole.Z <- t(as.matrix(rbindlist(mrna.tissues.Z.frame.t, fill = T)))
colnames(mrna.whole.Z) <- paste0(rep(tissues, each = 4), "_", rep(1:4, times = 30))
mrna.whole.Z.info <- data.frame(tissue = rep(tissues, each = 4), stringsAsFactors = F,
  stage = rep(1:4, times = 30))
```

```{r}
mrna.whole.Z.v <- mrna.whole.Z[rowSums(is.na(mrna.whole.Z)) < 0.5 * 120, ]
mrna.whole.Z.mean <- t(aggregate(t(mrna.whole.Z.v),
  by = list(mrna.whole.Z.info$stage), FUN = mean, na.rm = T))
mrna.whole.Z.mean  <- mrna.whole.Z.mean[-1, ]
mrna.whole.Z.mean.v <- mrna.whole.Z.mean[rowSums(is.na(mrna.whole.Z.mean))  == 0, ]
mrna.whole.Z.mean.eset <- new("ExpressionSet", exprs = mrna.whole.Z.mean.v)
```

```{r,fig.width=8, fig.height=4, out.width="100%", fig.align="center",dpi = 144}
set.seed(2025520)
mfuzz.mrna.whole <- mfuzz(mrna.whole.Z.mean.eset, c = 8, m = 1.5)
pdf("./results/trajactory_combine/mrna_mfuzz.pdf", width = 8, height = 4)
mfuzz.plot(mrna.whole.Z.mean.eset, mfuzz.mrna.whole, mfrow = c(2, 4),
  time.labels = c("Juvenile", "Young_adult", "Middle_aged",
    "Elderly"), new.window = F)
dev.off()

# just for plot visualization
mfuzz.plot(mrna.whole.Z.mean.eset, mfuzz.mrna.whole, mfrow = c(2, 4),
  time.labels = c("Juvenile", "Young_adult", "Middle_aged",
    "Elderly"), new.window = F)
```

## 9.5 Figure 5C-D use all omics data

### 9.5.1 construct data

```{r,fig.width=11, fig.height=7, out.width="100%", fig.align="center",dpi = 144}
# get data
tissue_trj_mrna <- get_trj_data(mrna.tissues, mrna.tissues.info,
  mrna.tissues.Z, mfuzz.mrna.whole,
  tissue.systems, tissue.color,
  outfile =
    "./results/trajactory_combine/FigureSX_tissue_trajectory_mrna.pdf",
  center = 1)

tissue_trj_promet1 <- get_trj_data(promet.tissues, promet.tissues.info,
  promet.tissues.Z, mfuzz.promet.whole,
  tissue.systems, tissue.color,
  outfile =
    "./results/trajactory_combine/Figure5B_tissue_trajectory_promet.pdf",
  center = 1)

# just for plot visualization
grid.arrange(arrangeGrob(grobs = tissue_trj_mrna$plot, ncol = 4, heights = c(4, 4),
  top = textGrob("mRNA aging trajectory in 30 tissues",
    gp = gpar(fontface = "bold",  fontsize = 20)),
  left = textGrob("Z values", gp = gpar(fontface = "bold",  fontsize = 18), rot = 90)))
```

### 9.5.2 defining subtypes

```{r}
# classification

## 1. consensus clustering
combine_trj <- rbind(tissue_trj_mrna$tissue_trajectory_matrix,
  tissue_trj_promet$tissue_trajectory_matrix)
consclass <- runclustering(combine_trj,
  title = "./results/trajactory_combine/consclass_tissue_trajectory_matrix")
tissueType_combine <- data.frame(cons3 = consclass[[3]]$consensusClass,
  cons4 = consclass[[4]]$consensusClass,
  # type_by_cons = rep("Undefined",length(pro.tissues.v)),
  row.names = names(consclass[[3]]$consensusClass))
tissueType_combine$cons3[tissueType_combine$cons3 == 3] <- 2
tissueType_combine$cons4[tissueType_combine$cons4 == 3] <- 1
tissueType_combine$cons4[tissueType_combine$cons4 == 4] <- 2


## 2. hc cluster by trajactory
enbrks <- 1.5 * c(-1, -0.8, -0.6, -0.4, -0.2, -0.1, -0.05, -0.02,
  0.02, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1)
dent <- pheatmap::pheatmap(combine_trj, scale = "none",
  cluster_rows = F, cutree_cols = 2, cutree_rows = 8,
  main = "Heatmap of molecular aging trajectory in 30 tissues",
  height = 10, width = 6, angle_col = 45,
  # clustering_method = "ward.D2",
  fontsize_row = 8, fontsize_col = 8, breaks = enbrks,
  treeheight_row = 20, treeheight_col = 20,
  file =
    "./results/trajactory_combine/Figure4C_heatmap_based_trajectory_combine.pdf",
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(15))
# color=colorRampPalette(c('#3C5488FF','gray95','#E64B35FF'))(15))


dendcol <- as.dendrogram(dent$tree_col)

labelColors <- c("#3C5488FF", "#E64B35FF")

clusMember <- cutree(dent$tree_col, 2)

tissueType_combine$type_by_hc <- clusMember

## 3. MAA
tmpaa <- cbind(tissue_trj_mrna$MAA, tissue_trj_promet$MAA)

colnames(tmpaa) <- c(paste0("mRNA_C", 1:8), paste0("promet_C", 1:8))
consclassmaa <- runclustering(t(tmpaa),
  title = "./results/trajactory_combine/consclass_tissue_trajectory_matrix_MAA")
dent.MAA <- pheatmap::pheatmap(tmpaa, scale = "none", height = 4, width = 3.5,
  main  = "Clustering by MAA",
  fontsize_row = 7, fontsize_col = 7,
  clustering_method = "ward.D2",
  treeheight_row = 20, treeheight_col = 20,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(15),
  # color=colorRampPalette(c('#3B4992','gray95','red'))(30),
  filename =
    "./results/trajactory_combine/FigureX_combine_heatmap_tissue_aging_byMAA.pdf"
)


dendrow <- as.dendrogram(dent.MAA$tree_row)

labelColors <- c("#3C5488FF", "#E64B35FF")

clusMember <- cutree(dent.MAA$tree_row, 2)
tissueType_combine$type_by_MAA <- clusMember
tissueType_combine$type_by_MAA_cons <- consclass[[3]]$consensusClass
tissueType_combine$type_by_MAA_cons[tissueType_combine$type_by_MAA_cons == 3] <- 2


# define by 3 methods
thistype <- rep("Undefined", nrow(tissueType_combine))
thistype[tissueType_combine$cons3 == 1 & tissueType_combine$cons4 == 1 &
  tissueType_combine$type_by_hc == 1 & tissueType_combine$type_by_MAA == 1 &
  tissueType_combine$type_by_MAA_cons == 1] <- "Type II"
thistype[tissueType_combine$cons3 == 2 & tissueType_combine$cons4 == 2 &
  tissueType_combine$type_by_hc == 2 & tissueType_combine$type_by_MAA == 2 &
  tissueType_combine$type_by_MAA_cons == 2] <- "Type I"

tissueType_combine$tissue <- rownames(tissueType_combine)
tissueType_combine$type_combine <- thistype
tissueType_combine$type_promet <- tissueType_promet[rownames(tissueType_combine), ]$type
tissueType_combine$type <- rep("Undefined", nrow(tissueType_combine))
tissueType_combine$type[tissueType_combine$type_combine == "Type I" &
  tissueType_combine$type_promet == "Type I"] <- "Type I"
tissueType_combine$type[tissueType_combine$type_combine == "Type II" &
  tissueType_combine$type_promet == "Type II"] <- "Type II"
tissueType_combine$color <- tissueType$type
tissueType_combine$color[tissueType_combine$type == "Type I"] <- "#FF6347"
tissueType_combine$color[tissueType_combine$type == "Type II"] <- "#1E90FF"
tissueType_combine$color[tissueType_combine$type == "Undefined"] <- "#DAB700"
tissueType_combine$system <- tissue.systems.v[rownames(tissueType_combine)]

tissueType <- tissueType_combine

writetxt_forGPMM(cbind(tissueType_promet, tissueType_combine[rownames(tissueType_promet), ]),
  filename = "./results/trajactory_combine/tissueType_combine_V.txt")

###

summary_vector(tissueType_combine$type_combine)
summary_vector(tissueType_combine$type)
writetxt_forGPMM(tissueType_combine,
  filename = "./results/trajactory_combine/tissueType_combine.txt")
#
```


### 9.5.3 plot Figure 5C and 5D

```{r,fig.width=10, fig.height=9.5, out.width="85%", fig.align="center",dpi = 144}
enbrks <- 1.5 * c(-1, -0.8, -0.6, -0.4, -0.2, -0.1, -0.05, -0.02,
  0.02, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1)

col_fun <- colorRamp2(
  breaks = enbrks,
  colors = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(length(enbrks))
)

col_dend <- hclust(dist(t(combine_trj), method = "euclidean"), method = "complete")

col_clusters <- cutree(col_dend, k = 2)

row_groups <- c(rep(paste0("mC", 1:8), each = 4), rep(paste0("pC", 1:8), each = 4))


column_ha <- HeatmapAnnotation(Type = tissueType$type,
  col = list(Type = c("Type I" = "#FF6347",
    "Type II" = "#1E90FF",
    "Undefined" = "#DAB700")))

row_ha <- rowAnnotation(dataType = c(rep("mRNA", 32), rep("protein & mets", 32)),
  col = list(dataType = c("mRNA" = "#3C5488FF",
    "protein & mets" = "#C71585")))


pdf("./results/trajactory_combine/Figure5C_heatmap_based_trajectory_combine.pdf",
  width = 10, height = 9.5)

pp <- Heatmap(
  combine_trj,
  name = "Z-value",
  col = col_fun,
  cluster_rows = FALSE,

  row_split = row_groups,
  cluster_columns = col_dend,
  column_split = 2,

  column_title = "Trajectories of mRNA and protein & metabolites",
  top_annotation = column_ha,
  left_annotation = row_ha,


  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 12),


  column_names_rot = 45,

  border_gp = gpar(col = "gray50", lwd = 1),


  column_dend_height = unit(10, "mm"),

)
print(pp)
dev.off()


pdf("./results/trajactory_combine/Figure5C_heatmap_based_trajectory_combine_nocolann.pdf",
  width = 10, height = 12)

Heatmap(
  combine_trj,
  name = "Z-value",
  col = col_fun,
  cluster_rows = FALSE,

  row_split = row_groups,

  cluster_columns = col_dend,
  column_split = 2,

  column_title = "Trajectories of mRNA and protein & metabolites",

  left_annotation = row_ha,

  row_names_gp = gpar(fontsize = 7),
  column_names_gp = gpar(fontsize = 12),

  column_names_rot = 45,

  border_gp = gpar(col = "gray50", lwd = 1),

  column_dend_height = unit(10, "mm"),
)

dev.off()

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(pp))
grid.draw(heatmap_grob)
```

```{r,fig.width=7, fig.height=7, out.width="60%", fig.align="center",dpi = 144}
tissue_pca <- prcomp(t(combine_trj), cor = F)
perc_tissue_pca <- 100 * summary(tissue_pca)$importance
tissue_names <- names(promet.tissues)
pdf("./results/trajactory_combine/Figure5D_PCA.pdf", width = 7, height = 7)
Type <- tissueType[tissue_names, ]$type
pp <- ggplot(, aes(tissue_pca$x[, 1], tissue_pca$x[, 2], color = Type)) +
  geom_point(size = 5, alpha = 0.7) +
  theme_classic() + lghplot.addtheme(legend.position = "top") +
  geom_text_repel(aes(label = tissue_names), size = 5, box.padding = 0.5, face = "bold") +
  xlab(paste("PC1(", as.character(round(perc_tissue_pca[2, 1], 1)), "%)", sep = "")) +
  ylab(paste("PC2(", as.character(round(perc_tissue_pca[2, 2], 1)), "%)", sep = "")) +
  scale_color_manual(values = c("#FF6347", "#1E90FF", "#DAB700")) +
  theme(legend.text = element_text(size = 12, face = "bold")) +
  labs(title = "Molecular trajectory PCA")
print(pp)
dev.off()

# just for plot visualization
print(pp)
```

# 10. Figure 6 differ type feature

## 10.1 Figure 6A

```{r,fig.width=6, fig.height=4, out.width="85%", fig.align="center",dpi = 144}
profc.all <- Metapro[, substr(colnames(Metapro), 1, 5) == "beta_"]
colnames(profc.all) <- gsub("beta_", "", colnames(profc.all))
profc.all <- profc.all[, tissueType_combine$tissue]
DEpro_profcall <- DEGenes.simplified(profc.all,
  catagory = tissueType_combine$type == "Type I",
  subset = tissueType_combine$type != "Undefined")
DEpro_profcall$log2FC[is.na(DEpro_profcall$log2FC)] <- 0
DEpro_profcall$Pvalue[is.na(DEpro_profcall$Pvalue)] <- 1
DEpro_profcall$manyNA <- rowSums(!is.na(profc.all[, tissueType$type == "Type I"])) < 2 |
  rowSums(!is.na(profc.all[, tissueType$type == "Type II"])) < 2 # at least 2 samples for each types
writetxt(DEpro_profcall, "./results/differTypes/DEprotein_DEbeta.txt")

## vocano
range <- max(abs(DEpro_profcall[!DEpro_profcall$manyNA, ]$log2FC), na.rm = T) + 0.002
pdf("./results/differTypes/Figure_S15A_vocano_DEpro_probeta_typeI_vs_typeII.pdf",
  height = 5.5)
p <- plot_DEflux(DEpro_profcall[!DEpro_profcall$manyNA, ],
  FCcutoff = 0.008, xlab = "Age effect difference",
  num.showlab = 5, alpha = 0.5, labSize = 6,
  ylab = bquote(~ -Log[10] ~ italic(P))) +
  ggtitle("Age effect difference typeI vs type II")
p <- p + xlim(c(-0.2, 0.2)) +
  theme(plot.title = element_text(size = 20)) + lghplot.addtheme()
print(p)
dev.off()



profc.all.v <- profc.all[DEpro_profcall$Pvalue < 0.05 &
  !DEpro_profcall$manyNA &
  abs(DEpro_profcall$log2FC) > 0.008, ]
sum(DEpro_profcall$Pvalue < 0.05 & !DEpro_profcall$manyNA &
  DEpro_profcall$log2FC > 0.008)
sum(DEpro_profcall$Pvalue < 0.05 & !DEpro_profcall$manyNA &
  DEpro_profcall$log2FC < -0.008)
tmptype <- data.frame(System = tissueType_combine$system,
  TissueType = tissueType_combine$type,
  row.names = tissueType_combine$tissue)

profc.all.v[is.na(profc.all.v)] <- 0
dim(profc.all.v)

annotation_colors <- list(System = tissue.color,
  TissueType = c("Type I" = "#FF6347",
    "Type II" = "#1E90FF",
    "Undefined" = "#DAB700"))
enbrks <- 0.15 * c(-1, -0.8, -0.4, -0.2, -0.1, -0.05, 0.05, 0.1, 0.2, 0.4, 0.8, 1)

dent.pro <- pheatmap::pheatmap(profc.all.v, scale = "none",
  cluster_rows = T, show_rownames = F,
  cluster_cols = T, show_colnames = T,
  main = "Heatmap of aging_beta in 30 tissues",
  cutree_rows = 2, cutree_cols = 2,
  height = 4, width = 6, angle_col = 45,
  fontsize_row = 8, fontsize_col = 8,
  clustering_method = "ward.D2",
  breaks = enbrks,
  annotation_col = tmptype,
  annotation_colors = annotation_colors,
  treeheight_row = 10, treeheight_col = 10,
  # file ="./results/differTypes/Figure 6A_DEbeta_proteinv1.pdf",
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11))

pdf("./results/differTypes/Figure 6A_DEbeta_proteinv1.pdf", height = 4, width = 6)
grid::grid.draw(dent.pro$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(dent.pro$gtable)
```

```{r,fig.width=6, fig.height=4, out.width="85%", fig.align="center",dpi = 144}
mrnafc.all <- Metamrna[, substr(colnames(Metamrna), 1, 5) == "beta_"]
colnames(mrnafc.all) <- gsub("beta_", "", colnames(mrnafc.all))
mrnafc.all <- mrnafc.all[, tissueType_combine$tissue]
DEmrna_mrnafcall <- DEGenes.simplified(mrnafc.all,
  catagory = tissueType_combine$type == "Type I",
  subset = tissueType_combine$type != "Undefined")
DEmrna_mrnafcall$log2FC[is.na(DEmrna_mrnafcall$log2FC)] <- 0
DEmrna_mrnafcall$Pvalue[is.na(DEmrna_mrnafcall$Pvalue)] <- 1
DEmrna_mrnafcall$manyNA <-
  rowSums(!is.na(mrnafc.all[, tissueType$type == "Type I"])) < 2 |
    rowSums(!is.na(mrnafc.all[, tissueType$type == "Type II"])) < 2


writetxt(DEmrna_mrnafcall, "./results/differTypes/DEmrna_DEbeta.txt")

# vocano
range <- max(abs(DEmrna_mrnafcall[!DEmrna_mrnafcall$manyNA, ]$log2FC), na.rm = T) + 0.002
pdf("./results/differTypes/Figure_S15A_vocano_DEmrna_probeta_typeI_vs_typeII.pdf",
  height = 5.5)
p <- plot_DEflux(DEmrna_mrnafcall[!DEmrna_mrnafcall$manyNA, ],
  FCcutoff = 0.008, xlab = "Age effect difference",
  num.showlab = 5, alpha = 0.5, labSize = 6, # fixpointsize = 3,
  ylab = bquote(~ -Log[10] ~ italic(P))) +
  ggtitle("Age effect difference typeI vs type II")
p <- p + xlim(c(-range, range)) +
  theme(plot.title = element_text(size = 20)) + lghplot.addtheme()
print(p)
dev.off()


mrnafc.all.v <- mrnafc.all[DEmrna_mrnafcall$Pvalue < 0.05 &
  !DEmrna_mrnafcall$manyNA &
  abs(DEmrna_mrnafcall$log2FC) > 0.008, ]
sum(DEmrna_mrnafcall$Pvalue < 0.05 & !DEmrna_mrnafcall$manyNA &
  DEmrna_mrnafcall$log2FC > 0.008)
sum(DEmrna_mrnafcall$Pvalue < 0.05 & !DEmrna_mrnafcall$manyNA &
  DEmrna_mrnafcall$log2FC < -0.008)
tmptype <- data.frame(System = tissueType_combine$system,
  TissueType = tissueType_combine$type,
  row.names = tissueType_combine$tissue)

mrnafc.all.v[is.na(mrnafc.all.v)] <- 0
dim(mrnafc.all.v)

annotation_colors <- list(System = tissue.color,
  TissueType = c("Type I" = "#FF6347",
    "Type II" = "#1E90FF",
    "Undefined" = "#DAB700"))
enbrks <- 0.15 * c(-1, -0.8, -0.4, -0.2, -0.1, -0.05, 0.05, 0.1, 0.2, 0.4, 0.8, 1)

dent.mrna <- pheatmap::pheatmap(mrnafc.all.v, scale = "none",
  cluster_rows = T, show_rownames = F,
  cluster_cols = T, show_colnames = T,
  main = "Heatmap of aging_beta in 30 tissues",
  height = 4, width = 6, angle_col = 45,
  cutree_rows = 2, cutree_cols = 2,
  fontsize_row = 8, fontsize_col = 8,
  breaks = enbrks,
  annotation_col = tmptype,
  annotation_colors = annotation_colors,
  treeheight_row = 10, treeheight_col = 10,
  # file ="./results/differTypes/Figure XS_DEbeta_mrna.pdf",
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11))

pdf("./results/differTypes/Figure XS_DEbeta_mrna.pdf", height = 4, width = 6)
grid::grid.draw(dent.mrna$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(dent.mrna$gtable)
```

```{r,fig.width=6, fig.height=4, out.width="85%", fig.align="center",dpi = 144}
metfc.all <- Metamet[, substr(colnames(Metamet), 1, 5) == "beta_"]
colnames(metfc.all) <- gsub("beta_", "", colnames(metfc.all))
metfc.all <- metfc.all[, tissueType_combine$tissue]
DEmet_metfcall <- DEGenes.simplified(metfc.all,
  catagory = tissueType_combine$type == "Type I",
  subset = tissueType_combine$type != "Undefined")
DEmet_metfcall$log2FC[is.na(DEmet_metfcall$log2FC)] <- 0
DEmet_metfcall$Pvalue[is.na(DEmet_metfcall$Pvalue)] <- 1
DEmet_metfcall$manyNA <-
  rowSums(!is.na(metfc.all[, tissueType$type == "Type I"])) < 2 |
    rowSums(!is.na(metfc.all[, tissueType$type == "Type II"])) < 2


writetxt(DEmet_metfcall, "./results/differTypes/DEmet_DEbeta.txt")

# vocano
range <- max(abs(DEmet_metfcall[!DEmet_metfcall$manyNA, ]$log2FC), na.rm = T) + 0.002
pdf("./results/differTypes/Figure_S15A_vocano_DEmet_probeta_typeI_vs_typeII.pdf",
  height = 5.5)
p <- plot_DEflux(DEmet_metfcall[!DEmet_metfcall$manyNA, ],
  FCcutoff = 0.008, xlab = "Age effect difference",
  num.showlab = 5, alpha = 0.5, labSize = 6, # fixpointsize = 3,
  ylab = bquote(~ -Log[10] ~ italic(P))) +
  ggtitle("Age effect difference typeI vs type II")
p <- p + xlim(c(-range, range)) +
  theme(plot.title = element_text(size = 20)) + lghplot.addtheme()
print(p)
dev.off()


metfc.all.v <- metfc.all[DEmet_metfcall$Pvalue < 0.05 &
  !DEmet_metfcall$manyNA &
  abs(DEmet_metfcall$log2FC) > 0.008, ]
sum(DEmet_metfcall$Pvalue < 0.05 & !DEmet_metfcall$manyNA &
  DEmet_metfcall$log2FC > 0.008)
sum(DEmet_metfcall$Pvalue < 0.05 & !DEmet_metfcall$manyNA &
  DEmet_metfcall$log2FC < -0.008)
tmptype <- data.frame(System = tissueType_combine$system,
  TissueType = tissueType_combine$type,
  row.names = tissueType_combine$tissue)

metfc.all.v[is.na(metfc.all.v)] <- 0
dim(metfc.all.v)

annotation_colors <- list(System = tissue.color,
  TissueType = c("Type I" = "#FF6347",
    "Type II" = "#1E90FF",
    "Undefined" = "#DAB700"))
enbrks <- 0.15 * c(-1, -0.8, -0.4, -0.2, -0.1, -0.05, 0.05, 0.1, 0.2, 0.4, 0.8, 1)

dent.met <- pheatmap::pheatmap(metfc.all.v, scale = "none",
  cluster_rows = T, show_rownames = F,
  cluster_cols = T, show_colnames = T,
  main = "Heatmap of aging_beta in 30 tissues",
  height = 4, width = 6, angle_col = 45,
  # cutree_rows = 2, cutree_cols = 2,
  fontsize_row = 8, fontsize_col = 8,
  breaks = enbrks,
  annotation_col = tmptype,
  annotation_colors = annotation_colors,
  treeheight_row = 10, treeheight_col = 10,
  # file ="./results/differTypes/Figure XS_DEbeta_met.pdf",
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(11))

pdf("./results/differTypes/Figure XS_DEbeta_met.pdf", height = 4, width = 6)
grid::grid.draw(dent.met$gtable)
dev.off()

# just for plot visualization
grid::grid.draw(dent.met$gtable)
```


## 10.2 Figure 6B metascape

```{r,fig.width=11, fig.height=8, out.width="75%", fig.align="center",dpi = 144}
### please use the pro_up  and pro down to perform GO enrichment, and then save the files in
# './results/differTypes/metascape/metascape_result_pro_up.xlsx', and
# in './results/differTypes/metascape/metascape_result_pro_down.xlsx'
# then run the follows:

# goup.pro = parse_metascape('./results/differTypes/metascape/metascape_result_pro_up.xlsx',
#                          gokegg = FALSE)
# godown.pro = parse_metascape('./results/differTypes/metascape/metascape_result_pro_down.xlsx',
#                          gokegg = FALSE)
# pdf('./results/differTypes/Figure_6B_beta_differ_protein_Enrichment.pdf',
#               width = 11,height = 8)
# p = plot_enrich(goup.pro, godown.pro,n_char = 40)
# print(p)
# dev.off()

# here is the results
goup.pro <- parse_metascape("./data/metascape_result_pro_up_typeI_vs_typeII.xlsx",
  gokegg = FALSE)
godown.pro <- parse_metascape("./data/metascape_result_pro_down_typeI_vs_typeII.xlsx",
  gokegg = FALSE)
pdf("./results/differTypes/Figure_6B_beta_differ_protein_Enrichment.pdf",
  width = 11, height = 8)
p <- plot_enrich(goup.pro, godown.pro, n_char = 40)
print(p)
dev.off()

# just for plot visualization
print(p)
```

## 10.3 meta differ tisses

```{r}
tissueClass <- tissueType
tissueClass$system <- tissue.systems.v[rownames(tissueClass)]
```

```{r}
typeI_tisses <- tissueClass$tissue[tissueClass$type == "Type I"]
# mrna
Metapro.typeI <- get_tissue_DEgenes_lm(pro.tissues.v[typeI_tisses],
  pro.tissues.info[typeI_tisses],
  tissue.systems = NULL)$MetaLimma
# Metapro$gene = pro.whole.nofilter.header[rownames(Metapro),]$Gene
writetxt(Metapro.typeI, "./results/differTypes/TypeI_Common_DEpro_from_metapro.txt")

Metamrna.typeI <- get_tissue_DEgenes_lm(mrna.tissues[typeI_tisses],
  mrna.tissues.info[typeI_tisses],
  tissue.systems = NULL)$MetaLimma
writetxt(Metamrna.typeI, "./results/differTypes/TypeI_Common_DEmrna_from_metamrna.txt")

Metamet.typeI <- get_tissue_DEgenes_lm(met.tissues[typeI_tisses],
  met.tissues.info[typeI_tisses],
  tissue.systems = NULL)$MetaLimma
writetxt(Metamet.typeI, "./results/differTypes/TypeI_Common_DEmet_from_metamet.txt")
```

```{r}
typeII_tisses <- tissueClass$tissue[tissueClass$type == "Type II"]
# mrna
Metapro.typeII <- get_tissue_DEgenes_lm(pro.tissues.v[typeII_tisses],
  pro.tissues.info[typeII_tisses],
  tissue.systems = NULL)$MetaLimma
writetxt(Metapro.typeII,
  "./results/differTypes/TypeII_Common_DEpro_from_metapro.txt")

Metamrna.typeII <- get_tissue_DEgenes_lm(mrna.tissues[typeII_tisses],
  mrna.tissues.info[typeII_tisses],
  tissue.systems = NULL)$MetaLimma
writetxt(Metamrna.typeII,
  "./results/differTypes/TypeII_Common_DEmrna_from_metamrna.txt")

Metamet.typeII <- get_tissue_DEgenes_lm(met.tissues[typeII_tisses],
  met.tissues.info[typeII_tisses],
  tissue.systems = NULL)$MetaLimma
writetxt(Metamet.typeII,
  "./results/differTypes/TypeII_Common_DEmet_from_metamet.txt")
```

```{r}
commonMols.filter$FDR005$macaca_upmrna_meta_typeI <-
  Metamrna.typeI$ID[Metamrna.typeI$MetaFDR < 0.05 &
    Metamrna.typeI$MetaBeta > beta_cutoff &
    !Metamrna.typeI$manyNA]

commonMols.filter$FDR005$macaca_upmrna_meta_typeII <-
  Metamrna.typeII$ID[Metamrna.typeII$MetaFDR < 0.05 &
    Metamrna.typeII$MetaBeta > beta_cutoff &
    !Metamrna.typeII$manyNA]

commonMols.filter$FDR005$macaca_downmrna_meta_typeI <-
  Metamrna.typeI$ID[Metamrna.typeI$MetaFDR < 0.05 &
    Metamrna.typeI$MetaBeta < -beta_cutoff &
    !Metamrna.typeI$manyNA]

commonMols.filter$FDR005$macaca_downmrna_meta_typeII <-
  Metamrna.typeII$ID[Metamrna.typeII$MetaFDR < 0.05 &
    Metamrna.typeII$MetaBeta < -beta_cutoff &
    !Metamrna.typeII$manyNA]


commonMols.filter$FDR005$macaca_uppro_meta_typeI <-
  Metapro.typeI$ID[Metapro.typeI$MetaFDR < 0.05 &
    Metapro.typeI$MetaBeta > beta_cutoff &
    !Metapro.typeI$manyNA]

commonMols.filter$FDR005$macaca_uppro_meta_typeII <-
  Metapro.typeII$ID[Metapro.typeII$MetaFDR < 0.05 &
    Metapro.typeII$MetaBeta > beta_cutoff &
    !Metapro.typeII$manyNA]

commonMols.filter$FDR005$macaca_downpro_meta_typeI <-
  Metapro.typeI$ID[Metapro.typeI$MetaFDR < 0.05 &
    Metapro.typeI$MetaBeta < -beta_cutoff &
    !Metapro.typeI$manyNA]

commonMols.filter$FDR005$macaca_downpro_meta_typeII <-
  Metapro.typeII$ID[Metapro.typeII$MetaFDR < 0.05 &
    Metapro.typeII$MetaBeta < -beta_cutoff &
    !Metapro.typeII$manyNA]

commonMols.filter$FDR005$macaca_upmet_meta_typeI <-
  Metamet.typeI$ID[Metamet.typeI$MetaFDR < 0.05 &
    Metamet.typeI$MetaBeta > beta_cutoff &
    !Metamet.typeI$manyNA]

commonMols.filter$FDR005$macaca_upmet_meta_typeII <-
  Metamet.typeII$ID[Metamet.typeII$MetaFDR < 0.05 &
    Metamet.typeII$MetaBeta > beta_cutoff &
    !Metamet.typeII$manyNA]

commonMols.filter$FDR005$macaca_downmet_meta_typeI <-
  Metamet.typeI$ID[Metamet.typeI$MetaFDR < 0.05 &
    Metamet.typeI$MetaBeta < -beta_cutoff &
    !Metamet.typeI$manyNA]

commonMols.filter$FDR005$macaca_downmet_meta_typeII <-
  Metamet.typeII$ID[Metamet.typeII$MetaFDR < 0.05 &
    Metamet.typeII$MetaBeta < -beta_cutoff &
    !Metamet.typeII$manyNA]

# write mrna
tmp <- data.frame(names = c("TypeI_up", "TypeII_up", "TypeI_down", "TypeII_down"),
  stringsAsFactors = F,
  genes = c(paste0(commonMols.filter$FDR005$macaca_upmrna_meta_typeI,
    collapse = ","),
  paste0(commonMols.filter$FDR005$macaca_upmrna_meta_typeII,
    collapse = ","),
  paste0(commonMols.filter$FDR005$macaca_downmrna_meta_typeI,
    collapse = ","),
  paste0(commonMols.filter$FDR005$macaca_downmrna_meta_typeII,
    collapse = ","))
)
writetxt(tmp, "./results/differTypes/input_mrna_differtypes_metascape.txt",
  col.names = F)


# write protein
tmp <- data.frame(names = c("TypeI_up", "TypeII_up", "TypeI_down", "TypeII_down"),
  stringsAsFactors = F,
  genes = c(paste0(commonMols.filter$FDR005$macaca_uppro_meta_typeI,
    collapse = ","),
  paste0(commonMols.filter$FDR005$macaca_uppro_meta_typeII,
    collapse = ","),
  paste0(commonMols.filter$FDR005$macaca_downpro_meta_typeI,
    collapse = ","),
  paste0(commonMols.filter$FDR005$macaca_downpro_meta_typeII,
    collapse = ","))
)
writetxt(tmp, "./results/differTypes/input_pro_differtypes_metascape.txt", col.names = F)

# write protein loose
up1 <-
  Metapro.typeI$ID[Metapro.typeI$MetaFDR < 0.1 &
    Metapro.typeI$MetaBeta > beta_cutoff &
    !Metapro.typeI$manyNA]
up2 <-
  Metapro.typeII$ID[Metapro.typeII$MetaFDR < 0.1 &
    Metapro.typeII$MetaBeta > beta_cutoff &
    !Metapro.typeII$manyNA]

down1 <-
  Metapro.typeI$ID[Metapro.typeI$MetaFDR < 0.1 &
    Metapro.typeI$MetaBeta < -beta_cutoff &
    !Metapro.typeI$manyNA]
down2 <-
  Metapro.typeII$ID[Metapro.typeII$MetaFDR < 0.1 &
    Metapro.typeII$MetaBeta < -beta_cutoff &
    !Metapro.typeII$manyNA]

tmp <- data.frame(names = c("TypeI_up", "TypeII_up", "TypeI_down", "TypeII_down"),
  stringsAsFactors = F,
  genes = c(paste0(up1, collapse = ","),
    paste0(up2, collapse = ","),
    paste0(down1, collapse = ","),
    paste0(down2, collapse = ","))
)

writetxt(tmp, "./results/differTypes/input_pro_differtypes_metascape_loose.txt",
  col.names = F)
c(length(up1), length(up2), length(down1), length(down2))



length(commonMols.filter$FDR005$macaca_uppro_meta_typeII)
length(commonMols.filter$FDR005$macaca_uppro_meta_typeI)

length(commonMols.filter$FDR005$macaca_downpro_meta_typeII)
length(commonMols.filter$FDR005$macaca_downpro_meta_typeI)


length(commonMols.filter$FDR005$macaca_upmrna_meta_typeII)
length(commonMols.filter$FDR005$macaca_upmrna_meta_typeI)

length(commonMols.filter$FDR005$macaca_downmrna_meta_typeII)
length(commonMols.filter$FDR005$macaca_downmrna_meta_typeI)


length(commonMols.filter$FDR005$macaca_upmet_meta_typeII)
length(commonMols.filter$FDR005$macaca_upmet_meta_typeI)

length(commonMols.filter$FDR005$macaca_downmet_meta_typeII)
length(commonMols.filter$FDR005$macaca_downmet_meta_typeI)
```

## 10.4 Deal with developmenat signal

```{r}
# gsea
bp_gene_sets <- msigdbr(species = "Homo sapiens",
  category = "C5", subcategory = "GO:BP")
gene_sets_bp <- split(bp_gene_sets$gene_symbol, bp_gene_sets$gs_name)

tmpfun_gsea <- function(Metapro.typeI, gene_sets_bp) {
  idx <- !Metapro.typeI$manyNA
  thisbeta <- Metapro.typeI$MetaBeta[idx]
  names(thisbeta) <- Metapro.typeI$ID[idx]
  thisbeta <- sort(thisbeta, decreasing = TRUE)
  tmpout <- fgsea(pathways = gene_sets_bp, stats = thisbeta)
  return(tmpout)
}

gsea.Difftype <- list()
# type 1 vs type II
# pro
vids <- rowSums(is.na(profc.all)) < 15
thisbeta <- DEpro_profcall$log2FC[vids & !is.na(DEpro_profcall$log2FC)]
names(thisbeta) <- DEpro_profcall$ID[vids & !is.na(DEpro_profcall$log2FC)]
thisbeta <- sort(thisbeta, decreasing = TRUE)
thisbetapro <- thisbeta
gsea.Difftype$pro.typeI_vs_typeII <- fgsea(pathways = gene_sets_bp, stats = thisbeta)

# mRNA
thisbeta <- DEmrna_mrnafcall$log2FC[!DEmrna_mrnafcall$manyNA &
  !is.na(DEmrna_mrnafcall$log2FC)]
names(thisbeta) <- DEmrna_mrnafcall$ID[!DEmrna_mrnafcall$manyNA &
  !is.na(DEmrna_mrnafcall$log2FC)]
thisbeta <- sort(thisbeta, decreasing = TRUE)
gsea.Difftype$mrna.typeI_vs_typeII <- fgsea(pathways = gene_sets_bp, stats = thisbeta)
```

```{r}
idx <- regexpr("development", gsea.Difftype$pro.typeI_vs_typeII$pathway, ignore.case = T) > 0 &
  gsea.Difftype$pro.typeI_vs_typeII$padj < 0.01
idx[is.na(idx)] <- FALSE
sum(idx)
gsea.Difftype$pro.typeI_vs_typeII[idx, ]
# plot
fgseaRes <- gsea.Difftype$pro.typeI_vs_typeII
topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n = 20), pathway]
topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n = 20), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
pathways1 <- gene_sets_bp
fgseaRes$pathway <- gsub("GOBP_", "", fgseaRes$pathway)
names(pathways1) <- gsub("GOBP_", "", names(pathways1))
topPathways <- gsub("GOBP_", "", topPathways)
p <- plotGseaTable(pathways1[topPathways], thisbetapro, fgseaRes, colwidths = c(7, 1, 1, 1, 1),
  gseaParam = 0.5, )
pdf(file = "./results/differTypes//GSEA_pro_typeI_vs_typeII.pdf", width = 9, height = 12)
print(p)
dev.off()


idx <- regexpr("development", fgseaRes$pathway, ignore.case = T) > 0
tmpaa <- as.data.frame(fgseaRes[idx, ])
tmpaa$log2FC <- tmpaa$NES
tmpaa$Pvalue <- tmpaa$padj
tmpaa$ID <- gsub("GOBP_", "", tmpaa$pathway)
rownames(tmpaa) <- tmpaa$ID
tmpaa <- tmpaa[!is.na(tmpaa$log2FC), ]
pdf("./results/differTypes/gsea_developement_typeI_vs_typeII.pdf",
  width = 12, height = 12)
plot_DEflux(tmpaa, num.showlab = 5, title = "GSEA developement analysis",
  pcutoff = 0.01, alpha = 2,
  ylab = bquote(italic(padj)),
  xlab = "NES") + lghplot.addtheme(size = 36, sizex = 32, sizey = 32)
dev.off()


idx <- regexpr("development", fgseaRes$pathway, ignore.case = T) > 0
tmpaa <- as.data.frame(fgseaRes)
tmpaa$log2FC <- tmpaa$NES
tmpaa$Pvalue <- tmpaa$padj
tmpaa$ID <- as.vector(gsub("GOBP_", "", tmpaa$pathway))
tmpaa <- tmpaa[!is.na(tmpaa$log2FC), ]

pdf("./results/differTypes/gsea_all_typeI_vs_typeII.pdf", width = 12, height = 12)
plot_DEflux(tmpaa, num.showlab = 5, title = "GSEA all analysis",
  pcutoff = 0.01, labSize = 6, alpha = 2,
  ylab = bquote(italic(padj)),
  xlab = "NES") + lghplot.addtheme(size = 36, sizex = 32, sizey = 32)
dev.off()
```

## 10.5 Figure S14 vocano diff pro

```{r}
volcono_plot_subtype <- function(Metapro.typeI, title, FCcutoff = 0.0080) {
  tmpxx <- Metapro.typeI[, substr(colnames(Metapro.typeI), 1, 5) == "beta_"]
  colnames(tmpxx) <- gsub("beta_", "", colnames(tmpxx))
  tmpmanyNA <- rowSums(is.na(tmpxx)) >= ncol(tmpxx) / 3
  tmpdata <- data.frame(ID = rownames(Metapro.typeI),
    log2FC = Metapro.typeI$MetaBeta,
    Pvalue = Metapro.typeI$MetaFDR,
    manyNA = tmpmanyNA)
  tmpdata <- tmpdata[!tmpdata$manyNA, ]
  range <- max(abs(tmpdata$log2FC), na.rm = T) + 0.002
  ntissue <- sum(substr(colnames(Metapro.typeI), 1, 4) == "beta")
  xlab <- paste0("Age effect (beta lm ", ntissue, " tissues)")
  p <- plot_DEflux(tmpdata, alpha = 0.5, num.showlab = 5, FCcutoff = FCcutoff,
    labSize = 6, xlab = xlab,
    ylab = bquote(~ -Log[10] ~ italic("FDR")),
    title = title)
  p <- p + xlim(c(-range, range)) +
    theme(plot.title = element_text(size = 20)) + lghplot.addtheme()
  return(p)
}
pdf("./results/differTypes//Figure_S14A_mRNA_typeI.pdf", height = 5.5)
p <- volcono_plot_subtype(Metamrna.typeI, "Aging related mRNA in Type I tissues")
print(p)
dev.off()
pdf("./results/differTypes/Figure_S14B_mRNA_typeII.pdf", height = 5.5)
p <- volcono_plot_subtype(Metamrna.typeII, "Aging related mRNA in Type II tissues")
print(p)
dev.off()

pdf("./results/differTypes/Figure_S14C_pro_typeI.pdf", height = 5.5)
p <- volcono_plot_subtype(Metapro.typeI, "Aging related protein in Type I tissues")
print(p)
dev.off()
pdf("./results/differTypes/Figure_S14D_pro_typeII.pdf", height = 5.5)
p <- volcono_plot_subtype(Metapro.typeII, "Aging related protein in Type II tissues")
print(p)
dev.off()


pdf("./results/differTypes/Figure_S14E_met_typeI.pdf", height = 5.5)
p <- volcono_plot_subtype(short_met_names(Metamet.typeI),
  "Aging related metabolites in Type I tissues")
print(p)
dev.off()
pdf("./results/differTypes/Figure_S14F_met_typeII.pdf", height = 5.5)
p <- volcono_plot_subtype(short_met_names(Metamet.typeII),
  "Aging related metabolites in Type II tissues")
print(p)
dev.off()
```

# 11. Differ ratio

## 11.1 construct data

```{r}
# construct data
tissues <- names(promet.tissues.Z)

ratio.tissues <- list()
ratio.tissues.info <- list()
pro.tissues.forRatio <- list()
mrna.tissues.forRatio <- list()
mrna.tissues.forRatio_noadj <- list()
tissues <- names(pro.tissues)
overlaptissues <- intersect(names(pro.tissues.v), names(mrna.tissues))
for (i in 1:length(overlaptissues)) {
  this_tissue <- overlaptissues[i]
  thispro <- pro.tissues.v[[this_tissue]]
  thismrna <- mrna.tissues[[this_tissue]]
  thismrna.noadj <- mrna.tissues.noadj[[this_tissue]]
  vcol <- intersect(colnames(thispro), colnames(thismrna))
  vrow <- intersect(rownames(thispro), rownames(thismrna))
  thispro <- thispro[vrow, vcol]
  thismrna  <- thismrna[vrow, vcol]
  thismrna.noadj  <- thismrna.noadj[vrow, vcol]
  pro.tissues.forRatio[[i]] <- thispro
  mrna.tissues.forRatio[[i]] <- thismrna
  mrna.tissues.forRatio_noadj[[i]] <- thismrna.noadj
  ratio.tissues[[i]] <- thispro - thismrna
  ratio.tissues.info[[i]] <- pro.tissues.info[[i]][vcol, ]
}
names(ratio.tissues) <- overlaptissues
names(pro.tissues.forRatio) <- overlaptissues
names(mrna.tissues.forRatio) <- overlaptissues
names(mrna.tissues.forRatio_noadj) <- overlaptissues
names(ratio.tissues.info) <- overlaptissues
```

## 11.2 mRNA_pro relation

### 11.2.1 mrna vs pro

```{r,fig.width=12.5, fig.height=15, out.width="100%", fig.align="center",dpi = 144}
mRNA.mean <- list()
mRNA.mean.noadj <- list()
pro.mean <- list()
for(i in 1:length(pro.tissues.forRatio)){
    mRNA.mean[[i]] <- rowMeans(mrna.tissues.forRatio[[i]],na.rm = T)
    mRNA.mean.noadj[[i]] <- rowMeans(mrna.tissues.forRatio_noadj[[i]],na.rm = T)
    pro.mean[[i]] <- rowMeans(pro.tissues.forRatio[[i]],na.rm = T)
}
names(mRNA.mean) <- names(mrna.tissues.forRatio)
names(mRNA.mean.noadj) <- names(mrna.tissues.forRatio_noadj)
names(pro.mean) <- names(pro.tissues.forRatio)
RNA.v <- list_to_matrix(mRNA.mean,names(mRNA.mean))
RNA.v.noadj <- list_to_matrix(mRNA.mean.noadj,names(mRNA.mean.noadj))
pro.v <- list_to_matrix(pro.mean,names(pro.mean))

pp <- list()
pp1 <- list()
for(i in 1:ncol(pro.v)){
    idaa <- !is.na(RNA.v[,i]) & !is.na(pro.v[,i])
    tcor <- cor.test(RNA.v[idaa,i],pro.v[idaa,i])$estimate
    tcor1 <- cor.test(RNA.v.noadj[idaa,i],pro.v[idaa,i])$estimate
    tmpdata <- data.frame(xx = RNA.v[idaa,i],
                        yy = pro.v[idaa,i],
                        xx.noajd = RNA.v.noadj[idaa,i])
    pp[[i]] <- ggplot(tmpdata,aes(x= xx,y = yy))+
        ggrastr::geom_point_rast(size =1, dpi = 144) +  theme_bw()+
        theme(plot.margin = margin(0.1,0.1,0.1,0.1,"cm"))+
        lghplot.addtheme(size = 16,sizex = 16,sizey = 16)+        
        annotate(geom="text", x=4, y=28, 
                label = paste('R=',signif(tcor,3)),
               color="darkblue",size = 8,face = "italic")+
        xlab('')+ylab('')+ggtitle(colnames(pro.v)[i])
    
    pp1[[i]] <- ggplot(tmpdata,aes(x= xx.noajd,y = yy))+
        ggrastr::geom_point_rast(size =1, dpi = 144) +  theme_bw()+
        theme(plot.margin = margin(0.1,0.1,0.1,0.1,"cm"))+
        lghplot.addtheme(size = 16,sizex = 16,sizey = 16)+        
        annotate(geom="text", x=6, y=25, 
                label = paste('R=',signif(tcor1,3)),
               color="darkblue",size = 8,face = "italic")+
        xlab('')+ylab('')+ggtitle(colnames(pro.v)[i])
}

# compare batch and RIN adjusted log2 (TMM normalized CPM) with protein
png(file = "./results/ratio/Figure_SX_mrna_vs_pro.png",
    width = 1250,height = 1500)
grid.arrange(arrangeGrob(grobs = pp, ncol = 5,
   bottom=textGrob('mRNA expression(log2 CPM)', 
     gp=gpar(fontface="bold",  fontsize=22)),
   left = textGrob('Protein abundance(log2 Peak Area)', 
     gp=gpar(fontface="bold",  fontsize=22),rot=90)))
dev.off()

# compare log2 (TMM normalized CPM) with protein
png(file = "./results/ratio/Figure_SX_mrna_vs_pro_using_noadjcpm.png",
   width = 1250,height = 1500)
grid.arrange(arrangeGrob(grobs = pp1, ncol = 5,
   bottom=textGrob('mRNA expression(log2 CPM)', 
     gp=gpar(fontface="bold",  fontsize=22)),
   left = textGrob('Protein abundance(log2 Peak Area)', 
     gp=gpar(fontface="bold",  fontsize=22),rot=90)))
dev.off()

# just for plot visualization
grid.arrange(arrangeGrob(grobs = pp1, ncol = 5,
   bottom=textGrob('mRNA expression(log2 CPM)', 
     gp=gpar(fontface="bold",  fontsize=22)),
   left = textGrob('Protein abundance(log2 Peak Area)', 
     gp=gpar(fontface="bold",  fontsize=22),rot=90)))

```

### 11.2.2 predict protein vs measured protein

```{r,fig.width=12.5, fig.height=15, out.width="100%", fig.align="center",dpi = 144}
ratio <- pro.v - RNA.v
ratio.noadj <- pro.v - RNA.v.noadj
ratio <- apply(ratio,1,median,na.rm =T)
ratio.noadj <- apply(ratio.noadj,1,median,na.rm =T)

pro.prediction <- RNA.v + ratio
pro.prediction.noadj <- RNA.v.noadj + ratio.noadj

pp <- list()
pp1 <- list()
for(i in 1:ncol(pro.v)){
    idaa <- !is.na(pro.prediction[,i]) & !is.na(pro.v[,i])
    tcor <- cor.test(pro.prediction[idaa,i],pro.v[idaa,i])$estimate
    tcor1 <- cor.test(pro.prediction.noadj[idaa,i],pro.v[idaa,i])$estimate
    tmpdata <- data.frame(xx = pro.prediction[idaa,i],
                        yy = pro.v[idaa,i],
                        xx.noadj = pro.prediction.noadj[idaa,i])
    pp[[i]] <- ggplot(tmpdata,aes(x= xx,y = yy))+
        ggrastr::geom_point_rast(size =1, dpi = 144) + theme_bw()+
        theme(plot.margin = margin(0.1,0.1,0.1,0.1,"cm"))+
        lghplot.addtheme(size = 16,sizex = 16,sizey = 16)+        
        annotate(geom="text", x=15, y=25, 
                label = paste('R=',signif(tcor,3)),
               color="darkblue",size = 8,face = "italic")+
        xlab('')+ylab('')+ggtitle(colnames(pro.v)[i])
    pp1[[i]] <- ggplot(tmpdata,aes(x= xx.noadj,y = yy))+
        ggrastr::geom_point_rast(size =1, dpi = 144) + theme_bw()+
        theme(plot.margin = margin(0.1,0.1,0.1,0.1,"cm"))+
        lghplot.addtheme(size = 16,sizex = 16,sizey = 16)+        
        annotate(geom="text", x=15, y=25, 
                label = paste('R=',signif(tcor1,3)),
               color="darkblue",size = 8,face = "italic")+
        xlab('')+ylab('')+ggtitle(colnames(pro.v)[i])
}

# predict proteins using batch and RIN adjusted log2 (TMM normalized CPM)
png(file = "./results/ratio/Figure SX_predPro_vs_pro.png",
  width = 1250,height = 1500)
grid.arrange(arrangeGrob(grobs = pp, ncol = 5,
  bottom=textGrob('Predicted protein abundance(log2 Peak Area)', 
    gp=gpar(fontface="bold",  fontsize=22)),
  left = textGrob('Protein abundance(log2 Peak Area)', 
                  gp=gpar(fontface="bold",  fontsize=22),rot=90)))
dev.off()

# predict proteins using log2 (TMM normalized CPM)
png(file = "./results/ratio/Figure SX_predPro_vs_pro_noadjcpm.png",
  width = 1250,height = 1500)
grid.arrange(arrangeGrob(grobs = pp1, ncol = 5,
  bottom=textGrob('Predicted protein abundance(log2 Peak Area)', 
    gp=gpar(fontface="bold",  fontsize=22)),
  left = textGrob('Protein abundance(log2 Peak Area)', 
                  gp=gpar(fontface="bold",  fontsize=22),rot=90)))
dev.off()

# just for plot visualization
grid.arrange(arrangeGrob(grobs = pp1, ncol = 5,
  bottom=textGrob('Predicted protein abundance(log2 Peak Area)', 
    gp=gpar(fontface="bold",  fontsize=22)),
  left = textGrob('Protein abundance(log2 Peak Area)', 
                  gp=gpar(fontface="bold",  fontsize=22),rot=90)))

```


## 11.3 DEratio use limma

```{r}
DEratio.limma <- get_tissue_DEgenes_lm(ratio.tissues,
  ratio.tissues.info, tissue.systems = NULL)
```

```{r,fig.width=25, fig.height=12.5, out.width="100%", fig.align="center",dpi = 144}
ratio_beta <- DEratio.limma$Aging
tissueClass <- tissueType[names(ratio_beta), ]
ratio_beta.mean <- rep(0, length(ratio_beta))
ratio_beta.median <- rep(0, length(ratio_beta))
for (i in 1:length(ratio_beta.mean)) {
  ratio_beta[[i]] <- ratio_beta[[i]][!is.na(ratio_beta[[i]]$beta), ]
  ratio_beta.mean[i] <- mean(ratio_beta[[i]]$beta, na.rm = T)
  ratio_beta.median[i] <- median(ratio_beta[[i]]$beta, na.rm = T)
}
names(ratio_beta.mean) <- names(ratio_beta)
names(ratio_beta.median) <- names(ratio_beta)


# hist plot
p1 <- list()
xplot.beta <- function(ratio_beta, xtissue, tissueClass) {
  tcolor <- tissueClass[xtissue, ]$color
  tmp <-  ggplot(, aes(x = ratio_beta[[xtissue]]$beta)) + theme_classic() +
    geom_vline(xintercept = 0, linetype = "dashed", color = "blue", size = 1) +
    geom_density(alpha = .6, fill = tcolor) + lghplot.addtheme() +
    theme(axis.line = element_line(size = 1.2)) + 
    scale_x_continuous(breaks = seq(-0.06, 0.06, by = 0.06)) +
    xlim(c(-0.065, 0.065)) +
    
    xlab("") + ylab("") + ggtitle(xtissue)
  return(tmp)
}

index <- sort.int(ratio_beta.median, decreasing = F, index.return = T)$ix
vclass <- tissueClass[names(ratio_beta.mean)[index], ]
idx <- sort.int(vclass$color, decreasing = T, index.return = T)$ix
vclass <- vclass[idx, ]
for (i in 1:nrow(vclass)) {
  xtissue <- vclass$tissue[i]
  p1[[i]] <- xplot.beta(ratio_beta, xtissue, tissueClass)
}
pdf(file = "./results/ratio/FigureSA_beta_ratio_distribute_based_on_promet_sortby.pdf",
  width = 25, height = 12.5)
grid.arrange(arrangeGrob(grobs = p1, ncol = 8,
  top = textGrob(
    "Distribution of changed translation efficiency to degradation (cTED) with aging",
    gp = gpar(fontface = "bold",  fontsize = 30)),
  bottom = textGrob("Changed normalized TED",
    gp = gpar(fontface = "bold",  fontsize = 30)),
  left = textGrob("Density",
    gp = gpar(fontface = "bold",  fontsize = 30),
    rot = 90)))
dev.off()

# just for plot visualization
grid.arrange(arrangeGrob(grobs = p1, ncol = 8,
  top = textGrob(
    "Distribution of changed translation efficiency to degradation (cTED) with aging",
    gp = gpar(fontface = "bold",  fontsize = 30)),
  bottom = textGrob("Changed normalized TED",
    gp = gpar(fontface = "bold",  fontsize = 30)),
  left = textGrob("Density",
    gp = gpar(fontface = "bold",  fontsize = 30),
    rot = 90)))
```

```{r,fig.width=12, fig.height=9, out.width="85%", fig.align="center",dpi = 144}
ratio_updown.beta <- zeros(length(ratio_beta), 2)
tissues <- names(ratio_beta)
rownames(ratio_updown.beta) <- tissues
colnames(ratio_updown.beta) <- c("pec_up", "pec_down")
for (i in 1:length(tissues)) {
  tmp <- ratio_beta[[i]]$beta
  tmpN <- sum(!is.na(tmp))
  ratio_updown.beta[i, 1] <- sum(tmp > 0, na.rm = T) / tmpN
  ratio_updown.beta[i, 2] <- sum(tmp < 0, na.rm = T) / tmpN
}
index <- sort.int(ratio_beta.median, decreasing = F, index.return = T)$ix
ratio_updown.beta <- ratio_updown.beta[index, ]

tmp <- reshape2::melt(ratio_updown.beta, stringsAsFactor = F)
colnames(tmp) <- c("tissues", "Direction", "perctage")
tmp$tissues <- factor(tmp$tissues, levels = rownames(vclass))
pdf("./results/ratio/Figure 6E_tissue_trans_effectiveness_beta_ratio.pdf",
  width = 12, height = 9)
pp <- ggplot(data = tmp, aes(x = tissues, y = perctage, fill = Direction)) + theme_classic() +
  geom_col(position = "fill", alpha = 0.8) +
  ylab("Relative ratio") + xlab("") +
  scale_fill_manual(values = c("#BB0021", "#3B4992")) +
  lghplot.addtheme(legend.position = "top", hjust = 1, size = 18) +
  ggtitle("Relative ratio of up- and down- TEDs") +
  theme(legend.text = element_text(size = 18))
print(pp)
dev.off()
print(pp)
```

```{r,fig.width=7, fig.height=5, out.width="75%", fig.align="center",dpi = 144}
tmpGeneList <- names(mfuzz.promet.whole$cluster)[mfuzz.promet.whole$cluster > 0 &
  mfuzz.promet.whole$cluster < 9]
tmpGeneList <- names(mfuzz.promet.whole$cluster)
probeta <- Metapro[is.element(Metapro$ID, tmpGeneList) & !Metapro$manyNA,
  substr(colnames(Metapro), 1, 5) == "beta_"]
colnames(probeta) <- gsub("beta_", "", colnames(probeta))

propvalue <- Metapro[is.element(Metapro$ID, tmpGeneList) & !Metapro$manyNA,
  substr(colnames(Metapro), 1, 7) == "Pvalue_"]
colnames(propvalue) <- gsub("Pvalue_", "", colnames(propvalue))
propvalue[is.na(propvalue)] <- 1
# probeta[is.na(probeta)] = 0
colmeanBeta <- colMeans(probeta, na.rm  = T)


tmpdata <- data.frame(ratioChange = ratio_beta.mean, stringsAsFactors = F,
  meanBeta = colmeanBeta[names(ratio_beta.mean)],
  tissues = names(ratio_beta.mean))
rownames(tmpdata) <- tmpdata$tissues
tmpdata$tissueclass <- tissueClass[rownames(tmpdata), ]$type
tmp1 <- tmpdata
cor.test(tmp1$ratioChange, tmp1$meanBeta)
pdf("./results/ratio/Figure6F_overall_cTED_vs_mean_aging_beta.pdf",
  width = 7, height = 5)
p <- ggplot(tmp1, aes(ratioChange, meanBeta, color = tissueclass)) +
  theme_classic() +
  geom_point(size = 6, aes(color = tissueclass)) +
  lghplot.addtheme() + geom_smooth(color = 4, size = 0.5, method = "lm") +
  geom_text_repel(aes(label = tissues), size = 4, box.padding = 0.5) +
  scale_color_manual(values = c("#FF6347", "#1E90FF", "#DAB700")) +
  theme(axis.line = element_line(size = 1.2)) +
  xlab("Averaged changed translation effeciency score") +
  ylab("Mean protein aging effect size (beta lm)") +
  ggtitle("Correlation between mean beta and cTED")
print(p)
dev.off()
# just for plot visualization
print(p)
```

## 11.4 TED pathways in different types

```{r}
bp_gene_sets <- msigdbr(species = "Homo sapiens",
  category = "C5", subcategory = "GO:BP")
gene_sets_bp <- split(bp_gene_sets$gene_symbol, bp_gene_sets$gs_name)
gsea_tissues.ratio.bp <- list()
ratiobeta <- DEratio.limma$MetaLimma[,
  substr(colnames(DEratio.limma$MetaLimma), 1, 5) == "beta_"]
colnames(ratiobeta) <- gsub("beta_", "", colnames(ratiobeta))
for (i in 1:ncol(ratiobeta)) {
  thisbeta <- ratiobeta[, i]
  names(thisbeta) <- rownames(ratiobeta)
  thisbeta <- thisbeta[!is.na(thisbeta)]
  thisbeta <- sort(thisbeta, decreasing = TRUE)
  gsea_tissues.ratio.bp[[i]] <- fgsea(pathways = gene_sets_bp, stats = thisbeta)
}
names(gsea_tissues.ratio.bp) <- colnames(ratiobeta)
gsea_tissues.ratio.bp <- gsea_tissues.ratio.bp[names(pro.tissues.v)]
```

```{r}
kegg_gene_sets <- msigdbr(species = "Homo sapiens",
  category = "C2", subcategory = "CP:KEGG")
gene_sets_kegg <- split(kegg_gene_sets$gene_symbol, kegg_gene_sets$gs_name)
gsea_tissues.ratio.kegg <- list()
ratiobeta <- DEratio.limma$MetaLimma[,
  substr(colnames(DEratio.limma$MetaLimma), 1, 5) == "beta_"]
colnames(ratiobeta) <- gsub("beta_", "", colnames(ratiobeta))
for (i in 1:ncol(ratiobeta)) {
  thisbeta <- ratiobeta[, i]
  names(thisbeta) <- rownames(ratiobeta)
  thisbeta <- thisbeta[!is.na(thisbeta)]
  thisbeta <- sort(thisbeta, decreasing = TRUE)
  gsea_tissues.ratio.kegg[[i]] <- fgsea(pathways = gene_sets_kegg, stats = thisbeta)
}
names(gsea_tissues.ratio.kegg) <- colnames(ratiobeta)
gsea_tissues.ratio.kegg <- gsea_tissues.ratio.kegg[names(pro.tissues.v)]
```

```{r}
cnames <- c("GOBP_TRANSLATIONAL_INITIATION",
  "GOBP_TRANSLATIONAL_ELONGATION", "GOBP_TRANSLATIONAL_TERMINATION")

bpratios <- matrix(0, length(cnames), length(gsea_tissues.ratio.bp))
for (i in 1:length(gsea_tissues.ratio.bp)) {
  tmpaa <- as.data.frame(gsea_tissues.ratio.bp[[i]])
  rownames(tmpaa) <- tmpaa$pathway
  bpratios[, i] <- tmpaa[cnames, ]$NES
}
rownames(bpratios) <- cnames
colnames(bpratios) <- names(gsea_tissues.ratio.bp)
bpratios.info <- tissueClass[colnames(bpratios), ]

# kegg
cnames <- c("KEGG_LYSOSOME",
  "KEGG_PROTEASOME", "KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS")
keggratios <- matrix(0, length(cnames), length(gsea_tissues.ratio.kegg))
for (i in 1:length(gsea_tissues.ratio.kegg)) {
  tmpaa <- as.data.frame(gsea_tissues.ratio.kegg[[i]])
  rownames(tmpaa) <- tmpaa$pathway
  keggratios[, i] <- tmpaa[cnames, ]$NES
}
rownames(keggratios) <- cnames
colnames(keggratios) <- names(gsea_tissues.ratio.kegg)
keggratios.info <- tissueClass[colnames(keggratios), ]
```

```{r,fig.width=10, fig.height=3, out.width="100%", fig.align="center",dpi = 144}
rationes <- rbind(bpratios, keggratios)
rationes <- rationes[, rownames(tissueClass)]
rownames(rationes) <- gsub("GOBP_", "", rownames(rationes))
rownames(rationes) <- gsub("KEGG_", "", rownames(rationes))
rownames(rationes) <- gsub("_", " ", rownames(rationes))
rownames(rationes) <- capitalize(tolower(rownames(rationes)))

vids <- c(which(tissueClass$type == "Type II"),
  which(tissueClass$type == "Undefined"),
  which(tissueClass$type == "Type I"))
tclass <- data.frame(tissueType = tissueClass$type, row.names = rownames(tissueClass))
annotation_colors <- list(tissueType = c("Type I" = "#FF6347",
  "Type II" = "#1E90FF",
  "Undefined" = "#DAB700"))
colnames(tclass) <- c("tissueType")
dentx <- pheatmap::pheatmap(rationes[, vids], cluster_rows = F,
  annotation_col = tclass,
  annotation_colors = annotation_colors,
  cluster_cols = F, fontsize_row = 9, fontsize_col = 10,
  fontsize = 10, treeheight_row = 20,
  treeheight_col = 10, legend = T,
  color = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(20),
  # file ="./results/ratio/Figure S14A_enrichment_GSEA_all_tissue_TEDs.pdf",
  height = 3, width = 10
)
pdf("./results/ratio/Figure S14A_enrichment_GSEA_all_tissue_TEDs.pdf",
  height = 3, width = 10)
print(dentx)
dev.off()

# just for plot visualization
grid::grid.draw(dentx$gtable)
```

```{r}
cor.test(rationes[1, ], ratio_beta.mean, method = "spearman")
cor.test(rationes[2, ], ratio_beta.mean, method = "spearman")
cor.test(rationes[3, ], ratio_beta.mean, method = "spearman")
cor.test(rationes[4, ], ratio_beta.mean, method = "spearman")
cor.test(rationes[5, ], ratio_beta.mean, method = "spearman")
cor.test(rationes[6, ], ratio_beta.mean, method = "spearman")
```

```{r,fig.width=10.5, fig.height=12, out.width="65%", fig.align="center",dpi = 144}

## ribosome
ribosome <- file2frame("./data/ribosome_proteins_from_kegg.txt", header = F)
ribosome <- ribosome$V1
profc.all <- Metapro[, substr(colnames(Metapro), 1, 5) == "beta_"]
colnames(profc.all) <- gsub("beta_", "", colnames(profc.all))
ribosome.matrix <- profc.all[intersect(ribosome, rownames(profc.all)), ]
idx <- rowSums(is.na(ribosome.matrix)) < 5
ribosome.matrix <- ribosome.matrix[idx, ]

enbrks <- 0.1 * c(-1, -0.8, -0.6, -0.4, -0.2, -0.1, -0.05, -0.02,
  0.02, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1)

col_fun <- colorRamp2(
  breaks = enbrks,
  colors = colorRampPalette(c("#4DBBD5FF", "gray95", "#E64B35FF"))(length(enbrks))
)
column_ha <- HeatmapAnnotation(tissueType = tissueType$type,
  col = list(tissueType = c("Type I" = "#FF6347",
    "Type II" = "#1E90FF",
    "Undefined" = "#DAB700")))

pdf("./results/ratio/Figure S14C_ribosome.pdf", width = 10.5, height = 12)

pp <- Heatmap(
  ribosome.matrix[order(rownames(ribosome.matrix)), ],
  name = "Beta age",
  col = col_fun,
  cluster_rows = FALSE,

  column_title = "Change in initiation and elongation factors",
  top_annotation = column_ha,
  row_names_gp = gpar(fontsize = 14),
  column_names_gp = gpar(fontsize = 14),
  column_names_rot = 45,
  rect_gp = gpar(col = "gray50", lwd = 1),
  # border_gp = gpar(col = "gray50", lwd = 1),
  column_dend_height = unit(10, "mm"),
)
print(pp)

dev.off()

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(pp))
grid.draw(heatmap_grob)
```

```{r,fig.width=10.5, fig.height=12, out.width="65%", fig.align="center",dpi = 144}
# EIFs and EEFs

trans_initial <- rownames(profc.all)[substr(rownames(profc.all), 1, 3) %in% c("EIF", "EEF")]
trans.matrix <- profc.all[trans_initial, ]
idx <- rowSums(is.na(trans.matrix)) < 5
trans.matrix <- trans.matrix[idx, ]

pdf("./results/ratio/Figure S14B_trans_initiation.pdf", width = 10.5, height = 11)

pp <- Heatmap(
  trans.matrix[order(rownames(trans.matrix)), ],
  name = "Beta age",
  col = col_fun,
  cluster_rows = FALSE,

  column_title = "Change in initiation and elongation factors",
  top_annotation = column_ha,
  row_names_gp = gpar(fontsize = 14),
  column_names_gp = gpar(fontsize = 14),
  column_names_rot = 45,
  rect_gp = gpar(col = "gray50", lwd = 1),
  column_dend_height = unit(10, "mm"),
)
print(pp)

dev.off()

# just for plot visualization
heatmap_grob <- grid.grabExpr(draw(pp))
grid.draw(heatmap_grob)
```


# 12 Experiment validation

### 12.1 P16/P21

```{r}
p16 <- file2frame("./data/P16_tissues_v20230419.txt")
utissue <- unique(p16$Tissue)
tpvalues <- data.frame(tissue = utissue,
  p_Juvenile_vs_Elderly = rep(1, length(utissue)),
  p_Young_vs_Elderly = rep(1, length(utissue)))
breaks_A <- c(0.01, 0.08, 0.08, 0.08, 0.08, 0.01, 0.01, 0.01)
breaks_B <- c(3, 3, 3, 3, 2, 2, 0.06, 3)
for (i in 1:length(utissue)) {
  tmpdata <- p16[p16$Tissue == utissue[i], ]
  tmpdata$Area.Density.log2 <- tmpdata$Area.Density * 1e3
  tmpdata$class <- factor(x = tmpdata$class,
    levels = c("Juvenile", "Young", "Elderly"))
  tmpp1 <- t.test(Area.Density ~ class, data =
    tmpdata, subset = tmpdata$class != "Young")$p.value
  tmpp2 <- t.test(Area.Density ~ class, data =
    tmpdata, subset = tmpdata$class != "Juvenile")$p.value
  tpvalues$p_Juvenile_vs_Elderly[i] <- tmpp1
  tpvalues$p_Young_vs_Elderly[i] <- tmpp2

  df2 <- data.frame(tmean = aggregate(tmpdata$Area.Density.log2,
    by = list(tmpdata$class), FUN = mean, na.rm = T)$x,
  tsd = aggregate(tmpdata$Area.Density.log2,
    by = list(tmpdata$class), FUN = sd, na.rm = T)$x,
  class = factor(x = c("Juvenile", "Young", "Elderly"),
    levels = c("Juvenile", "Young", "Elderly"))
  )

  thepath <- paste0("./results/experiments/Figure5x_p16_stats", utissue[i], ".pdf")
  pdf(thepath)
  p <- ggplot(df2, aes(x = class, y = tmean, fill = class)) +
    geom_bar(stat = "identity", color = "black",
      position = position_dodge(), alpha = 0.5, size = 1.5) +
    geom_errorbar(aes(ymin = tmean - tsd, ymax = tmean + tsd),
      width = .3, size = 1.5) +
    lghplot.addtheme(size = 28, sizex = 26, sizey = 26) +
    scale_fill_manual(values = c("#008B45FF", "#3B4992FF", "#EE0000FF")) +
    theme(axis.line = element_line(size = 1.2)) + xlab("") +
    ylab(bquote("Area Density " ~ italic(x10)^italic(3))) +
    ggtitle(utissue[i])
  if (i <= 8) {
    p <- p + scale_y_break(breaks = c(breaks_A[i], breaks_B[i]),
      scales = "free", space  = 0.2)
  }
  print(p)
  dev.off()
}

tpvalues
```

```{r}
# p21
p21 <- file2frame("./data/P21_tissues_v20230419.txt")
breaks_A <- c(0.05, 0.4, 0.4, 0.4, 0.1, 0.4, 0.4, 0.05)
breaks_B <- c(1, 4, 4, 4, 2, 4, 4, 1)
utissue <- unique(p21$Tissue)
tpvalues <- data.frame(tissue = utissue,
  p_Juvenile_vs_Elderly = rep(1, length(utissue)),
  p_Young_vs_Elderly = rep(1, length(utissue)))

for (i in 1:length(utissue)) {
  tmpdata <- p21[p21$Tissue == utissue[i], ]
  tmpdata$Area.Density.log2 <- tmpdata$Area.Density * 1e3
  tmpdata$class <- factor(x = tmpdata$class,
    levels = c("Juvenile", "Young", "Elderly"))
  tmpp1 <- t.test(Area.Density ~ class, data = tmpdata,
    subset = tmpdata$class != "Young")$p.value
  tmpp2 <- t.test(Area.Density ~ class, data = tmpdata,
    subset = tmpdata$class != "Juvenile")$p.value
  tpvalues$p_Juvenile_vs_Elderly[i] <- tmpp1
  tpvalues$p_Young_vs_Elderly[i] <- tmpp2

  df2 <- data.frame(tmean = aggregate(tmpdata$Area.Density.log2,
    by = list(tmpdata$class), FUN = mean, na.rm = T)$x,
  tsd = aggregate(tmpdata$Area.Density.log2,
    by = list(tmpdata$class), FUN = sd, na.rm = T)$x,
  class = factor(x = c("Juvenile", "Young", "Elderly"),
    levels = c("Juvenile", "Young", "Elderly"))
  )
  thepath <- paste0("./results/experiments/Figure5x_p21_stats", utissue[i], ".pdf")
  pdf(thepath)
  p <- ggplot(df2, aes(x = class, y = tmean, fill = class)) +
    geom_bar(stat = "identity",
      color = "black", position = position_dodge(),
      alpha = 0.5, size = 1.5) +
    geom_errorbar(aes(ymin = tmean - tsd, ymax = tmean + tsd),
      width = .3, size = 1.5) +
    lghplot.addtheme(size = 28, sizex = 26, sizey = 26) +
    scale_fill_manual(values = c("#008B45FF", "#3B4992FF", "#EE0000FF")) +
    theme(axis.line = element_line(size = 1.2)) + xlab("") +
    ylab(bquote("Area Density " ~ italic(x10)^italic(3))) +
    ggtitle(utissue[i])
  if (i <= 8) {
    p <- p + scale_y_break(c(breaks_A[i], breaks_B[i]),
      scales = "free", space  = 0.2)
  }
  print(p)
  dev.off()
}

tpvalues
```

### 12.2 cell counts

```{r}
Cellcounts <- file2frame("./data/cell_counts_20240802.txt")

utissue <- unique(Cellcounts$Tissue)
tpvalues <- data.frame(tissue = utissue,
  p_Juvenile_vs_Elderly = rep(1, length(utissue)),
  p_Young_vs_Elderly = rep(1, length(utissue)))
for (i in 1:length(utissue)) {
  tmpdata <- Cellcounts[Cellcounts$Tissue == utissue[i], ]
  tmpdata$class <- factor(x = tmpdata$class,
    levels = c("Juvenile", "Young", "Elderly"))
  tmpp1 <- t.test(number ~ class, data = tmpdata,
    subset = tmpdata$class != "Young")$p.value
  tmpp2 <- t.test(number ~ class, data = tmpdata,
    subset = tmpdata$class != "Juvenile")$p.value
  tpvalues$p_Juvenile_vs_Elderly[i] <- tmpp1
  tpvalues$p_Young_vs_Elderly[i] <- tmpp2

  df2 <- data.frame(tmean = aggregate(tmpdata$number,
    by = list(tmpdata$class), FUN = mean, na.rm = T)$x,
  tsd = aggregate(tmpdata$number,
    by = list(tmpdata$class), FUN = sd, na.rm = T)$x,
  class = factor(x = c("Juvenile", "Young", "Elderly"),
    levels = c("Juvenile", "Young", "Elderly"))
  )
  thepath <- paste0("./results/experiments/Extendfigure_HE_cellcounts_",
    utissue[i], ".pdf")
  pdf(thepath)
  p <- ggplot(df2, aes(x = class, y = tmean, fill = class)) +
    geom_bar(stat = "identity", color = "black",
      position = position_dodge(), alpha = 0.5, size = 1.5) +
    geom_errorbar(aes(ymin = tmean - tsd, ymax = tmean + tsd),
      width = .3, size = 1.5) +
    lghplot.addtheme(size = 28, sizex = 26, sizey = 26) +
    scale_fill_manual(values = c("#008B45FF", "#3B4992FF", "#EE0000FF")) +
    theme(axis.line = element_line(size = 1.2)) + xlab("") +
    ylab("Number of parenchymal cells") + ggtitle(utissue[i])
  print(p)


  dev.off()

  thepath1 <- paste0("./results/experiments/Extendfigure_HE_cellcounts_",
    utissue[i], ".png")
  png(thepath1)
  print(p)
  dev.off()

}

tpvalues
```

# end
