---
title: "A multi-omics molecular landscape of 30 tissues in aging rhesus macaques(Macaca mulatta)"
author: "Gong-Hua Li"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: no
  pdf_document: 
    toc: yes
    latex_engine: xelatex
---

# 1. setwd and env

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE, message = FALSE)
```

```{r}
library(ggplot2)
library(reshape2)
#library(Rmisc)
library(Hmisc)
library(EnhancedVolcano)
library(pracma)
library(car)
library(ggrepel)
library(edgeR)
library(grid)
library(gridExtra)
library(Mfuzz)
library(M3C)
library(preprocessCore)
library(rlist)
library(ggsci)
library(scales)
library(data.table)
library(RColorBrewer)
library(readxl)
#setwd("/home/ligh/github/code_forpublication/macaca_multiple_tissue/")
source('./subroutines.R')
source('./subroutines_for_MCMT_aging.R')
```

# 2. load data

```{r}
#whole body data
load('./data/pro.whole.fdr0.01_v20210108_from_NOVO_remap_solid_tissues.Rdata')
load('./data/mrna.whole_v20210108_solid_tissues.Rdata')
load('./data/met_whole_from_novo_v20210108_solid_tissues.Rdata')

# tissue data
load('./data/pro.tissues_v20210108_solid_tissues.Rdata')
load('./data/mrna.tissues_v20210108_solid_tissues.Rdata')
load('./data/met.tissues_v20210108_solid_tissues_new.Rdata')


load('./data/met.header.all.hmdb_curated.Rdata')
idx = !is.na(met.header.all.hmdb$hmdbid_highconfidence)
met.header.all.hmdb.v = met.header.all.hmdb[idx,]

```

# 3. quality control

```{r}
# protein remove outliners
tmp = pro.whole
pp = prcomp(t(tmp),cor=F)
outlinerids = c()
thetissues = unique(pro.whole.info$tissue_en)
for(i in 1:length(thetissues)){
    idx = pro.whole.info$tissue_en == thetissues[i]
    tmpinfo = pro.whole.info[idx,]
    outx = is.outliner(pp$x[idx,1])
    outy = is.outliner(pp$x[idx,2])
    #outy = is.outliner(pp$x[idx,2],coef = 3)
    if(sum(outx | outy) > 0){
        outlinerids = c(outlinerids,rownames(tmpinfo)[outx | outy])
    } 
}
outlinerids
vid = !is.element(colnames(pro.whole),outlinerids)
pro.whole = pro.whole[,vid]
pro.whole.info = pro.whole.info[vid,]

```

```{r}
# met remove outliners
tmp = met.whole
pp = prcomp(t(tmp),cor=F)
outlinerids = c()
thetissues = unique(met.whole.info$tissue_en)
for(i in 1:length(thetissues)){
    idx = met.whole.info$tissue_en == thetissues[i]
    tmpinfo = met.whole.info[idx,]
    outx = is.outliner(pp$x[idx,1])
    outy = is.outliner(pp$x[idx,2])
    #outy = is.outliner(pp$x[idx,2],coef = 3)
    if(sum(outx | outy) > 0){
        outlinerids = c(outlinerids,rownames(tmpinfo)[outx | outy])
    } 
}
outlinerids
vid = !is.element(colnames(met.whole),outlinerids)
met.whole = met.whole[,vid]
met.whole.info = met.whole.info[vid,]
```

```{r}
# mrna remove outliners
tmp = mrna.whole
pp = prcomp(t(tmp),cor=F)
outlinerids = c()
thetissues = unique(mrna.whole.info$tissue_en)
for(i in 1:length(thetissues)){
    idx = mrna.whole.info$tissue_en == thetissues[i]
    tmpinfo = mrna.whole.info[idx,]
    outx = is.outliner(pp$x[idx,1])
    outy = is.outliner(pp$x[idx,2])
    #outy = is.outliner(pp$x[idx,2],coef = 3)
    if(sum(outx | outy) > 0){
        outlinerids = c(outlinerids,rownames(tmpinfo)[outx | outy])
    } 
}
outlinerids
vid = !is.element(colnames(mrna.whole),outlinerids)
mrna.whole = mrna.whole[,vid]
mrna.whole.info = mrna.whole.info[vid,]
```

# 4. set colors

```{r}
alltissues = names(pro.tissues)

tissue.systems = c('Integumentary','Endocrine','Brain','Respiratory','Digestive',
             'Cardiovascular','Cardiovascular','Brain','Digestive','Endocrine',
             'Cardiovascular','Muscle','Reproductive','Digestive','Brain',
             'Immune','Renal','Endocrine','Digestive','Reproductive',
             'Digestive','Brain','Muscle','Immune','Integumentary','Endocrine',
            'Brain','Immune','Cardiovascular','Reproductive')
names(tissue.systems) = alltissues
tissue.color = pal_npg()(10)
names(tissue.color) = levels(factor(tissue.systems))

tissue.systems
tissue.color
```

```{r}
mypal = pal_aaas()(10)
mypal
show_col(mypal)
```

# 5. Figure 1: overall data distribution

## 5.1 Figure 1A flowchart

## 5.2 Figure 1B

```{r}
num_omics = data.frame(num_mrna = rep(0,length(alltissues)),stringsAsFactors = F,
                       tissues = alltissues,
                       tissue_systems = tissue.systems,
                      num_protein = rep(0,length(alltissues)),
                       num_met = rep(0,length(alltissues))
                       )
for(i in 1:length(alltissues)){
    num_omics$num_mrna[i] = nrow(mrna.tissues[[alltissues[i]]])
    num_omics$num_protein[i] = nrow(pro.tissues[[alltissues[i]]])
    num_omics$num_met[i] = nrow(met.tissues[[alltissues[i]]])
}
rownames(num_omics) = alltissues
```

```{r}
idx = sort.int(num_omics$num_protein,decreasing = F,index.return = T)$ix
num_omics.v = num_omics[idx,]
idx = sort.int(num_omics.v$tissue_systems,decreasing = F,index.return = T)$ix
num_omics.v = num_omics.v[idx,]
writetxt(num_omics.v[,c(2,3,1,4,5)],'./out/20230217_aging/Figure1_Overall_data_distribution/Figure_1B_number_omics_data.txt')
```

```{r}
idx = sort.int(num_omics$num_protein,decreasing = F,index.return = T)$ix
num_omics.v = num_omics[idx,]
idx = sort.int(num_omics.v$tissue_systems,decreasing = F,index.return = T)$ix
num_omics.v = num_omics.v[idx,]

pomics = list()
pomics[[1]] = ggplot(num_omics.v,aes(x = factor(tissues,level = tissues),
                                     y = num_met,
                       color = tissue_systems,fill = tissue_systems))+ 
  geom_bar(stat="identity")+
theme_classic()+lghplot.addtheme(legend.position = 'none',hjust = 1,size = 10)+ 
  scale_color_aaas()+scale_fill_aaas()+ 
 theme(axis.text.x=element_blank(),axis.title.x=element_blank())+ 
  ylab('Number of metabolites')


pomics[[2]] = ggplot(num_omics.v,aes(x = factor(tissues,level = tissues),
                                     y = num_protein,
                       color = tissue_systems,fill = tissue_systems))+ 
  geom_bar(stat="identity")+
theme_classic()+lghplot.addtheme(legend.position = 'none',hjust = 1,size = 10)+
  scale_color_aaas()+scale_fill_aaas()+
ylab('Number of proteins')+ xlab('')

#pdf(file = './out/20230217_aging/Figure1_Overall_data_distribution/
#    Figure1B_number_of_omicsV__promet.pdf',width = 7,height = 5)
grid.arrange(arrangeGrob(grobs = pomics,ncol = 1,
                         top = 'Identified proteins and metabolites',
                         heights = c(3.2,5.8)))

#dev.off()

```

```{r}
# total identified molecules
vmrna = c()
for (i in 1:length(mrna.tissues)){
    vmrna = unique(c(vmrna,rownames(mrna.tissues[[i]])))
}
length(vmrna)


vproteins = c()
for (i in 1:length(pro.tissues)){
    vproteins = unique(c(vproteins,pro.tissues.header[[i]]$Gene))
}
length(vproteins)

vmets = c()
for (i in 1:length(met.tissues)){
    vmets = unique(c(vmets,rownames(met.tissues[[i]])))
}
length(vmets)
```

## 5.3 Figure 1C_1E

### 5.3.1 Figure 1C mRNA

```{r Figure_1C, fig.height = 4, fig.width = 4, fig.align = "center"}
### for mRNA
mrna.whole.std = standardise_matrix(mrna.whole)

p = tsne(mrna.whole.std,labels=tissue.systems[mrna.whole.info$tissues],
         legendtextsize = 10,dotsize = 2)
p = p + theme_classic()+lghplot.addtheme(legend.position = 'none')+ 
  scale_color_aaas()+ xlab('tsne 1') + ylab('tsne 2')+
theme(axis.line = element_line(size = 1.0)) + ggtitle('Transcriptome')
#pdf(file ="./out/20230217_aging/Figure1_Overall_data_distribution/
#    Figure1C_tSNE_mrna_using_standardised.pdf",height = 4,width = 4)
print(p)
#dev.off()

```

### 5.3.2 Figure 1D protein

```{r Figure_1D, fig.height = 4, fig.width = 4, fig.align = "center"}

ida = rowSums(pro.whole < 0.1) < 0.2 * ncol(pro.whole)
sum(ida)
pro.whole.cons = pro.whole[ida,]
pro.whole.std = standardise_matrix(pro.whole.cons)

p = tsne(pro.whole.std,labels=tissue.systems[pro.whole.info$tissue_en],
         legendtextsize = 10,dotsize = 2)
p = p + theme_classic()+lghplot.addtheme(legend.position = 'none')+ 
  scale_color_aaas()+ xlab('tsne 1') + ylab('tsne 2') + ggtitle('Proteome')
#pdf(file ="./out/20230217_aging/Figure1_Overall_data_distribution/
#    Figure1D_tSNE_protein_using_standardised.pdf",height = 4,width = 4)
print(p)
#dev.off()

```

### 5.3.3 Figure 1E metabolism

```{r Figure_1E, fig.height = 4, fig.width = 4, fig.align = "center"}
met.whole.std = standardise_matrix(met.whole)
p = tsne(met.whole.std,labels=tissue.systems[met.whole.info$Tissues],
         legendtextsize = 10,dotsize = 2)
p = p + theme_classic()+lghplot.addtheme(legend.position = 'none')+ 
  scale_color_aaas()+ xlab('tsne 1') + ylab('tsne 2')+
theme(axis.line = element_line(size = 1.0)) + ggtitle('Metabolome')
#pdf(file ="./out/20230217_aging/Figure1_Overall_data_distribution/
#    Figure1E_tSNE_met_using_std.pdf",height = 4,width = 4)
print(p)
#dev.off()
```

# 6 Figure 2: tissue aging DEGs and GO

```{r}
# sub routine plot volcano
plot_Volcano <- function(res2, title){
    

  tmpup = res2$Pvalue
  tmpup[is.na(tmpup)] = 1
  tmpup[res2$log2FC < 0] = 1
  sortid_up = sort.int(-log10(tmpup),decreasing = T,index.return = T)$ix
  tmpid.up = res2$ID[sortid_up[1:5]]

  tmpdown = res2$Pvalue
  tmpdown[is.na(tmpdown)] = 1
  tmpdown[res2$log2FC > 0] = 1
  sortid_down = sort.int(-log10(tmpdown),decreasing = T,index.return = T)$ix
  tmpid.down = res2$ID[sortid_down[1:5]]

  vid = is.element(res2$ID,c(tmpid.up,tmpid.down))
  tlab = res2$ID
  tlab[!vid] = NA

  keyvals <- rep('gray50', nrow(res2))
  names(keyvals) <- rep('NS', nrow(res2))   

  keyvals[which(res2$log2FC > 0.58 & res2$Pvalue < 0.05)] <- "Brown"    
  names(keyvals)[which(res2$log2FC > 0.58 & res2$Pvalue < 0.05)] <- 'High'    

  keyvals[which(res2$log2FC < -0.58 & res2$Pvalue < 0.05)] <- "darkblue"    
  names(keyvals)[which(res2$log2FC < -0.58 & res2$Pvalue < 0.05)] <- 'Low'   
  p = EnhancedVolcano(res2,
                        lab = tlab,
                        x = 'log2FC',
                        y = 'Pvalue',
                        caption = NULL,
                        title = title,
                        border = 'full',
                        titleLabSize = 12,
                        FCcutoff = 0.58,
                        cutoffLineWidth = F,
                        axisLabSize = 12,
                        subtitle = NULL,
                        cutoffLineCol = 'white',
                        gridlines.minor = F,
                        gridlines.major = F,
                        xlab = bquote(~Log[2]~ 'fold change'),
                        pCutoff = 0.05,
                        colCustom = keyvals,
                        colAlpha = 4/5,
                        legendPosition = 'none',
                        legendLabSize = 5,
                        legendIconSize = 3,
                        drawConnectors = TRUE,
                        widthConnectors = 0.5,
                        pointSize = -0.3*log10(res2$Pvalue),labSize = 3,
                        colConnectors = 'black')
    return(p)
}


list_to_matrix <- function(DEproFC,alltissues){
    DEproFC_matrix = list()
    for(i in 1:length(alltissues)){
        tmp = matrix(DEproFC[[i]],1,length(DEproFC[[i]]))
        tmp = as.data.frame(tmp)
        colnames(tmp) = names(DEproFC[[i]])
        DEproFC_matrix[[i]] = tmp
    }
    DEproFC_matrix = t(as.matrix(rbindlist(DEproFC_matrix,fill = T)))
    colnames(DEproFC_matrix) = names(DEproFC)
    #vid = rowSums(is.na(DEproFC_matrix)) < ncol(DEproFC_matrix)/2
    #DEproFC_matrix = DEproFC_matrix[vid,]
    return(DEproFC_matrix)
}

met.class_enrichment <- function(mets,annote){
  require(clusterProfiler)

  vmet = intersect(mets,rownames(annote))
  fluxgmt = data.frame(ont = annote$sub_class,
                       gene = rownames(annote),stringsAsFactors = F) 
  
  Recon3D <- enricher(gene = vmet,
                TERM2GENE=fluxgmt,
                pAdjustMethod = "BH",
                minGSSize = 1,
                pvalueCutoff =1,
                qvalueCutoff = 1
                )
  Recon3Dout = Recon3D@result
  Recon3Dout$DB = rep('HMDBclass',dim(Recon3Dout)[1])
  n = dim(Recon3Dout)[2]
  Recon3Dout = Recon3Dout[,c(n,1:(n-1))]
  return(Recon3Dout)
}
```

## 6.1 protein

```{r}
DEproFC = list()
DEproPvalue = list()
DEproAging = list()
DEproFC.develop = list()
DEproPvalue.develop = list()
for(i in 1:length(alltissues)){
    #Mfuzz
    thistissue = alltissues[i]
    thispro = pro.tissues[[thistissue]]
    thispro.header = pro.tissues.header[[thistissue]]
    thispro = delete_dup_genes_forprotein(thispro,pro.tissues.header[[thistissue]])
    thispro.info = pro.tissues.info[[thistissue]]
    thisDEpro = DEGenes.simplified(thispro,catagory = thispro.info$stage == 4,
                    subset = thispro.info$stage == 4 | thispro.info$stage == 1)
    thisDEpro.develop = DEGenes.simplified(thispro,catagory = thispro.info$stage == 2,
                    subset = thispro.info$stage == 2 | thispro.info$stage == 1)
    forsort = thisDEpro$Pvalue
    forsort[is.na(forsort)] = 1
    idx = sort.int(forsort,decreasing = F,index.return = T)$ix
    DEproAging[[i]] = thisDEpro[idx,-5]
    DEproAging[[i]]$log2FC.devControl = thisDEpro.develop$log2FC
    DEproAging[[i]]$Pvalue.devControl = thisDEpro.develop$Pvalue
    
    DEproFC[[i]] = thisDEpro$log2FC
    names(DEproFC[[i]]) = rownames(thisDEpro)
    DEproPvalue[[i]] = thisDEpro$Pvalue
    names(DEproPvalue[[i]])  = rownames(thisDEpro)
    
    DEproFC.develop[[i]] = thisDEpro.develop$log2FC
    names(DEproFC.develop[[i]]) = rownames(thisDEpro.develop)
    DEproPvalue.develop[[i]] = thisDEpro.develop$Pvalue
    names(DEproPvalue.develop[[i]])  = rownames(thisDEpro.develop)
}
names(DEproAging) = alltissues
names(DEproFC) = alltissues
names(DEproPvalue) = alltissues
names(DEproFC.develop) = alltissues
names(DEproPvalue.develop) = alltissues
```

```{r}
proteinVoconoPlot = list()
names.DEproAging = names(DEproAging)
for(i in 1:length(DEproAging)){
    res2 = DEproAging[[i]]
    
    proteinVoconoPlot[[i]] = plot_Volcano(res2,names.DEproAging[i])
}
```

```{r Figure_S2_DEpro, fig.height = 18, fig.width = 16, fig.align = "center"}
pdf(file = './out/20230217_aging/Figure2_DEG_GO_tissue/
    Figure2A_DEpro_each_tissueV1.pdf',width = 3*5+1,height =3*6)
grid.arrange(arrangeGrob(grobs = proteinVoconoPlot,ncol = 5))
dev.off()
```

```{r}
#Data S2
openxlsx::write.xlsx(DEproAging, file = "./out/20230217_aging/Figure2_DEG_GO_tissue/Data S2_DE_tissue_Aging_pro.xlsx")

# reducce size Data S2
tpath = "./out/20230217_aging/Figure2_DEG_GO_tissue/Data S2_DE_tissue_Aging_pro.xlsx"
sheetNames = openxlsx::getSheetNames(tpath)
xx = list()
for(i in 1:length(sheetNames)){
    tmp = openxlsx::read.xlsx(tpath,sheet = sheetNames[i])
    tmp[c(2,3,4,7,8)] = signif(tmp[c(2,3,4,7,8)],3)
    tmp = tmp[,-c(5,6)]
    xx[[i]] = tmp
}
names(xx) = sheetNames
openxlsx::write.xlsx(xx, file = "./out/20230217_aging/Figure2_DEG_GO_tissue/reduce_Data S2_DE_tissue_Aging_pro.xlsx")
```

```{r}
# num up and down proteins
DEproFC_matrix = list_to_matrix(DEproFC,alltissues)
DEproPvalue_matrix = list_to_matrix(DEproPvalue,alltissues)

DEproFC_matrix.develop = list_to_matrix(DEproFC.develop,alltissues)
DEproPvalue_matrix.develop = list_to_matrix(DEproPvalue.develop,alltissues)

Aging_pro_sigup_matrix = (DEproFC_matrix > 0.58 & DEproPvalue_matrix < 0.05  
                          & DEproPvalue_matrix.develop > 0.05) +0
Aging_pro_sigdown_matrix = -((DEproFC_matrix < -0.58 & DEproPvalue_matrix < 0.05 
                              & DEproPvalue_matrix.develop > 0.05) +0)
Aging_pro_sigall_matrix = Aging_pro_sigup_matrix + Aging_pro_sigdown_matrix

Aging_pro_updown = data.frame(stringsAsFactors = F,
                       num.up = colSums(Aging_pro_sigup_matrix,na.rm =T)/
                          colSums(!is.na(Aging_pro_sigup_matrix)),
                       num.down = colSums(Aging_pro_sigdown_matrix,na.rm =T)/
                          colSums(!is.na(Aging_pro_sigdown_matrix)),
                       num.all = colSums(abs(Aging_pro_sigall_matrix),na.rm =T)/
                         colSums(!is.na(Aging_pro_sigall_matrix)),
                        tissues = colnames(Aging_pro_sigup_matrix),
                        tissue_systems = tissue.systems)
rownames(Aging_pro_updown) = Aging_pro_updown$tissues

Aging_pro_updown
mean(Aging_pro_updown$num.all)
```

## 6.2 mRNA

```{r}
DEmrnaFC = list()
DEmrnaPvalue = list()
DEmrnaAging = list()
DEmrnaFC.develop = list()
DEmrnaPvalue.develop = list()
for(i in 1:length(alltissues)){
    #Mfuzz
    thistissue = alltissues[i]
    thismrna = mrna.tissues[[thistissue]]
    #thismrna.header = mrna.tissues.header[[thistissue]]
    thismrna.info = mrna.tissues.info[[thistissue]]
    if (sum(thismrna.info$stage == 1) ==1){ 
        thismrna = cbind(thismrna[,thismrna.info$stage == 1],thismrna)
        thismrna.info = rbind(thismrna.info[thismrna.info$stage == 1,],thismrna.info)
    }
    if (sum(thismrna.info$stage == 4) ==1){ 
        thismrna = cbind(thismrna,thismrna[,thismrna.info$stage == 4])
        thismrna.info = rbind(thismrna.info,thismrna.info[thismrna.info$stage == 4,])
    }
    
    thisDEmrna = DEGenes.simplified(thismrna,catagory = thismrna.info$stage == 4,
                  subset = thismrna.info$stage == 4 | thismrna.info$stage == 1)
    
    thisDEmrna.develop = DEGenes.simplified(thismrna,catagory = thismrna.info$stage == 2,
                    subset = thismrna.info$stage == 2 | thismrna.info$stage == 1)
    
    idx = sort.int(thisDEmrna$Pvalue,decreasing = F,index.return = T)$ix
    DEmrnaAging[[i]] = thisDEmrna[idx,-5]
    DEmrnaAging[[i]]$log2FC.devControl = thisDEmrna.develop$log2FC
    DEmrnaAging[[i]]$Pvalue.devControl = thisDEmrna.develop$Pvalue
    
    
    DEmrnaFC[[i]] = thisDEmrna$log2FC
    names(DEmrnaFC[[i]]) = rownames(thisDEmrna)
    DEmrnaPvalue[[i]] = thisDEmrna$Pvalue
    names(DEmrnaPvalue[[i]])  = rownames(thisDEmrna)
    
    DEmrnaFC.develop[[i]] = thisDEmrna.develop$log2FC
    names(DEmrnaFC.develop[[i]]) = rownames(thisDEmrna.develop)
    DEmrnaPvalue.develop[[i]] = thisDEmrna.develop$Pvalue
    names(DEmrnaPvalue.develop[[i]])  = rownames(thisDEmrna.develop)
    
    
}
names(DEmrnaFC) = alltissues
names(DEmrnaPvalue) = alltissues
names(DEmrnaAging) = alltissues
names(DEmrnaFC.develop) = alltissues
names(DEmrnaPvalue.develop) = alltissues



DEmrnaFC_matrix = list_to_matrix(DEmrnaFC,alltissues)
DEmrnaPvalue_matrix = list_to_matrix(DEmrnaPvalue,alltissues)

DEmrnaFC_matrix.develop = list_to_matrix(DEmrnaFC.develop,alltissues)
DEmrnaPvalue_matrix.develop = list_to_matrix(DEmrnaPvalue.develop,alltissues)
```

```{r}
#data S1
openxlsx::write.xlsx(DEmrnaAging, file = "./out/20230217_aging/Figure2_DEG_GO_tissue/Data S1_DE_tissue_Aging_mRNA.xlsx")

# reducce size Data S1
tpath = "./out/20230217_aging/Figure2_DEG_GO_tissue/Data S1_DE_tissue_Aging_mRNA.xlsx"
sheetNames = openxlsx::getSheetNames(tpath)
xx = list()
for(i in 1:length(sheetNames)){
    tmp = openxlsx::read.xlsx(tpath,sheet = sheetNames[i])
    tmp[c(2,3,4,7,8)] = signif(tmp[c(2,3,4,7,8)],3)
    tmp = tmp[,-c(5,6)]
    xx[[i]] = tmp
}
names(xx) = sheetNames
openxlsx::write.xlsx(xx, file = "./out/20230217_aging/Figure2_DEG_GO_tissue/reduce_Data S1_DE_tissue_Aging_mRNA.xlsx")
```

```{r}
Aging_mrna_sigup_matrix = (DEmrnaFC_matrix > 0.58 & DEmrnaPvalue_matrix < 0.05 &
                             DEmrnaPvalue_matrix.develop > 0.05) +0
Aging_mrna_sigdown_matrix = -((DEmrnaFC_matrix < -0.58 & DEmrnaPvalue_matrix < 0.05 &
                                 DEmrnaPvalue_matrix.develop > 0.05) +0)
Aging_mrna_sigall_matrix = Aging_mrna_sigup_matrix + Aging_mrna_sigdown_matrix

Aging_mrna_updown = data.frame(stringsAsFactors = F,
                               num.up = colSums(Aging_mrna_sigup_matrix,na.rm =T)/
                                 colSums(!is.na(Aging_mrna_sigup_matrix)),
                              num.down = colSums(Aging_mrna_sigdown_matrix,na.rm =T)/
                                colSums(!is.na(Aging_mrna_sigdown_matrix)),
                              num.all = colSums(abs(Aging_mrna_sigall_matrix),na.rm =T)/
                                colSums(!is.na(Aging_mrna_sigall_matrix)),
                             tissues = colnames(Aging_mrna_sigup_matrix),
                             tissue_systems = tissue.systems)
rownames(Aging_mrna_updown) = Aging_mrna_updown$tissues
Aging_mrna_updown
mean(Aging_mrna_updown$num.all)
```

```{r Figure_S1_DEmrna, fig.height = 18, fig.width = 16, fig.align = "center"}
mrnaVoconoPlot = list()
names.DEmrnaAging = names(DEmrnaAging)
for(i in 1:length(DEmrnaAging)){
    res2 = DEmrnaAging[[i]]
    
    mrnaVoconoPlot[[i]] = plot_Volcano(res2,names.DEmrnaAging[i])
}
#this plot is large, plot to file
pdf(file = './out/20230217_aging/Figure2_DEG_GO_tissue/FigureS2B_DEmrna_each_tissueV1.pdf',
    width = 3*5+1,height =3*6)
grid.arrange(arrangeGrob(grobs = mrnaVoconoPlot,ncol = 5))
dev.off()
```

## 6.3 metabolism

```{r}
DEmetFC = list()
DEmetPvalue = list()
DEmetAging = list()
DEmetFC.develop = list()
DEmetPvalue.develop = list()
for(i in 1:length(alltissues)){
    #Mfuzz
    thistissue = alltissues[i]
    thismet = met.tissues[[thistissue]]
    thismet.header = met.tissues.header[[thistissue]]
    thismet.info = met.tissues.info[[thistissue]]
    if (sum(thismet.info$stage == 1) ==1){ 
        thismet = cbind(thismet[,thismet.info$stage == 1],thismet)
        thismet.info = rbind(thismet.info[thismet.info$stage == 1,],thismet.info)
    }
    if (sum(thismet.info$stage == 4) ==1){ 
        thismet = cbind(thismet,thismet[,thismet.info$stage == 4])
        thismet.info = rbind(thismet.info,thismet.info[thismet.info$stage == 4,])
    }
    
    thisDEmet = DEGenes.simplified(thismet,catagory = thismet.info$stage == 4,
                               subset = thismet.info$stage == 4 | thismet.info$stage == 1)
    
    thisDEmet.develop = DEGenes.simplified(thismet,catagory = thismet.info$stage == 2,
                               subset = thismet.info$stage == 2 | thismet.info$stage == 1)
    
    forsort = thisDEmet$Pvalue
    forsort[is.na(forsort)] = 1
    idx = sort.int(forsort,decreasing = F,index.return = T)$ix
    DEmetAging[[i]] = thisDEmet[idx,-5]
    DEmetAging[[i]]$log2FC.devControl = thisDEmet.develop$log2FC
    DEmetAging[[i]]$Pvalue.devControl = thisDEmet.develop$Pvalue
    
    DEmetFC[[i]] = thisDEmet$log2FC
    names(DEmetFC[[i]]) = rownames(thisDEmet)
    DEmetPvalue[[i]] = thisDEmet$Pvalue
    names(DEmetPvalue[[i]])  = rownames(thisDEmet)
    
    DEmetFC.develop[[i]] = thisDEmet.develop$log2FC
    names(DEmetFC.develop[[i]]) = rownames(thisDEmet.develop)
    DEmetPvalue.develop[[i]] = thisDEmet.develop$Pvalue
    names(DEmetPvalue.develop[[i]])  = rownames(thisDEmet.develop)
    
    
}
names(DEmetAging) = alltissues
names(DEmetFC) = alltissues
names(DEmetPvalue) = alltissues
DEmetFC_matrix = list_to_matrix(DEmetFC,alltissues)
DEmetPvalue_matrix = list_to_matrix(DEmetPvalue,alltissues)

DEmetFC_matrix.develop = list_to_matrix(DEmetFC.develop,alltissues)
DEmetPvalue_matrix.develop = list_to_matrix(DEmetPvalue.develop,alltissues)
```

```{r}
#write Data S3
openxlsx::write.xlsx(DEmetAging, file = "./out/20230217_aging/Figure2_DEG_GO_tissue/Data S3_DE_tissue_Aging_metabolite.xlsx")

# reducce size Data S2
tpath = "./out/20230217_aging/Figure2_DEG_GO_tissue/Data S3_DE_tissue_Aging_metabolite.xlsx"
sheetNames = openxlsx::getSheetNames(tpath)
xx = list()
for(i in 1:length(sheetNames)){
    tmp = openxlsx::read.xlsx(tpath,sheet = sheetNames[i])
    tmp[c(2,3,4,7,8)] = signif(tmp[c(2,3,4,7,8)],3)
    tmp = tmp[,-c(5,6)]
    xx[[i]] = tmp
}
names(xx) = sheetNames
openxlsx::write.xlsx(xx, file = "./out/20230217_aging/Figure2_DEG_GO_tissue/reduce_Data S3_DE_tissue_Aging_metabolite.xlsx")
```

```{r}
# num DEmets
Aging_met_sigup_matrix = (DEmetFC_matrix > 0.58 & DEmetPvalue_matrix < 0.05 &
                            DEmetPvalue_matrix.develop > 0.05) + 0
Aging_met_sigdown_matrix = -((DEmetFC_matrix < -0.58 & DEmetPvalue_matrix < 0.05  &
                                DEmetPvalue_matrix.develop > 0.05) + 0)
Aging_met_sigall_matrix = Aging_met_sigup_matrix + Aging_met_sigdown_matrix

Aging_met_updown = data.frame(stringsAsFactors = F,
                              num.up = colSums(Aging_met_sigup_matrix,na.rm =T)/
                                colSums(!is.na(Aging_met_sigup_matrix)),
                              num.down = colSums(Aging_met_sigdown_matrix,na.rm =T)/
                                colSums(!is.na(Aging_met_sigdown_matrix)),
                              num.all = colSums(abs(Aging_met_sigall_matrix),na.rm =T)/
                                colSums(!is.na(Aging_met_sigall_matrix)),
                             tissues = colnames(Aging_met_sigup_matrix),
                             tissue_systems = tissue.systems)
rownames(Aging_met_updown) = Aging_met_updown$tissues
Aging_met_updown
mean(Aging_met_updown$num.all)
```

```{r Figure_S3_DEmet, fig.height = 18, fig.width = 16, fig.align = "center"}
metVoconoPlot = list()
names.DEmetAging = names(DEmetAging)
for(i in 1:length(DEmetAging)){
    res2 = DEmetAging[[i]]
    
    metVoconoPlot[[i]] = plot_Volcano(res2,names.DEmetAging[i])
}

pdf(file = './out/20230217_aging/Figure2_DEG_GO_tissue/
    FigureS2C_DEmet_each_tissueV1.pdf',width = 3*5+1,height =3*6)
grid.arrange(arrangeGrob(grobs = metVoconoPlot,ncol = 5))
dev.off()
```

## 6.4 Figure 2A perc changed mols each tissue

```{r}
fmt_dcimals <- function(decimals=0){
    function(x) format(x,nsmall = decimals,scientific = FALSE)
}
```

```{r Figure_2A, fig.height = 5, fig.width = 12, fig.align = "center"}
plotAgingNum = list()
idx = sort.int(Aging_pro_updown$num.all,decreasing = F,index.return = T)$ix
Aging_pro_updown.v = Aging_pro_updown[idx,]
idx = sort.int(Aging_pro_updown.v$tissue_systems,decreasing = F,index.return = T)$ix
Aging_pro_updown.v = Aging_pro_updown.v[idx,]
tissueindex = rownames(Aging_pro_updown.v)

#metabolites
Aging_met_updown.v = Aging_met_updown[tissueindex,]

p1 = ggplot(Aging_met_updown.v,aes(x = factor(tissues,level = tissues),y = num.up,
                       color = tissue_systems,fill = tissue_systems))+ 
  geom_bar(stat="identity",alpha = 0.8)

plotAgingNum[[1]] = p1+ geom_bar(stat="identity",aes(y = num.down),alpha = 0.6)+
  geom_hline(yintercept = 0)+
theme_classic()+lghplot.addtheme(legend.position = 'none',hjust = 1,size = 13.5)+
theme(axis.text.y = element_text(size = 9.5, face = "bold", color = "black"))+
  scale_color_aaas(alpha = 0.6)+
scale_fill_aaas(alpha = 0.6)+theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),axis.line.x =element_blank())+
scale_y_continuous(labels = fmt_dcimals(2))+
ylab('Metabolites')+ xlab('')

#protein
Aging_pro_updown.v = Aging_pro_updown[tissueindex,]

p1 = ggplot(Aging_pro_updown.v,aes(x = factor(tissues,level = tissues),y = num.up,
                       color = tissue_systems,fill = tissue_systems))+ 
  geom_bar(stat="identity",alpha = 0.8)

plotAgingNum[[2]] = p1+ geom_bar(stat="identity",aes(y = num.down),alpha = 0.6)+
  geom_hline(yintercept = 0)+
theme_classic()+lghplot.addtheme(legend.position = 'none',hjust = 1,size = 13.5)+
theme(axis.text.y = element_text(size = 9.5, face = "bold", color = "black"))+
  scale_color_aaas(alpha = 0.6)+
scale_fill_aaas(alpha = 0.6)+  
scale_y_continuous(labels = fmt_dcimals(2))+
ylab('Proteins')+ xlab('')

#mrna
Aging_mrna_updown.v = Aging_mrna_updown[tissueindex,]

p1 = ggplot(Aging_mrna_updown.v,aes(x = factor(tissues,level = tissues),y = num.up,
                       color = tissue_systems,fill = tissue_systems))+ 
  geom_bar(stat="identity",alpha = 0.8)

plotAgingNum[[3]] = p1+ geom_bar(stat="identity",aes(y = num.down),alpha = 0.6)+
  geom_hline(yintercept = 0)+
theme_classic()+lghplot.addtheme(legend.position = 'none',hjust = 1,size = 13.5)+
theme(axis.text.y = element_text(size = 9.5, face = "bold", color = "black"))+
  scale_color_aaas(alpha = 0.6)+
scale_fill_aaas(alpha = 0.6)+scale_y_continuous(labels = fmt_dcimals(2))+
ylab('mRNAs')+ xlab('')

#pdf(file = './out/20230217_aging/Figure2_DEG_GO_tissue/
#   Figure2A_number_of_Aging_moleculars_prometv1_aaas.pdf',width = 12,height =5)
grid.arrange(arrangeGrob(grobs = plotAgingNum[1:2],
                         top = 'Overall molecular changes with aging',
                         ncol = 1,heights = c(1.7,3.3)))
#dev.off()
```

## 6.5 Figure 2B GO enrichment

```{r}
#write for metascape up
inputGOenrich = data.frame(tissues = names(DEproAging),stringsAsFactors = F,
                      genes = rep('c',length(DEproAging)))
#colnames(inputGOenrich) = c('tissue','GeneSymbol')
for(i in 1:length(DEproAging)){
    tmp = DEproAging[[i]]
    tgenes = unique(rownames(tmp)[tmp$Pvalue < 0.05 & tmp$log2FC > 0.58 &
                                    tmp$Pvalue.devControl > 0.05])
    tgenes = tgenes[!is.na(tgenes)]
    tgenes = tgenes[tgenes != '']
    tgenes = paste0(tgenes,collapse =',')
    inputGOenrich$genes[i] = tgenes
}
writetxt(inputGOenrich,'./out/20230217_aging/Figure2_DEG_GO_tissue/protein/_inputGOenrich_age_DEpro_up.txt',row.names = F,col.names = F)


#write for metascape down
inputGOenrich = data.frame(tissues = names(DEproAging),stringsAsFactors = F,
                      genes = rep('c',length(DEproAging)))
#colnames(inputGOenrich) = c('tissue','GeneSymbol')
for(i in 1:length(DEproAging)){
    tmp = DEproAging[[i]]
    tgenes = unique(rownames(tmp)[tmp$Pvalue < 0.05 & tmp$log2FC < -0.58 &
                                    tmp$Pvalue.devControl > 0.05])
    tgenes = tgenes[!is.na(tgenes)]
    tgenes = tgenes[tgenes != '']
    tgenes = paste0(tgenes,collapse =',')
    inputGOenrich$genes[i] = tgenes
}
writetxt(inputGOenrich,'./out/20230217_aging/Figure2_DEG_GO_tissue/protein/_inputGOenrich_age_DEpro_down.txt',row.names = F,col.names = F)
```

```{r Figure_2B, fig.height = 10, fig.width = 10, fig.align = "center"}
outids = c()
tpath1 = './out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_up/Enrichment_heatmap/HeatmapSelectedGO.csv'
tpath2 = './out/20230217_aging/Figure2_DEG_GO_tissue/protein//metascape_DEpro_down/Enrichment_heatmap/HeatmapSelectedGO.csv'
selectGO = readxl::read_xlsx(path = './out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_up/metascape_result.xlsx',
                             sheet = 'Enrichment')
go1 = file2frame(tpath1,sep = ',',header = T,row.names =2)
rownames(go1) = paste0(go1$GO,'-',rownames(go1))
go1term = go1$GO
tpath1a = './out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_up/Enrichment_GO/GO_AllLists.csv'
go1qval = file2frame(tpath1a,sep = ',',header = T)
rownames(go1qval) = paste0(go1qval$GO,'X_LogP_',go1qval$GeneList)
go1 = abs(as.matrix(go1[,-1]))
cname = colnames(go1)
for(i in 1:nrow(go1)){
    for(j in 1:ncol(go1)){
        tmpname = paste0(go1term[i],cname[j])
        go1[i,j] = abs(go1qval[tmpname,]$Log.q.value.)
    }
}
go1[is.na(go1)] = 0

go1for_write = cbind(data.frame(GO = rownames(go1)),-go1)
outlist = list()
outlist[[1]] = go1for_write;
outlist[[2]] = selectGO;
tmpnames = sort(unique(go1qval$GeneList))
for(i in 1:length(tmpnames)){
    outlist[[i+2]] = go1qval[go1qval$GeneList == tmpnames[[i]],]
}
names(outlist) = c('GOenrichmentALL','selectGO',tmpnames)
openxlsx::write.xlsx(outlist, file = "./out/20230217_aging/Figure2_DEG_GO_tissue/Data S4_GO_tissue_upregulated_protein.xlsx")


selectGO = readxl::read_xlsx(path = './out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_down/metascape_result.xlsx',sheet = 'Enrichment')
go2 = file2frame(tpath2,sep = ',',header = T,row.names =2)
rownames(go2) = paste0(go2$GO,'-',rownames(go2))
go2term = go2$GO
tpath2a = './out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_down/Enrichment_GO/GO_AllLists.csv'
go2qval = file2frame(tpath2a,sep = ',',header = T)
rownames(go2qval) = paste0(go2qval$GO,'X_LogP_',go2qval$GeneList)
go2 = abs(as.matrix(go2[,-1]))
cname = colnames(go2)
for(i in 1:nrow(go2)){
    for(j in 1:ncol(go2)){
        tmpname = paste0(go2term[i],cname[j])
        go2[i,j] = -abs(go2qval[tmpname,]$Log.q.value.)
    }
}
go2[is.na(go2)] = 0

go1for_write = cbind(data.frame(GO = rownames(go2)),go2)
outlist = list()
outlist[[1]] = go1for_write;
outlist[[2]] = selectGO;
tmpnames = sort(unique(go2qval$GeneList))
for(i in 1:length(tmpnames)){
    outlist[[i+2]] = go2qval[go2qval$GeneList == tmpnames[[i]],]
}
names(outlist) = c('GOenrichmentALL','selectGO',tmpnames)
openxlsx::write.xlsx(outlist, file = "./out/20230217_aging/Figure2_DEG_GO_tissue/Data S5_GO_tissue_downregulated_protein.xlsx")


thisgo = rbind(go1,go2)
thisgo.matrix = as.matrix(thisgo)
colnames(thisgo.matrix) = capitalize(gsub('X_LogP_','',colnames(thisgo.matrix)))
thisgo.matrix[abs(thisgo.matrix) < -log10(0.05)] = 0
writetxt(thisgo.matrix,'./out/20230217_aging/Figure2_DEG_GO_tissue/Figure2_heatmap_GOenrichment_metascape_DEpro_eachtissue_fdr.txt',row.names = T)
thisgo.matrix[thisgo.matrix > 4] = 4
thisgo.matrix[thisgo.matrix < -4] = -4

rownames(thisgo.matrix)[rownames(thisgo.matrix) == 'R-HSA-9716542-Signaling by Rho GTPases, Miro GTPases and RHOBTB3'] = 'R-HSA-9716542-Signaling by Rho GTPases'
rownames(thisgo.matrix)[rownames(thisgo.matrix) == 'R-HSA-1428517-The citric acid (TCA) cycle and respiratory electron transport'] = 'R-HSA-1428517-TCA cycle and respiratory electron transport'

graphics.off()
pheatmap::pheatmap(thisgo.matrix,cluster_rows = T,cluster_cols = T,
                   main ='Enrichment of proteome',
                   fontsize_row = 11,fontsize_col = 11,fontsize = 14,
                   treeheight_row = 20,treeheight_col = 20,legend = T,
                  color=colorRampPalette(c('#3B4992','gray99','#BB0021FF'))(50),
                  height = 10,width = 10
                  #file ="./out/20230217_aging/Figure2_DEG_GO_tissue/Figure2_heatmap_GOenrichment_metascape_DEpro_eachtissue.v1_aaas.pdf"
                   )
```

# 7 Figure 3 clustering aging type

## 7.1 mfuzz promet tissues

```{r}
tissues = names(pro.tissues)
promet.tissues = list()
promet.tissues.info = list()
promet.tissues.Z = list()
mfuzz.promet.tissues = list()
promet.mstd.eset = list()
for(i in 1:length(tissues)){
#for(i in 1:1){
    tt = tissues[i]
    thispro = pro.tissues[[tt]]
    thispro = delete_dup_genes_forprotein(thispro,pro.tissues.header[[tt]])
    thispro.header = pro.tissues.header[[tt]]
    thispro = thispro[rowSums(is.na(thispro)) < 1/3*ncol(thispro), ]
    thismet = met.tissues[[tt]]
    rownames(thismet) = paste0('met_',rownames(thismet))
    thispromet = rbind2(thispro,thismet)
    thisinfo = pro.tissues.info[[tt]]
    thisinfo = thisinfo[colnames(thispromet),]
    thispromet.median = t(aggregate(t(thispromet), by=list(thisinfo$stage), 
                                    FUN=median, na.rm = T))
    mstd = standardise_matrix(thispromet.median)
    promet.tissues.Z[[i]] = mstd
    promet.tissues[[i]] = thispromet
    promet.tissues.info[[i]] = thisinfo
    mstd.v = mstd[rowSums(is.na(mstd))  == 0,]
    mstd.eset = new("ExpressionSet",exprs = mstd.v)
    promet.mstd.eset[[i]] = mstd.eset
    mfuzz.promet.tissues[[i]] = mfuzz(mstd.eset, c = 8,m = 1.5)

}
names(mfuzz.promet.tissues) = tissues
names(promet.tissues) = tissues
names(promet.tissues.Z) = tissues
names(promet.tissues.info) = tissues
names(promet.mstd.eset) = tissues
```

```{r_}
## not run

tissues = names(promet.tissues.mstd.eset)
for(i in 1:length(mfuzz.promet.tissues)){
    tpath = paste0('./out/20210428_aging/promet/tissues/',tissues[i])
    dir.create(tpath)
    pdf(paste0(tpath,'/mfuzz_plot_8A_promet.pdf'),width = 8,height = 4)
    mfuzz.plot(promet.tissues.mstd.eset[[i]],mfuzz.promet.tissues[[i]],
    mfrow=c(2,4),time.labels = c("Juvenile", "Young_adult","Middle_aged",
                                   "Elderly"),new.window = F)
    dev.off()
}

```

```{r}
promet.tissues.mstd.eset = promet.mstd.eset
promet.tissues.Z.t = list()
for(i in 1:length(promet.tissues.Z)){
    tmp = as.data.frame(t(promet.tissues.Z[[i]]))
    promet.tissues.Z.t[[i]] = tmp
}
names(promet.tissues.Z.t) = names(promet.tissues.Z)

promet.whole.Z = t(as.matrix(as.data.frame(rbindlist(promet.tissues.Z.t,fill = T))))
colnames(promet.whole.Z) = paste0(rep(names(promet.tissues.Z),each = 4),'_',rep(1:4,times = 30))
promet.whole.Z.info = data.frame(tissue = rep(names(promet.tissues.Z),each = 4),
                                 stringsAsFactors = F,
                                stage = rep(1:4,times = 30))
rownames(promet.whole.Z.info) = colnames(promet.whole.Z)
```

## 7.2 Figure 3A whole body mfuzz

```{r}
#this is used to reproduce the figures in manuscript
load('./out/promet_outdata.Rdata')
```

```{r}
promet.whole.Z.v = promet.whole.Z[rowSums(is.na(promet.whole.Z)) < 0.5*120,]
promet.whole.Z.mean = t(aggregate(t(promet.whole.Z.v), 
                      by=list(promet.whole.Z.info$stage), FUN=mean, na.rm = T))
promet.whole.Z.mean  =promet.whole.Z.mean[-1,]
promet.whole.Z.mean.v = promet.whole.Z.mean[rowSums(is.na(promet.whole.Z.mean))  == 0,]
promet.whole.Z.mean.eset = new("ExpressionSet",exprs = promet.whole.Z.mean.v)
dim(promet.whole.Z.mean.v)
sum(substr(rownames(promet.whole.Z.mean.v),1,4) == 'met_')
#mfuzz.promet.whole = mfuzz(promet.norm.eset.stand, c = 8,m = 1.5)

```

```{r Figure_3A, fig.height = 4, fig.width = 8, fig.align = "center"}
mfuzz.plot(promet.whole.Z.mean.eset,mfuzz.promet.whole,mfrow=c(2,4),
           time.labels = c("Juvenile", "Young_adult","Middle_aged",
                                   "Elderly"),new.window = F)
```

## 7.3 Figure 3B Go enrichment

```{r Figure_3B, fig.height = 4.5, fig.width = 6.5, fig.align = "center"}
outids = c()
tpath = paste0('./out/20230217_aging/Figure3_trajactory_analysis/cluster1-8_metascape/metascape_result_curated.xlsx')
    my_data <- read_excel(tpath, sheet = "Enrichment")
    idx = regexpr('Summary',my_data$GroupID) >0
    outids = c(outids,my_data$Term[idx])
outids

tpath = paste0('./out/20230217_aging/Figure3_trajactory_analysis/cluster1-8_metascape/Enrichment_GO/','GO_membership.csv')
thisgo = file2frame(tpath,sep = ',')
thisgo= thisgo[!duplicated(thisgo$GO),]
xid = is.element(thisgo$GO,outids)
thisgo = thisgo[xid,]

#rownames(thisgo) = paste0(thisgo$GO,':',thisgo$Description)
rownames(thisgo) = capitalize(paste0(thisgo$Description))
thisgo.matrix = as.matrix(thisgo[,substr(colnames(thisgo),1,6)== 'X_LogP'])
writetxt(thisgo.matrix,'./out/20230217_aging/Figure3_trajactory_analysis/Table S1_GOenrichment_metascape_promet_clusters.txt',row.names = T)

thisgo.matrix[thisgo.matrix > -3] = 0
colnames(thisgo.matrix) = capitalize(gsub('X_LogP_','',colnames(thisgo.matrix)))
colnames(thisgo.matrix) = gsub('Cluster','C',colnames(thisgo.matrix))

pheatmap::pheatmap(thisgo.matrix,cluster_rows = T,cluster_cols = T,
                   main = 'Enrichment analysis of each cluster',
                   fontsize_row = 9,fontsize_col = 10,fontsize = 10,
                   treeheight_row = 20,treeheight_col = 20,legend = T,
                  color=colorRampPalette(c('#3C5488FF','gray95','gray95'))(7),
                  breaks = c(-32,-24,-16,-8,0),
                  #file ="./out/20230217_aging/Figure3_trajactory_analysis/Figure3B_GOenrichment_metascape_prometA.pdf",
                 height = 3.5,width = 5.5
                  )
```

## 7.4 mfuzz for each tissue

### 7.4.1 get data

```{r}
# construct data
tissue_trajectory = data.frame()
tissue_names = names(promet.tissues)
for(k in 1:8){
    tmpGeneList = names(mfuzz.promet.whole$cluster)[mfuzz.promet.whole$cluster == k]
    for(i in 1:length(promet.tissues)){
        
        M1 = promet.tissues[[i]]
        metadata.tissue = promet.tissues.info[[i]]
        idgene1 = intersect(tmpGeneList,rownames(M1))
        cc = repmat(as.matrix(rowMedians(M1[,metadata.tissue$stage == '1'],na.rm = T)),1,ncol(M1))
        tsd = repmat(as.matrix(apply(M1,1,sd,na.rm = T)),1,ncol(M1))
        
        M1.Z = (M1 - cc)/tsd
        M1.Z.v = M1.Z[idgene1,]
        M1.Z.v.mean = t(aggregate(t(M1.Z.v), 
                                  by=list(metadata.tissue$stage), FUN=median, na.rm = T))
        M1.Z.v.mean = M1.Z.v.mean[-1,]
        mean_x = colMeans(M1.Z.v.mean,na.rm  =T)
        mean_x = mean_x - mean_x[1]
        
        tmpdata2 = data.frame(expr = mean_x,stringsAsFactors = F,
                              stage = c(1,2,3,4),
                              group = factor(c('Juvenile','Young','Middle_aged','Elderly'),
                                     level = c('Juvenile','Young','Middle_aged','Elderly')),
                              cluster = rep(k,4),
                              tissue = rep(tissue_names[i],4)
                     )
        if (nrow(tissue_trajectory) <1){
            tissue_trajectory = tmpdata2
        }else{
            tissue_trajectory = rbind(tissue_trajectory,tmpdata2)
        }
    }

}
```

```{r}
tissue_trajectory$tissue_system = tissue.systems[tissue_trajectory$tissue]
tissue_trajectory$tissue_system_color = tissue.color[tissue_trajectory$tissue_system]
# to matrix
tissue_trajectory_matrix = matrix(0,nrow(tissue_trajectory)/length(tissue_names),length(tissue_names))
for(i in 1:length(tissue_names)){
    tmptr = tissue_trajectory[tissue_trajectory$tissue == tissue_names[i],]
    tissue_trajectory_matrix[,i] = tmptr$expr
}
rownames(tissue_trajectory_matrix) = paste(tmptr$group,tmptr$cluster,sep = '_C')
colnames(tissue_trajectory_matrix) = tissue_names
```

```{r Figure_3C, fig.height = 9, fig.width = 8, fig.align = "center"}
tissue_tr_plot = list()
for (i in 1:8){
    idx = tissue_trajectory$cluster == i
    tissue_tr_plot [[i]] = ggplot(tissue_trajectory[idx,],
                                  aes(x= group,y = expr,group  =tissue)) +
        geom_line(aes(color = tissue_system),alpha = 0.3 ,
                  position = position_dodge(0.2),size = 2)+ scale_color_aaas()+ 
        lghplot.addtheme(hjust = 1,size = 14)+ ggtitle(paste0('Cluster ',i))+
         theme(axis.line = element_line(size = 1.2))+ xlab('')+ ylab('')+
        scale_y_continuous(labels = scales::comma_format(accuracy =0.1))
}
#pdf(file = "./out/20230217_aging/Figure3_trajactory_analysis/
#    Figure3c_trajectory_for_each_tissue_aaas.pdf",height = 9,width = 8)
grid.arrange(arrangeGrob(grobs = tissue_tr_plot,ncol = 3,heights = c(4,4,4),
            top = textGrob('Average molecular aging trajectory in 30 tissues',
                           gp=gpar(fontface="bold",  fontsize=20)),
            bottom=textGrob('Tissues', gp=gpar(fontface="bold",  fontsize=18)),
            left = textGrob('Z values', gp=gpar(fontface="bold",  fontsize=18),rot=90)))
#dev.off()
```

### 7.4.3 Figure 3D heatmap clustering

```{r Figure_3D, fig.height = 5, fig.width = 6, fig.align = "center"}
#heatmap cluster '#E64B35FF''#4DBBD5FF'
idx = substr(rownames(tissue_trajectory_matrix),1,3) != 'Juv'
dent = pheatmap::pheatmap(tissue_trajectory_matrix[idx,],scale = 'row',
                    main = 'Heatmap of molecular aging trajectory in 30 tissues',
                          height = 5,width = 6,angle_col = 45,             
                          fontsize_row = 8,fontsize_col = 8,
                          treeheight_row = 20,treeheight_col = 20,
                          #color=colorRampPalette(c('#3B4992','gray95','red'))(30),
                          #file ="./out/20230217_aging/Figure3_trajactory_analysis/Figure3D_heatmap_based_trajectoryV1.pdf",
                          color=colorRampPalette(c('#3C5488FF','gray95','#E64B35FF'))(30))
```

### 7.4.4 Figure 3E dendgrogam

```{r Figure_3E, fig.height = 5, fig.width = 7, fig.align = "center"}
dendcol = as.dendrogram(dent$tree_col)

labelColors = c('#3C5488FF','#E64B35FF')

clusMember = cutree(dent$tree_col,2)

# function to get color labels
colLab <- function(n) {
  if (is.leaf(n)) {
    a <- attributes(n)
    labCol <- labelColors[clusMember[which(names(clusMember) == a$label)]]
    attr(n, "nodePar") <- c(a$nodePar, lab.col = labCol)
  }
  n
}
clusDendro = dendrapply(dendcol, colLab)

dendcolsplit = cut(dendcol,12)
class1 = unlist(cut(dendcol,12)$lower[[1]])
class2 = unlist(cut(dendcol,12)$lower[[2]])
tissueType = data.frame(tissue = c(tissue_names[class1],tissue_names[class2]),
                        stringsAsFactors = F,
              type = c(rep('Type II',length(class1)),rep('Type I',length(class2))))
rownames(tissueType) = tissueType$tissue
tissueType$class = tissueType$type
tissueClass  = tissueType

#pdf('./out/20230217_aging/Figure3_trajactory_analysis/Figure3E_dendrogamV1.pdf',width = 7,height = 5)
plot(clusDendro, main = "Molecular trajactory dendgrogram")
#dev.off()
#dev.off()




```

### 7.4.5 Figure 3F trajactory pca

```{r Figure_3F, fig.height = 7, fig.width = 7, fig.align = "center"}
# pca
tissue_pca = prcomp(t(tissue_trajectory_matrix),cor=F)
perc_tissue_pca = 100*summary(tissue_pca)$importance

#pdf('./out/20230217_aging/Figure3_trajactory_analysis/Figure3F_PCA.pdf',width = 7,height = 7)
TissueType = tissueType[tissue_names,]$type
ggplot(,aes(tissue_pca$x[,1],tissue_pca$x[,2],color = TissueType)) + 
   geom_point(size =5,alpha = 0.6) + 
   theme_classic() +lghplot.addtheme(legend.position = 'top')+  
   #stat_ellipse(lwd=1,level = 0.95) +
   geom_text_repel(aes(label = tissue_names),size = 5,box.padding = 0.5,face = 'bold')+
   xlab(paste("PC1(",as.character(round(perc_tissue_pca[2,1],1)),'%)',sep = '')) + 
   ylab(paste("PC2(",as.character(round(perc_tissue_pca[2,2],1)),'%)',sep = '')) + 
   scale_color_manual(values = c('#E64B35FF','#3C5488FF'))+
   theme(legend.text = element_text(size=12,face = 'bold'))+
   labs(title = "Molecular trajectory PCA")
#dev.off()
```

# 8 Figure 4 cluster by MAA

## 8.1 Figure 4A

```{r Figure_S4, fig.height = 7, fig.width = 7, fig.align = "center"}
promet.norm.eset.stand.matrix = as.matrix(promet.whole.Z.mean.eset)
variability = rep(0,8) 
for(j in 1:8){
    tgene = names(mfuzz.promet.whole$cluster)[mfuzz.promet.whole$cluster == j]
    tgene = intersect(tgene,rownames(promet.norm.eset.stand.matrix))
    bx = t(t(promet.norm.eset.stand.matrix[tgene,]) - mfuzz.promet.whole$centers[j,])
    variability[j] = mean(sqrt(rowSums(bx^2)/(ncol(bx)-1)))
}
tmpdata = data.frame(amplitude = as.vector(mfuzz.promet.whole$centers[,4]-
                                             mfuzz.promet.whole$centers[,1]),
                     stringsAsFactors = F,
                     variability = variability,
                     class = paste0('Cluster ',1:8))
#pdf(file ='./out/20230217_aging/Figure4_MAA_analysis/
#    FigureS4_cluster_amplitude_variability_8_promet.pdf',width = 7,height = 5)
ggplot(tmpdata,aes(variability,amplitude)) + geom_point(size = 4) + lghplot.addtheme(size = 18)+
  geom_text_repel(aes(label = class),size = 7)+theme(axis.line = element_line(size = 1.0))+
  ggtitle('Figure S4 cluster_amplitude_variability')
#dev.off()
```

```{r Figure_4A, fig.height = 9, fig.width = 16, fig.align = "center"}
tissues = names(promet.tissues.Z)
clusterdist = matrix(0,length(tissues),8)
clusteramplitude = matrix(0,length(tissues),8)
clusteramplitude_xx = matrix(0,length(tissues),8)
for(i in 1:length(tissues)){
    mstd =   promet.tissues.Z[[i]] 
    for(j in 1:8){
    tgene = names(mfuzz.promet.whole$cluster)[mfuzz.promet.whole$cluster == j]
    tgene = intersect(tgene,rownames(mstd))
    bx = t(t(mstd[tgene,]) - mfuzz.promet.whole$centers[j,])
    clusterdist[i,j] = mean(sqrt(rowSums(bx^2)/(ncol(bx)-1)),na.rm = T)
    if(length(tgene) < 2){
      clusteramplitude[i,j] = NA
      next;
    }
    tamp = colMeans(mstd[tgene,],na.rm = T)
    clusteramplitude[i,j] = abs(tamp[4]-tamp[1])
    clusteramplitude_xx[i,j] = tamp[4]-tamp[1]
  }
}
rownames(clusteramplitude) = tissues
rownames(clusteramplitude_xx) = tissues
rownames(clusterdist) = tissues

px = list()
for(j in 1:8){
  tmpdata = data.frame(amplitude = clusteramplitude_xx[,j],
                       stringsAsFactors = F,
                       variability = clusterdist[,j],
                       class = tissues,
                      tissue.systems = tissue.systems)
    px[[j]] = ggplot(tmpdata,aes(variability,amplitude,color = tissue.systems)) + 
      geom_point(size = 6) + lghplot.addtheme(size = 14)+
    scale_color_aaas()+scale_fill_aaas()+
    geom_text_repel(aes(label = class),size = 3,box.padding = 0.5)+
    theme(axis.line = element_line(size = 1.5))+ ggtitle(paste0('Cluster ',j))+ 
      xlab('') + ylab('')
}

#pdf(file = "./out/20230217_aging/Figure4_MAA_analysis/
#    FigureXS_cluster8_tissues_amplitude_variability_allV2.pdf",height = 9,width = 16)
grid.arrange(arrangeGrob(grobs = px,ncol = 4,margin = c(1,1,1,1),
                top = textGrob('Molecular alteration amplitude of each cluster in 30 tissues', 
                                         gp=gpar(fontface="bold",  fontsize=24)),
                         bottom=textGrob('Molecular alteration variability', 
                                         gp=gpar(fontface="bold",  fontsize=24)),
                         #bottom = 'mRNA expression(log2 FPKM)',
                        left = textGrob('Molecular alteration amplitude(MAA)', 
                                        gp=gpar(fontface="bold",  fontsize=24),rot=90)))
#dev.off()
```

## 8.2 Figure 4B

```{r Figure_4B, fig.height = 4, fig.width = 3.5, fig.align = "center"}
tmpaa = clusteramplitude_xx
colnames(tmpaa) = paste0("C",1:8)
dent.MAA = pheatmap::pheatmap(tmpaa,scale = 'none',height = 4,width = 3.5,
                          main  = 'Aging types in solid tissues using MAA',
                          fontsize_row = 7,fontsize_col = 7,
                          treeheight_row = 20,treeheight_col = 20,
                          color=colorRampPalette(c('#3B4992','gray95','red'))(30),
             #filename = './out/20230217_aging/Figure4_MAA_analysis             /FigureX_promet_heatmap_tissue_aging_byMAA.pdf'
             )


```

## 8.3 Figure 4C

```{r Figure_4add_1, fig.height = 5, fig.width = 7, fig.align = "center"}
dendrow = as.dendrogram(dent.MAA$tree_row)

labelColors = c('#3C5488FF','#E64B35FF')

clusMember = cutree(dent.MAA$tree_row,2)

# function to get color labels
colLab <- function(n) {
  if (is.leaf(n)) {
    a <- attributes(n)
    labCol <- labelColors[clusMember[which(names(clusMember) == a$label)]]
    attr(n, "nodePar") <- c(a$nodePar, lab.col = labCol)
  }
  n
}
clusDendro = dendrapply(dendrow, colLab)

#pdf('./out/20230217_aging/Figure4_MAA_analysis/
#    FigureX_dendrogam_by_MAA.pdf',width = 7,height = 5)
plot(clusDendro, main = "Molecular trajactory dendgrogram by MAA")
#dev.off()
```

```{r}
mean_clusteramplitude_xx = rowMeans(clusteramplitude_xx)
names(mean_clusteramplitude_xx) = rownames(clusteramplitude_xx)
tissues_pro = rownames(clusteramplitude_xx)
dendrow = as.dendrogram(dent.MAA$tree_row)
dendrowsplit = cut(dendrow,2.5)
class1 = unlist(cut(dendrow,2.5)$lower[[1]])
class2 = unlist(cut(dendrow,2.5)$lower[[2]])
tissueClass.MAA = data.frame(tissue = c(tissues_pro[class1],
                                        tissues_pro[class2]),stringsAsFactors = F,
                         class = c(rep('Type II',length(class1)),
                                   rep('Type I',length(class2))))
rownames(tissueClass.MAA)  = tissueClass.MAA$tissue
tissueClass.MAA$mean_aging_amplitude=mean_clusteramplitude_xx[rownames(tissueClass.MAA)]


tissueClass$class.MAA = tissueClass.MAA[rownames(tissueClass),]$class
tissueClass$mean_aging_amplitude = tissueClass.MAA[rownames(tissueClass),]$mean_aging_amplitude

tissueClass$type = rep('Undefine',nrow(tissueClass))
tissueClass$type[tissueClass$class == 'Type I' & tissueClass$class.MAA == 'Type I'] = 'Type I'
tissueClass$type[tissueClass$class == 'Type II' & tissueClass$class.MAA == 'Type II'] = 'Type II'
tissueClass

```

```{r_ Figure_4C, fig.height = 9, fig.width = 9, fig.align = "center"}
require(scatterplot3d)
#typecolor = '#3B4992','gray99','#EE0000FF',
#pdf('./out/20230217_aging/Figure4_MAA_analysis/Figure4C_3d_scatterplotAA.pdf')
typecolor = tissueClass[rownames(clusteramplitude_xx),]$color


s3d <- scatterplot3d(clusteramplitude_xx[,8], clusteramplitude_xx[,2], clusteramplitude_xx[,4], grid = T,size = 1,  cex.lab = 2, cex.symbols = 4,  color=typecolor,
pch=19, 
scale.y=.75,

main="Aging amplitude",

xlab='C8',

ylab='C2 ',

zlab='C4')
s3d1 <- scatterplot3d(clusteramplitude_xx[,8], clusteramplitude_xx[,2], clusteramplitude_xx[,4], grid = T,size = 1,  cex.lab = 2, cex.symbols = 4,  # x y and z axis
              color=typecolor,

pch=19,       

scale.y=0.75,
scale.Z=1.25,

main="Aging amplitude",

xlab='C8',

ylab='C2 ',

zlab='C4',
                     

text(s3d$xyz.convert(clusteramplitude_xx[,c(8,2,4)]), labels = rownames(clusteramplitude_xx),
     cex= 1))
#dev.off()
```

## 8.4 Figure 4D

```{r}
require(VennDiagram)
tmpnames = rownames(tissueClass)
Trajactory_Type_I = tmpnames[tissueClass$class == 'Type I']
Trajactory_Type_II = tmpnames[tissueClass$class == 'Type II']
MAA_Type_I = tmpnames[tissueClass$class.MAA == 'Type I']
MAA_Type_II = tmpnames[tissueClass$class.MAA == 'Type II']
venn.diagram(list(Trajactory_Type_I=Trajactory_Type_I,Trajactory_Type_II = Trajactory_Type_II,
                  MAA_Type_I = MAA_Type_I,MAA_Type_II=MAA_Type_II), 
             #fill=c("red","blue",),
             fill = c("cornflowerblue", "green", "yellow", "darkorchid1"),
             alpha=c(0.5,0.5,0.5,0.5), cex=2, cat.fontface=3, cat.cex = 1.3,margin = 0.1,
             filename="./out/20230217_aging/Figure4_MAA_analysis/Venn_trajactory_and_MAA.tiff")
```

# 9 Figure 5 experimental validation

```{r Figure_S5A, fig.height = 4, fig.width = 7, fig.align = "center"}
# Figure 5A all changed protein
tx = colSums(abs(Aging_pro_sigall_matrix),na.rm = T)
yy1 = tissueClass[names(tx),]
#pdf('./out/20230217_aging/Figure5_markers_HE/Figure_S5_allchanged_proteins.pdf')
ggplot(,aes(x = yy1$type,y = tx,color = yy1$type,fill = yy1$type)) +#geom_violin()+
     geom_boxplot(outlier.size = -1,alpha = 0.1,size = 0.8)+
     geom_jitter(size = 5,width = 0.3)+ theme_classic() + 
     ggtitle('# of changed proteins')+
     lghplot.addtheme()+ scale_color_manual(values=c('#EE0000','#3B4992','gray10'))+
     xlab('Tissue Type')+ylab('Number of changed proteins')
#dev.off()
wilcox.test(tx~yy1$type,subset= yy1$type != 'Undefine')     

```

```{r Figure_S5B, fig.height = 4, fig.width = 7, fig.align = "center"}
tx = colSums(abs(Aging_pro_sigall_matrix),na.rm = T)/
  colSums(!is.na(Aging_pro_sigall_matrix))
yy1 = tissueClass[names(tx),]
#pdf('./out/20230217_aging/Figure5_markers_HE/Figure4B_allchanged_proteins_percentageV2.pdf')
ggplot(,aes(x = yy1$type,y = tx*100,color = yy1$type,fill = yy1$type)) +#geom_violin()+
     geom_boxplot(outlier.size = -1,alpha = 0.1,size = 0.8)+
     geom_jitter(size = 5,width = 0.3)+ theme_classic() +
     ggtitle('% of changed proteins')+
     lghplot.addtheme()+ scale_color_manual(values=c('#EE0000','#3B4992','gray10'))+
     xlab('Tissue Type')+ylab('Percentage of changed proteins(%)')
#dev.off()
wilcox.test(tx~yy1$type,subset= yy1$type != 'Undefine')
```

## 9.1 Figure 5B

```{r p16_plot}
p16 = file2frame('./data/P16_tissues_v20230419.txt')
utissue = unique(p16$Tissue)
tpvalues = data.frame(tissue = utissue,
                     p_Juvenile_vs_Elderly = rep(1,length(utissue)),
                     p_Young_vs_Elderly = rep(1,length(utissue)))
breaks_A = c(0.01,0.08,0.08,0.08,0.08,0.01,0.01,0.01)
breaks_B = c(3,3,3,3,2,2,0.06,3)
for(i in 1:length(utissue)){
    tmpdata = p16[p16$Tissue == utissue[i],]
    tmpdata$Area.Density.log2 = tmpdata$Area.Density*1e3#log2(tmpdata$Area.Density*1e3)
    tmpdata$class = factor(x = tmpdata$class,levels = c('Juvenile','Young','Elderly'))
    tmpp1 = t.test(Area.Density~class,data = tmpdata,subset = tmpdata$class != 'Young')$p.value
    tmpp2 = t.test(Area.Density~class,data = tmpdata,subset = tmpdata$class != 'Juvenile')$p.value
    tpvalues$p_Juvenile_vs_Elderly[i] = tmpp1
    tpvalues$p_Young_vs_Elderly[i] = tmpp2
    
    df2 = data.frame(tmean = aggregate(tmpdata$Area.Density.log2, 
                                       by=list(tmpdata$class), FUN=mean, na.rm = T)$x,
                tsd = aggregate(tmpdata$Area.Density.log2, by=list(tmpdata$class), 
                                FUN=sd, na.rm = T)$x,
                    class = factor(x= c('Juvenile','Young','Elderly'),
                                   levels = c('Juvenile','Young','Elderly'))
                 )
    thepath = paste0('./out/20230217_aging/Figure5_markers_HE/Figure5x_p16_stats', utissue[i],'.pdf')
    pdf(thepath)
    p = ggplot(df2, aes(x = class, y = tmean, fill = class)) + 
      geom_bar(stat = 'identity',color = 'black',position = position_dodge(),
               alpha = 0.5,size = 1.5)+
      geom_errorbar(aes(ymin = tmean,ymax = tmean+tsd),width = .3,size = 1.5)+
      lghplot.addthemeA(size = 28,sizex = 26,sizey = 26)+
      scale_fill_manual(values = c('#008B45FF','#3B4992FF','#EE0000FF'))+ 
      theme(axis.line = element_line(size = 1.2))+ xlab('')+ 
    ylab(bquote('Area Density ' ~ italic(x10) ^italic(3)))+
    #ylab('Log10 Area Density x 1e-7')+
    ggtitle(utissue[i])
    #if(i <=8){
    #    p = p+ scale_y_break(c(breaks_A[i],breaks_B[i]), scales = "free",space  = 0.2)
    #}
    
    print(p)
    dev.off()
}

tpvalues
```

## 9.2 Figure 5C

```{r p21_plot}
p21 = file2frame('./data/P21_tissues_v20230419.txt')
utissue = unique(p21$Tissue)
tpvalues = data.frame(tissue = utissue,
                     p_Juvenile_vs_Elderly = rep(1,length(utissue)),
                     p_Young_vs_Elderly = rep(1,length(utissue)))
breaks_A = c(0.05,0.4,0.4,0.4,0.1,0.4,0.4,0.05)
breaks_B = c(1,4,4,4,2,4,4,1)
for(i in 1:length(utissue)){
    tmpdata = p21[p21$Tissue == utissue[i],]
    tmpdata$Area.Density.log2 = tmpdata$Area.Density*1e3#log2(tmpdata$Area.Density*1e3)
    tmpdata$class = factor(x = tmpdata$class,levels = c('Juvenile','Young','Elderly'))
    tmpp1 = t.test(Area.Density~class,data = tmpdata,subset = tmpdata$class != 'Young')$p.value
    tmpp2 = t.test(Area.Density~class,data = tmpdata,subset = tmpdata$class != 'Juvenile')$p.value
    tpvalues$p_Juvenile_vs_Elderly[i] = tmpp1
    tpvalues$p_Young_vs_Elderly[i] = tmpp2
    
    df2 = data.frame(tmean = aggregate(tmpdata$Area.Density.log2, 
                                       by=list(tmpdata$class), FUN=mean, na.rm = T)$x,
                tsd = aggregate(tmpdata$Area.Density.log2, by=list(tmpdata$class), 
                                FUN=sd, na.rm = T)$x,
                    class = factor(x= c('Juvenile','Young','Elderly'),
                                   levels = c('Juvenile','Young','Elderly'))
                 )
    thepath = paste0('./out/20230217_aging/Figure5_markers_HE/Figure4x_p21_stats', utissue[i],'.pdf')
    pdf(thepath)
    p = ggplot(df2, aes(x = class, y = tmean, fill = class)) + 
      geom_bar(stat = 'identity',color = 'black',position = position_dodge(),
               alpha = 0.5,size = 1.5)+
      geom_errorbar(aes(ymin = tmean,ymax = tmean+tsd),width = .3,size = 1.5)+
      lghplot.addthemeA(size = 28,sizex = 26,sizey = 26)+
      scale_fill_manual(values = c('#008B45FF','#3B4992FF','#EE0000FF'))+ 
      theme(axis.line = element_line(size = 1.2))+ xlab('')+ 
    ylab(bquote('Area Density ' ~ italic(x10) ^italic(3)))+
    #ylab('Log10 Area Density x 1e-7')+
    ggtitle(utissue[i])
    #if(i <=8){
    #    p = p+ scale_y_break(c(breaks_A[i],breaks_B[i]), scales = "free",space  = 0.2)
    #}
    
    print(p)
    dev.off()
}

tpvalues
```

## 9.3 Figure 5D

```{r}
Cellcounts = file2frame('./data/cell_counts_v20230407.txt')

utissue = unique(Cellcounts$Tissue)
tpvalues = data.frame(tissue = utissue,
                     p_Juvenile_vs_Elderly = rep(1,length(utissue)),
                     p_Young_vs_Elderly = rep(1,length(utissue)))
for(i in 1:length(utissue)){
    tmpdata = Cellcounts[Cellcounts$Tissue == utissue[i],]
    tmpdata$class = factor(x = tmpdata$class,
                           levels = c('Juvenile','Young','Elderly'))
    tmpp1 = t.test(number~class,data = tmpdata,subset = tmpdata$class != 'Young')$p.value
    tmpp2 = t.test(number~class,data = tmpdata,subset = tmpdata$class != 'Juvenile')$p.value
    tpvalues$p_Juvenile_vs_Elderly[i] = tmpp1
    tpvalues$p_Young_vs_Elderly[i] = tmpp2
    
    df2 = data.frame(tmean = aggregate(tmpdata$number, by=list(tmpdata$class), 
                                       FUN=mean, na.rm = T)$x,
                tsd = aggregate(tmpdata$number, by=list(tmpdata$class), FUN=sd, na.rm = T)$x,
                    class = factor(x= c('Juvenile','Young','Elderly'),
                                   levels = c('Juvenile','Young','Elderly'))
                 )
    thepath = paste0('./out/20230217_aging/Figure5_markers_HE/Figure4x_HE_cellcounts_', utissue[i],'_v1.pdf')
    pdf(thepath)
    p = ggplot(df2, aes(x = class, y = tmean, fill = class)) + 
      geom_bar(stat = 'identity',color = 'black',position = position_dodge(),alpha = 0.5,size = 1.5)+
      geom_errorbar(aes(ymin = tmean,ymax = tmean+tsd),width = .3,size = 1.5)+
      lghplot.addthemeA(size = 28,sizex = 26,sizey = 26)+
      scale_fill_manual(values = c('#008B45FF','#3B4992FF','#EE0000FF'))+ 
      theme(axis.line = element_line(size = 1.2))+ xlab('')+ 
      ylab('Number of Parenchymal Cells')+ggtitle(utissue[i])
    print(p)
    dev.off()
}

tpvalues
```

# 10 Figure 6: **Translation efficiency**

## 10.1 ratio data construction

```{r}
tissues = names(promet.tissues.Z)
clusterdist = matrix(0,length(tissues),8)
clusteramplitude = matrix(0,length(tissues),8)
clusteramplitude_xx = matrix(0,length(tissues),8)
for(i in 1:length(tissues)){
    mstd =   promet.tissues.Z[[i]] 
    for(j in 1:8){
    tgene = names(mfuzz.promet.whole$cluster)[mfuzz.promet.whole$cluster == j]
    tgene = intersect(tgene,rownames(mstd))
    bx = t(t(mstd[tgene,]) - mfuzz.promet.whole$centers[j,])
    clusterdist[i,j] = mean(sqrt(rowSums(bx^2)/(ncol(bx)-1)),na.rm = T)
    if(length(tgene) < 2){
      clusteramplitude[i,j] = NA
      next;
    }
    tamp = colMeans(mstd[tgene,],na.rm = T)
    clusteramplitude[i,j] = abs(tamp[4]-tamp[1])
    clusteramplitude_xx[i,j] = tamp[4]-tamp[1]
  }
}
rownames(clusteramplitude) = tissues
rownames(clusteramplitude_xx) = tissues
rownames(clusterdist) = tissues
```

```{r}
ratio.tissues = list()
ratio.tissues.info = list()
pro.tissues.forRatio = list()
mrna.tissues.forRatio = list()
tissues = names(pro.tissues)
overlaptissues = intersect(names(pro.tissues),names(mrna.tissues))
for( i in 1:length(overlaptissues)){
    this_tissue = overlaptissues[i]
    thispro = delete_dup_genes_forprotein(pro.tissues[[this_tissue]],
                                          pro.tissues.header[[this_tissue]])
    thismrna = mrna.tissues[[this_tissue]]
    vcol = intersect(colnames(thispro),colnames(thismrna))
    vrow = intersect(rownames(thispro),rownames(thismrna))
    thispro = thispro[vrow,vcol]
    thismrna  = thismrna[vrow,vcol]
    pro.tissues.forRatio[[i]] = thispro
    mrna.tissues.forRatio[[i]] = thismrna
    ratio.tissues[[i]] = thispro - thismrna
    ratio.tissues.info[[i]] = pro.tissues.info[[i]][vcol,]
}
names(ratio.tissues) = overlaptissues
names(pro.tissues.forRatio) = overlaptissues
names(mrna.tissues.forRatio) = overlaptissues
names(ratio.tissues.info) = overlaptissues
```

```{r}
ratio_amplitude = list()
pro_amplitude = list()
mrna_amplitude = list()
n = length(ratio.tissues)
ratio_out = data.frame(meanChangeRatio = rep(0,n),stringsAsFactors = F,
                fc_up_down = rep(0,n),
                tissues = names(ratio.tissues),
                meanChangeproZ = rep(0,n),
                fc_up_down_pro = rep(0,n),
                meanChangemrnaZ = rep(0,n),
                fc_up_down_mrna = rep(0,n))
for (i in 1:length(ratio.tissues)){
    bx = ratio.tissues[[i]]
    bx.info = ratio.tissues.info[[i]]
    id = bx.info$stage < 5
    bx = bx[,id]
    bx.info = bx.info[id,]
    cx = t(aggregate(t(bx), by=list(bx.info$stage), FUN=mean, na.rm = T))
    cx = as.matrix(standardise_1(new("ExpressionSet",exprs = cx)))
    cx = cx[-1,]
    ee = cx[,4]-cx[,1]
    ratio_amplitude[[i]] = ee;
    #hist(ee,30)
    ratio_out$meanChangeRatio[i] = mean(ee,na.rm = T)
    ratio_out$fc_up_down[i] = log2(sum(ee > 0,na.rm = T)/sum(ee < -0,na.rm = T))
    
    bx = pro.tissues.forRatio[[i]]
    bx.info = ratio.tissues.info[[i]]
    id = bx.info$stage < 5
    bx = bx[,id]
    bx.info = bx.info[id,]
    cx = t(aggregate(t(bx), by=list(bx.info$stage), FUN=mean, na.rm = T))
    cx = as.matrix(standardise_1(new("ExpressionSet",exprs = cx)))
    cx = cx[-1,]
    ee = cx[,4]-cx[,1]
    pro_amplitude[[i]] = ee
    #hist(ee,30)
    ratio_out$meanChangeproZ[i] = mean(ee,na.rm = T)
    ratio_out$fc_up_down_pro[i] = log2(sum(ee > 0,na.rm = T)/sum(ee < -0,na.rm = T))
    
    bx = mrna.tissues.forRatio[[i]]
    bx.info = ratio.tissues.info[[i]]
    id = bx.info$stage < 5
    bx = bx[,id]
    bx.info = bx.info[id,]
    cx = t(aggregate(t(bx), by=list(bx.info$stage), FUN=mean, na.rm = T))
    cx = as.matrix(standardise_1(new("ExpressionSet",exprs = cx)))
    cx = cx[-1,]
    ee = cx[,4]-cx[,1]
    mrna_amplitude[[i]] = ee
    #hist(ee,30)
    ratio_out$meanChangemrnaZ[i] = mean(ee,na.rm = T)
    ratio_out$fc_up_down_mrna[i] = log2(sum(ee > 0,na.rm = T)/sum(ee < -0,na.rm = T))
}
rownames(ratio_out) = names(ratio.tissues)
names(ratio_amplitude) = names(ratio.tissues)
names(pro_amplitude) = names(ratio.tissues)
names(mrna_amplitude) = names(ratio.tissues)
```

## 10.2 Figure 6A design

## 10.3 Figure S9 plot mRNA and protein

```{r}
mRNA.mean = list()
pro.mean = list()
for(i in 1:length(pro.tissues.forRatio)){
    mRNA.mean[[i]] = rowMeans(mrna.tissues.forRatio[[i]],na.rm = T)
    pro.mean[[i]] = rowMeans(pro.tissues.forRatio[[i]],na.rm = T)
}
names(mRNA.mean) = names(mrna.tissues.forRatio)
names(pro.mean) = names(pro.tissues.forRatio)
RNA.v = list_to_matrix(mRNA.mean,names(mRNA.mean))
pro.v = list_to_matrix(pro.mean,names(pro.mean))

pp = list()
for(i in 1:ncol(pro.v)){
    idaa = !is.na(RNA.v[,i]) & !is.na(pro.v[,i])
    tcor = cor.test(RNA.v[idaa,i],pro.v[idaa,i])$estimate
    tmpdata = data.frame(xx = RNA.v[idaa,i],
                        yy = pro.v[idaa,i])
    pp[[i]] = ggplot(tmpdata,aes(x= xx,y = yy))+geom_point(size =1,) + theme_bw()+
        theme(plot.margin = margin(0.1,0.1,0.1,0.1,"cm"))+
        lghplot.addthemeA(size = 16,sizex = 16,sizey = 16)+        
        annotate(geom="text", x=4, y=25, 
                label = paste('R=',signif(tcor,3)),
               color="darkblue",size = 8,face = "italic")+
        xlab('')+ylab('')+ggtitle(colnames(pro.v)[i])
}

png(file = "./out/20230217_aging/Figure6_ratio_mrna_pro/Figure S9_mrna_vs_pro.png",
    width = 1250,height = 1500)
grid.arrange(arrangeGrob(grobs = pp, ncol = 5,
                         bottom=textGrob('mRNA expression(log2 CPM)', 
                                         gp=gpar(fontface="bold",  fontsize=22)),
                        left = textGrob('Protein abundance(log2 Peak Area)',
                                  gp=gpar(fontface="bold",  fontsize=22),rot=90)))
dev.off()

```

## 10.4 Figure S10 predict protein vs measured protein

```{r}
ratio = pro.v - RNA.v
ratio = apply(ratio,1,median,na.rm =T)

# remove ratio = 0 and NA
#id = is.na(ratio) | ratio == 0
#ratio = ratio[!id]

pro.prediction = RNA.v + ratio

pp = list()
for(i in 1:ncol(pro.v)){
    idaa = !is.na(pro.prediction[,i]) & !is.na(pro.v[,i])
    tcor = cor.test(pro.prediction[idaa,i],pro.v[idaa,i])$estimate
    tmpdata = data.frame(xx = pro.prediction[idaa,i],
                        yy = pro.v[idaa,i])
    pp[[i]] = ggplot(tmpdata,aes(x= xx,y = yy))+geom_point(size =1,) + theme_bw()+
        theme(plot.margin = margin(0.1,0.1,0.1,0.1,"cm"))+
        lghplot.addthemeA(size = 16,sizex = 16,sizey = 16)+        
        annotate(geom="text", x=15, y=25, 
                label = paste('R=',signif(tcor,3)),
               color="darkblue",size = 8,face = "italic")+
        xlab('')+ylab('')+ggtitle(colnames(pro.v)[i])
}

png(file = "./out/20230217_aging/Figure6_ratio_mrna_pro/Figure S10_predPro_vs_pro.png",
    width = 1250,height = 1500)
grid.arrange(arrangeGrob(grobs = pp, ncol = 5,
                         bottom=textGrob('Predicted protein abundance(log2 Peak Area)', 
                                         gp=gpar(fontface="bold",  fontsize=22)),
                        left = textGrob('Protein abundance(log2 Peak Area)', 
                                        gp=gpar(fontface="bold",  fontsize=22),rot=90)))
dev.off()

```

## 10.5 Figure 6B density plot

```{r Figure_6B, fig.height = 12.5, fig.width = 25, fig.align = "center"}
systems.Color = substr(pal_aaas()(10),1,7)
names(systems.Color) = unique(tissue.systems)
systems.Color.fortissue = systems.Color[tissue.systems]
names(systems.Color.fortissue) = names(ratio_amplitude)
tissueClass$color = tissueClass$type
tissueClass$color[tissueClass$color == 'Type I'] = '#BB0021'
tissueClass$color[tissueClass$color == 'Type II'] = '#3B4992'
tissueClass$color[tissueClass$color == 'Undefine'] = '#B09C85'


# hist plot
p1 = list()
xplot <- function(ratio_amplitude,xtissue,tissueClass){
    tcolor = tissueClass[xtissue,]$color
    tmp =  ggplot(,aes(x = ratio_amplitude[[xtissue]]))+theme_classic()+
         #geom_histogram(binwidth=0.1,aes(y=..density..), colour="black", fill="white")+
         geom_vline(xintercept=0, linetype="dashed", color = "blue", size=1)+
         geom_density(alpha=.6, fill=tcolor) +lghplot.addtheme()+
      theme(axis.line = element_line(size = 1.2))+ 
      xlab('')+ylab('')+ggtitle(xtissue)
    return(tmp)
}

index = sort.int(ratio_out$meanChangeRatio,decreasing = F,index.return = T)$ix
vclass = tissueClass[rownames(ratio_out)[index],]
idx = sort.int(vclass$class,decreasing = F,index.return = T)$ix
vclass = vclass[idx,]
for(i in 1:nrow(vclass)){
    xtissue = vclass$tissue[i]
    p1[[i]] = xplot(ratio_amplitude,xtissue,tissueClass)
}
#pdf(file = './out/20230217_aging/Figure6_ratio_mrna_pro/Figure6B_Changed_ratio_distribute_based_on_promet_sortby_v2.pdf',width = 25,height = 12.5)
grid.arrange(arrangeGrob(grobs = p1,ncol = 8,
           top=textGrob('Distribution of changed translation efficiency to degradation (cTED) with aging', 
                                        gp=gpar(fontface="bold",  fontsize=30)),
            bottom=textGrob('Changed normalized TED', 
                                        gp=gpar(fontface="bold",  fontsize=30)),
            left = textGrob('Density', 
                                        gp=gpar(fontface="bold",  fontsize=30),rot=90)))
#graphics.off()
```

## 10.6 Figure 6C ratio plot

```{r Figure_6C, fig.height = 7, fig.width = 8, fig.align = "center"}


ratio_updown = zeros(length(ratio_amplitude),2)
tissues =names(ratio_amplitude)
rownames(ratio_updown) = tissues
colnames(ratio_updown) = c('pec_up','pec_down')
for(i in 1:length(tissues)){
    tmp = as.vector(ratio_amplitude[[i]])
    tmpN = sum(!is.na(tmp))
    ratio_updown[i,1] = sum(tmp > 0,na.rm =T)/tmpN
    ratio_updown[i,2] = sum(tmp < 0,na.rm =T)/tmpN
}
index = sort.int(ratio_out$meanChangeRatio,decreasing = F,index.return = T)$ix
ratio_updown = ratio_updown[index,]

tmp = melt(ratio_updown,stringsAsFactor =F)
colnames(tmp) = c('tissues','Direction','perctage')
tmp$tissues = factor(tmp$tissues,levels = rownames(vclass))
#pdf('./out/20230217_aging/Figure6_ratio_mrna_pro/
#   Figure6C_tissue_trans_effectiveness_amplitude_ratio_V2.pdf',width = 8,height = 7)
ggplot(data=tmp, aes(x=tissues, y=perctage, fill=Direction))+theme_classic()+
       geom_col(position = "fill",alpha = 0.8)+
      ylab('Relative ratio')+ xlab('')+scale_fill_manual(values=c('#BB0021', '#3B4992'))+
      lghplot.addtheme(legend.position = 'top',hjust = 1,size = 14)+
      ggtitle('Relative ratio of up- and down- TEDs')+
      theme(legend.text=element_text(size=20))
#graphics.off()
```

## 10.7 Figure 6D correlation

```{r Figure_6D, fig.height = 5.3, fig.width = 7, fig.align = "center"}
mean_aging_amplitude = rowMeans(clusteramplitude_xx[,])
mean_aging_amplitude = mean_aging_amplitude[rownames(ratio_out)]
tmpdata = data.frame(ratioChange = ratio_out$meanChangeRatio,stringsAsFactors = F,
                     ratioFC = ratio_out$fc_up_down,
                     amplitude = mean_aging_amplitude,
                    tissues = rownames(ratio_out))
rownames(tmpdata) = tmpdata$tissues
tmpdata$tissueclass = tissueClass[rownames(tmpdata),]$type
tmp1 = tmpdata
cor.test(tmp1$ratioChange,tmp1$amplitude)
#pdf('./out/20230217_aging/Figure6_ratio_mrna_pro/Figure6D_overall_translation_effective_score_vs_aging_amplitute_promet_V2.pdf',width = 7,height = 5)
p = ggplot(tmp1,aes(ratioChange,amplitude,color = tissueclass)) + 
  theme_classic()+
      geom_point(size = 6,aes(color = tissueclass)) + 
      lghplot.addtheme()+geom_smooth(color = 4,size = 0.5,method = 'lm')+
      geom_text_repel(aes(label = tissues),size = 4,box.padding = 0.5)+
      #scale_color_aaas(alpha = 0.6)+scale_fill_aaas(alpha = 0.6)+
      scale_color_manual(values=c('#BB0021', '#3B4992','#B09C85'))+
      theme(axis.line = element_line(size = 1.2))+
      xlab('Averaged changed translation effective score')+
      ylab('Aging amplitude')+
      ggtitle('Correlation between MAA and cTED')
print(p)
   
#dev.off()

```

# 11 Figure 7

## 11.1 GO data construct

```{r}
#outids = c('R-HSA-72766','GO:0006413','hsa03010','ko04142','hsa04120','ko03050')
outids = c('R-HSA-72766','hsa03010','hsa04142','hsa04120','hsa03050')
#tpath = paste0('./out/20210428_aging/promet/tissues/metascape/Aging_up_genes_metascape/Enrichment_GO/','GO_membership.csv')
tpath = paste0('./out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_up/Enrichment_GO/','GO_membership.csv')
thisgo = file2frame(tpath,sep = ',')
thisgo= thisgo[!duplicated(thisgo$GO),]
xid = is.element(thisgo$GO,outids)
thisgo = thisgo[xid,]
#rownames(thisgo) = paste0(thisgo$GO,':',thisgo$Description)
rownames(thisgo) = paste0(thisgo$GO)
thisgo.matrix = -as.matrix(thisgo[,substr(colnames(thisgo),1,6)== 'X_LogP'])

# replace qval
tpath1a = './out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_up/Enrichment_GO/GO_AllLists.csv'
go1qval = file2frame(tpath1a,sep = ',',header = T)
rownames(go1qval) = paste0(go1qval$GO,'X_LogP_',go1qval$GeneList)
go1term = rownames(thisgo.matrix)
cname = colnames(thisgo.matrix)
for(i in 1:nrow(thisgo.matrix)){
    for(j in 1:ncol(thisgo.matrix)){
        tmpname = paste0(go1term[i],cname[j])
        thisgo.matrix[i,j] = abs(go1qval[tmpname,]$Log.q.value.)
    }
}
thisgo.matrix[is.na(thisgo.matrix)] = 0


thisgo.matrix[abs(thisgo.matrix) < 3] = 0
colnames(thisgo.matrix) = capitalize(gsub('X_LogP_','',colnames(thisgo.matrix)))
colnames(thisgo.matrix) = gsub('Cluster','C',colnames(thisgo.matrix))
thisgo.matrix.up = thisgo.matrix
tmpname = rownames(thisgo.matrix.up)
#add hsa03050 
thisgo.matrix.up = rbind(thisgo.matrix.up,rep(0,30))
rownames(thisgo.matrix.up) = c(tmpname,'hsa03050')
thisgo.matrix.up = thisgo.matrix.up[outids,]
thisgo.matrix.up
```

```{r}
outids = c('R-HSA-72766','hsa03010','hsa04142','hsa04120','hsa03050')
outids_des = c('Translation','Ribosome','Lysosome',
                               'Ubiquitin mediated proteolysis', 'Proteasome') 
tpath = paste0('./out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_down/Enrichment_GO/','GO_membership.csv')
thisgo = file2frame(tpath,sep = ',')
thisgo= thisgo[!duplicated(thisgo$GO),]
xid = is.element(thisgo$GO,outids)
thisgo = thisgo[xid,]
rownames(thisgo) = paste0(thisgo$GO)

# replace qval
tpath1a = './out/20230217_aging/Figure2_DEG_GO_tissue/protein/metascape_DEpro_down/Enrichment_GO/GO_AllLists.csv'
go1qval = file2frame(tpath1a,sep = ',',header = T)
rownames(go1qval) = paste0(go1qval$GO,'X_LogP_',go1qval$GeneList)
go1term = rownames(thisgo.matrix)
cname = colnames(thisgo.matrix)
for(i in 1:nrow(thisgo.matrix)){
    for(j in 1:ncol(thisgo.matrix)){
        tmpname = paste0(go1term[i],cname[j])
        thisgo.matrix[i,j] = abs(go1qval[tmpname,]$Log.q.value.)
    }
}
thisgo.matrix[is.na(thisgo.matrix)] = 0


thisgo.matrix = as.matrix(thisgo[,substr(colnames(thisgo),1,6)== 'X_LogP'])
thisgo.matrix[abs(thisgo.matrix) < 3 ] = 0
colnames(thisgo.matrix) = capitalize(gsub('X_LogP_','',colnames(thisgo.matrix)))
colnames(thisgo.matrix) = gsub('Cluster','C',colnames(thisgo.matrix))
rownames(thisgo.matrix)
thisgo.matrix.down = thisgo.matrix
thisgo.matrix.down = thisgo.matrix.down[outids,]
thisgo.matrix.down
```

## 11.2 Figure 7A

```{r Figure_7A, fig.height = 3.2, fig.width = 10, fig.align = "center"}
outids = c('R-HSA-72766','hsa03010','hsa04142','hsa04120','hsa03050')
outids_des = c('Translation','Ribosome','Lysosome',
                               'Ubiquitin mediated proteolysis', 'Proteasome')
thisgo.matrix.all = thisgo.matrix.up+ thisgo.matrix.down

tmpnames = rownames(tissueClass)
tmpnames = c(tmpnames[tissueClass$type == 'Type II'],tmpnames[tissueClass$type == 'Undefine'],
             tmpnames[tissueClass$type == 'Type I'])

thisgo.matrix.all = thisgo.matrix.all[,tmpnames]
thisgo.matrix.all[thisgo.matrix.all >= 3] = 1
thisgo.matrix.all[thisgo.matrix.all <= -3] = -1

tclass = data.frame(class = tissueClass[tmpnames,]$type,row.names = tmpnames)
colnames(tclass) <- c("Tissue_type")
ann_colors = list(
   Tissue_type= c('#FD60A7','gray85','#4fffF7')
    
)
names(ann_colors$Tissue_type) = c('Type I','Undefine', 'Type II')
#sort tissues
tmpname  = colnames(thisgo.matrix.all)
rownames(thisgo.matrix.all) = outids_des
pheatmap::pheatmap(thisgo.matrix.all,cluster_rows = F,annotation_colors = ann_colors,
                   annotation_col = tclass, 
                 main = 'Enrichment in mRNA translation and protein degradation related pathways',
                   cluster_cols = F,fontsize_row = 9,fontsize_col = 10,
                 fontsize = 10,treeheight_row = 20,treeheight_col = 20,legend = T,
                  color=colorRampPalette(c('#3B4992','gray99','#EE0000FF'))(6),
                  #file ="./out/20230217_aging/Figure6_heatmap_TED_machanism/Figure 6A_enrichment_metascape_promet_all_tissue_TEDs_V6.pdf",
                 height = 3,width = 10
                  )
```

```{r}
idx = tclass$Tissue_type != 'Undefine'
tmpaa = thisgo.matrix.all[,idx]
tmpclass = tclass$Tissue_type[idx]
rownames(tmpaa)
cor.test(abs(tmpaa[1,]),(tmpclass == 'Type I')+0,method = 'spearman')
cor.test(abs(tmpaa[2,]),(tmpclass == 'Type I')+0,method = 'spearman')
cor.test(abs(tmpaa[3,]),(tmpclass == 'Type I')+0,method = 'spearman')
cor.test(abs(tmpaa[4,]),(tmpclass == 'Type I')+0,method = 'spearman')
cor.test(abs(tmpaa[5,]),(tmpclass == 'Type I')+0,method = 'spearman')
```

## 11.3 Figure 7B

```{r}
require(data.table)
amplitude_matrix = list()

for(i in 1:length(pro_amplitude)){
    tmp = matrix(pro_amplitude[[i]],1,length(pro_amplitude[[i]]))
    tmp = as.data.frame(tmp)
    colnames(tmp) = names(pro_amplitude[[i]])
    amplitude_matrix[[i]] = tmp
}
#amplitude_matrix = rbindlist(amplitude_matrix,fill = T)
amplitude_matrix = t(as.matrix(rbindlist(amplitude_matrix,fill = T)))
colnames(amplitude_matrix) = names(pro_amplitude)

vid = rowSums(is.na(amplitude_matrix)) < ncol(amplitude_matrix)/2
amplitude_matrix.v = amplitude_matrix[vid,]
amplitude_matrix.v[is.na(amplitude_matrix.v)] = NA
dim(amplitude_matrix.v)
tcor = cor(t(amplitude_matrix.v),ratio_out$meanChangeRatio,use = "pairwise")
names(tcor) = rownames(amplitude_matrix.v)
gsea_input = sort(tcor,decreasing = T)

calcp <- function(x,y){
    return(cor.test(x,y,use = "pairwise")$p.value)
}
tpvalue = apply(amplitude_matrix.v,1,calcp,ratio_out$meanChangeRatio)
tpvalue = tpvalue[names(gsea_input)]
```

```{r Figure_7B, fig.height = 8.2, fig.width = 8, fig.align = "center"}
xnames = names(tpvalue)
trans_initialA  = xnames[substr(xnames,1,3) == 'EIF']
trans_initialA = sort(trans_initialA)
trans_initialA
trans_initialB  = xnames[substr(xnames,1,3) == 'EEF']
trans_initialB = sort(trans_initialB)
trans_initialB
trans_initialA = c(trans_initialA,trans_initialB)
for(i in 1:length(pro_amplitude)){
    tx = pro_amplitude[[i]][trans_initialA]
    #tx= tx[!is.na(tx)]
    tmp = data.frame(genes = trans_initialA,stringsAsFactors = F,
                    amplitude = as.vector(tx))
    if(i ==1){
        tmpout = tmp;
    }else{
        tmpout = cbind(tmpout,tmp[,'amplitude'])
    }
    
}
tmpout = t(tmpout[,-1])

rownames(tmpout) = names(pro_amplitude)
colnames(tmpout) = trans_initialA
idx = colSums(is.na(tmpout)) < 3
sum(idx)
#tmpout = tmpout[rowSums(is.na(tmpout)) < 3,]
tmpout = tmpout[,idx]


metadata <- tissueClass[rownames(tmpout),c('type','type')]
tclass = data.frame(class = metadata$type,row.names = rownames(metadata))
colnames(tclass) <- c("Tissue_type")

ann_colors = list(
   Tissue_type= c('#FD60A7','gray85','#4fffF7')
    
)
names(ann_colors$Tissue_type) = c('Type I','Undefine', 'Type II')

pheatmap::pheatmap(t(tmpout),annotation_col = tclass,cluster_rows = F,
                   annotation_colors = ann_colors,
                   main = 'Change in initiation and elongation factors',
                   color=colorRampPalette(c('#3B4992','gray99','#EE0000'))(30),
                    #file ="./out/20230217_aging/Figure6_heatmap_TED_machanism/Figure6B_heatmap_translation_elongationV3.pdf",
                  height = 8,width = 8
                  )  
```

## 11.4 Figure 7C

```{r Figure_7C, fig.height = 8.2, fig.width = 8, fig.align = "center"}
xnames = names(tpvalue)
ribosome = file2frame('./data/ribosome_proteins_from_kegg.txt',header = F)
ribosome = ribosome$V1
trans_initialA = intersect(ribosome,xnames)
trans_initialA
for(i in 1:length(pro_amplitude)){
    tx = pro_amplitude[[i]][trans_initialA]
    #tx= tx[!is.na(tx)]
    tmp = data.frame(genes = trans_initialA,stringsAsFactors = F,
                    amplitude = as.vector(tx))
    if(i ==1){
        tmpout = tmp;
    }else{
        tmpout = cbind(tmpout,tmp[,'amplitude'])
    }
    
}
tmpout = t(tmpout[,-1])

rownames(tmpout) = names(pro_amplitude)
colnames(tmpout) = trans_initialA
idx = colSums(is.na(tmpout)) < 3
sum(idx)
#tmpout = tmpout[rowSums(is.na(tmpout)) < 5,]
tmpout = tmpout[,idx]
tmpout = tmpout[rownames(tissueClass),]

metadata <- tissueClass#[rownames(tmpout),]
#metadata <- tissueClass[rownames(tmpout),c('class','class')]
tclass = data.frame(class = metadata$type,row.names = rownames(metadata))
colnames(tclass) <- c("Tissue_type")

ann_colors = list(
   Tissue_type= c('#FD60A7','gray85','#4fffF7')
    
)
names(ann_colors$Tissue_type) = c('Type I','Undefine', 'Type II')

pheatmap::pheatmap(t(tmpout),annotation_col = tclass,cluster_rows = F,
                   annotation_colors = ann_colors,
                   main = 'Change in ribosomal proteins',
                   cluster_cols = T,color=colorRampPalette(c('#3B4992','gray99','#EE0000FF'))(30),
                    #file ="./out/20230217_aging/Figure6_heatmap_TED_machanism/Figure6C_heatmap_ribosomeV3.pdf",
                  height = 8,width = 8
                  )
```
