---
title: "Get cell line aging datasets"
author: "Gong-Hua Li"
date: "`r Sys.Date()`"
output:
  html_document: 
    toc: yes
    toc_float: yes
    toc_depth: 3
  pdf_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = 'E:/iHuman/projects/Cell_line_aging/data/cell_line_data_matrix')
knitr::opts_chunk$set(echo = TRUE,warning = FALSE, message = FALSE)
```

# 1. setwd and env

```{r}
library(rGPMM)
library(pheatmap)
#library(DESeq2)
library(edgeR)
library(ggsci)
#library(pracma)a
library(data.table)
library(fgsea)
library(GEOquery)
library(EnhancedVolcano)
library(ggplot2)

```

```{r}
geneheader = file2frame('./Homo_sapiens.GRCh38.99.ensemble_symbol_biotype.txt')
rownames(geneheader) = geneheader$gene_id

str2num <- function(x){
  rname = rownames(x)
  cname = colnames(x)
  y = matrix(as.numeric(x),ncol = ncol(x))
  rownames(y) = rname
  colnames(y) = cname
  return(y)
}

cbindlist <- function(list, key, col_names = NULL){
  result <- list[[1]]
  lapply(list, function(x) data.table::setkeyv(x, key))
  for(i in list[-1]){
    result <- result[i]
  }
  
  if(!is.null(col_names)){
    
    if(length(col_names)!=length(colnames(result)[-1])){
      stop('Vector with new column names must be the same length as the list', call. = FALSE)
    }
    
    colnames(result)[-1] <- col_names
  }
  
  return(result)
}

writetxt_forGPMM <- function(x,filename){
  name = rownames(x)
  y = cbind(as.data.frame(name),x)
  writetxt(y, filename)
}
```

# 2. dataset

## 2.1 GSE156384_iPSCs

**Information GSE156384**

-   **Citation**: Schaniel C, Dhanan P, Hu B, Xiong Y et al. A library of induced pluripotent stem cells from clinically well-characterized, diverse healthy human individuals. Stem Cell Reports 2021 Dec 14;16(12):3036-3049. PMID: [34739849](https://www.ncbi.nlm.nih.gov/pubmed/34739849 "Link to PubMed record")

-   **Summary**: Generation and extensive characterization of a gender-balanced, racially/ethnically diverse library of iPSCs from forty clinically healthy human individuals who range in age from 22-61. Specifically, here we provide mRNAseq data of duplicate samples of one clone from each of these forty iPSC lines

-   **Submission information**

    |                   |                                         |
    |-------------------|-----------------------------------------|
    | Submission date   | Aug 18, 2020                            |
    | Last update date  | Jan 11, 2022                            |
    | Contact name      | Ravi Iyengar                            |
    | Organization name | Icahn School of Medicine at Mount Sinai |
    | Department        | Pharmacological Sciences                |
    | Street address    | 1425 Madison Avenue, Room 12-70C        |
    | City              | New York                                |
    | State/province    | NY                                      |
    | ZIP/Postal code   | 10029                                   |
    | Country           | USA                                     |

```{r}
## readmatrix
rawCounts = read.table(gzfile("./GSE156384/GSE156384_Gene-Read-Counts.tsv.gz"))


## get clincal data
clin = readxl::read_excel("./GSE156384/mmc3.xlsx")
colnames(clin)[4:6]  =c('Sex','Age','Race')
clin = clin[-1,]
ids = colnames(rawCounts)
clin$ID_short = clin$`iPSC ID`
clin$ID_short = gsub('-','_',clin$ID_short)
rownames(clin) = clin$ID_short
ids_st = substr(ids,1,9)
clin = clin[ids_st,]
clin$ID = ids
rownames(clin) = clin$ID

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

#write
writetxt_forGPMM(clin,'./GSE156384/clin_GSE156384.txt')
writetxt_forGPMM(fpkm,'./GSE156384/fpkm_GSE156384.txt')
writetxt_forGPMM(tpm,'./GSE156384/tpm_GSE156384.txt')
writetxt_forGPMM(rawCounts,'./GSE156384/rawCounts_GSE156384.txt')

```

## 2.2 GSE205867_neutrophils

**Information GSE205867**

-   **Citation**: Yu X, Li L, Zhang M, Liu J et al. Transcriptional analysis of neutrophils from patients with Behçet's disease reveals activation and chemotaxis of neutrophils. Clin Immunol 2022 Dec;245:109161. PMID: [36243348](https://www.ncbi.nlm.nih.gov/pubmed/36243348 "Link to PubMed record")

-   **Summary**: Behçet's Disease (BD) is a chronic and systemic vasculitis with unknown etiology. Although BD is considered a condition linking both autoimmunity and autoinflammation, aberrant innate immunity has emerged in its significant pathogenetic role, among which neutrophils directly drive inflammation in BD. To investigate neutrophil aberrance in BD, we performed bulk RNA-sequencing on isolated peripheral neutrophils from 10 pairs of BD and sex- and age-matched healthy control (HCs).

-   **Submission information**

    +-------------------+-------------------------------------------------------------+
    | Submission date   | Jun 10, 2022                                                |
    +-------------------+-------------------------------------------------------------+
    | Last update date  | Nov 01, 2022                                                |
    +-------------------+-------------------------------------------------------------+
    | Contact name      | Wenjie Zheng                                                |
    +-------------------+-------------------------------------------------------------+
    | E-mail(s)         | [wenjzheng\@gmail.com](mailto:wenjzheng@gmail.com)          |
    +-------------------+-------------------------------------------------------------+
    | Organization name | Peking Union Medical College Hospital                       |
    +-------------------+-------------------------------------------------------------+
    | Street address    | No1 ShuaiFuYuan，Dong Cheng District ，Beijing, Beijing, CN |
    +-------------------+-------------------------------------------------------------+
    | City              | Beijing                                                     |
    +-------------------+-------------------------------------------------------------+
    | ZIP/Postal code   | 100005                                                      |
    +-------------------+-------------------------------------------------------------+
    | Country           | China                                                       |
    +-------------------+-------------------------------------------------------------+

```{r}
#readmatrix
rawCounts = file2frame('./GSE205867/GSE205867_count_matrix.csv',sep = ',')
rawCounts[,1] = geneheader[rawCounts[,1],]$gene_name
rawCounts = rawCounts[!is.na(rawCounts[,1]),]
rawCounts = del_duplicate_genes(rawCounts)

# only used contrl
vid = substr(colnames(rawCounts),1,2) == 'HC'
rawCounts = rawCounts[,vid]

# get clin data
clin  = file2frame('./GSE205867/GSE205867_sample_info.csv',sep = ',')
rownames(clin) = clin$sample
clin = clin[colnames(rawCounts),]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

#write
writetxt_forGPMM(clin,'./GSE205867/clin_GSE205867.txt')
writetxt_forGPMM(fpkm,'./GSE205867/fpkm_GSE205867.txt')
writetxt_forGPMM(tpm,'./GSE205867/tpm_GSE205867.txt')
writetxt_forGPMM(rawCounts,'./GSE205867/rawCounts_GSE205867.txt')

```

## 2.3 GSE51518_skin fibroblasts

**Information GSE51518**

-   **Citation**: Jung M, Jin SG, Zhang X, Xiong W et al. Longitudinal epigenetic and gene expression profiles analyzed by three-component analysis reveal down-regulation of genes involved in protein translation in human aging. Nucleic Acids Res 2015 Sep 3;43(15):e100. PMID: [25977295](https://www.ncbi.nlm.nih.gov/pubmed/25977295 "Link to PubMed record")

-   **Summary**: Aging signatures developed from a longitudinal study design are dominated by reduced transcription of genes involved in protein synthesis.Aging is a multifactorial process where the impact of singular components still remains unclear. Furthermore, previous studies were focused on measuring specific traits such as DNA -methylation and used categorical group-wise designs, unable to capture intra-individual signature changes. Here we have developed a new method for a longitudinal, age-related analysis combining the merits of a pair-wise design with the statistical power of gene set enrichment analysis. We present an integrated analysis, including transcriptional changes and genome-wide epigenetic changes in DNA- methylation, H3K4- and H3K27- histone methylation in promoter regions. We tested our method on a rare collection of paired skin fibroblast samples from male middle age to old age transitions and obtained functional, age-related clusters. By using a set of only ten individuals, we could demonstrate a high overlap of functional terms to previously established tissue-independent age signatures including extracellular matrix, apoptosis and oxidative stress. Importantly, we identify protein translation-related processes as the main cluster of age-driven, specific down regulation.

-   **Submission information**

    |                   |                                        |
    |-------------------|----------------------------------------|
    | Submission date   | Oct 21, 2013                           |
    | Last update date  | May 15, 2019                           |
    | Contact name      | Marc Jung                              |
    | E-mail(s)         | [mjung\@coh.org](mailto:mjung@coh.org) |
    | Organization name | City Of Hope                           |
    | Department        | Biology                                |
    | Lab               | Gerd Pfeifer                           |
    | Street address    | 1500 East Duarte Street                |
    | City              | Duarte                                 |
    | State/province    | CA                                     |
    | ZIP/Postal code   | 91010                                  |
    | Country           | USA                                    |

```{r}
# download from raw data and using salmon to get fpkm and tpm
```

## 2.4 GSE102215 neutrophils

**Information GSE102215**

-   **Citation**: Knight JS, Meng H, Coit P, Yalavarthi S et al. Activated signature of antiphospholipid syndrome neutrophils reveals potential therapeutic target. JCI Insight 2017 Sep 21;2(18). PMID: [28931754](https://www.ncbi.nlm.nih.gov/pubmed/28931754 "Link to PubMed record")

-   **Summary**: Antiphospholipid antibodies are a leading cause of thrombosis. To further elucidate the role of neutrophils in APS, we performed a comprehensive transcriptome analysis using the total RNA of neutrophils isolated from 9 patients with primary APS and 9 age, sex and ethnicity-matched healthy controls using next-generation RNA-seq.

-   **Submission information**

    |                   |                                                  |
    |-------------------|--------------------------------------------------|
    | Submission date   | Aug 03, 2017                                     |
    | Last update date  | May 15, 2019                                     |
    | Contact name      | Amr Sawalha                                      |
    | E-mail(s)         | [asawalha\@umich.edu](mailto:asawalha@umich.edu) |
    | Organization name | University of Michigan                           |
    | Street address    | 1150 W. Medical Center Drive                     |
    | City              | Ann Arbor                                        |
    | State/province    | MI                                               |
    | ZIP/Postal code   | 48109                                            |
    | Country           | USA                                              |

```{r}
rawCounts = read.table(gzfile("./GSE102215/GSE102215_raw_counts.txt.gz"))
rownames(rawCounts) = rawCounts[,1]
colnames(rawCounts)  = rawCounts[1,]
rawCounts = as.matrix(rawCounts[-1,-1])
rawCounts = str2num(rawCounts)

# get clin data
clin_org = getGEO(filename="./GSE102215/GSE102215_series_matrix.txt.gz")
clin = clin_org@phenoData@data
rownames(clin)= clin$title
colnames(clin) = gsub(':','_',colnames(clin))
colnames(clin) = gsub(' ','_',colnames(clin))
vid = clin$disease_ch1 == 'healthy'
clin = clin[vid,]

rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

#write
writetxt_forGPMM(clin,'./GSE102215/clin_GSE102215.txt')
writetxt_forGPMM(fpkm,'./GSE102215/fpkm_GSE102215.txt')
writetxt_forGPMM(tpm,'./GSE102215/tpm_GSE102215.txt')
writetxt_forGPMM(rawCounts,'./GSE102215/rawCounts_GSE102215.txt')

```

## 2.5 GSE135902 Monocytes

**Information GSE135902**

-   **Citation**: Franzini A, Pomicter AD, Yan D, Khorashad JS et al. The transcriptome of CMML monocytes is highly inflammatory and reflects leukemia-specific and age-related alterations. Blood Adv 2019 Oct 22;3(20):2949-2961. PMID: [31648319](https://www.ncbi.nlm.nih.gov/pubmed/31648319 "Link to PubMed record")

-   **Summary**: Chronic myelomonocytic leukemia (CMML) is an aggressive myeloid neoplasm of older individuals characterized by persistent monocytosis. Somatic mutations in CMML are heterogeneous and only partially explain the variability in clinical outcomes. Recent data suggest that cardiovascular morbidity is increased in CMML and contributes to reduced survival. Clonal hematopoiesis of indeterminate potential (CHIP), the presence of mutated blood cells in hematologically normal individuals, is a precursor of age-related myeloid neoplasms and associated with increased cardiovascular risk. To isolate CMML-specific alterations from those related to aging, we performed RNA sequencing and DNA methylation profiling on purified monocytes from CMML patients and from age-matched (old) and young healthy controls. We found that the transcriptional signature of CMML monocytes is highly pro-inflammatory, with upregulation of multiple inflammatory pathways, including tumor necrosis factor, IL-6 and IL-17 signaling, while age per se does not significantly contribute to this pattern. We observed no consistent correlations between aberrant gene expression and CpG island methylation, suggesting that pro-inflammatory signaling in CMML monocytes is governed by multiple and complex regulatory mechanisms. We propose that pro-inflammatory monocytes may contribute to cardiovascular morbidity in CMML patients, and promote progression by selection of mutated cell clones. Our data raise questions whether asymptomatic CMML patients may benefit from monocyte-depleting or anti-inflammatory therapies.

-   **Submission information**

    |                   |                           |
    |-------------------|---------------------------|
    | Submission date   | Aug 16, 2019              |
    | Last update date  | Nov 19, 2019              |
    | Contact name      | Chris Stubben             |
    | Organization name | Huntsman Cancer Institute |
    | Department        | University of Utah        |
    | Street address    | 2000 Cir of Hope Dr       |
    | City              | Salt Lake City            |
    | State/province    | Utah                      |
    | ZIP/Postal code   | 84103                     |
    | Country           | USA                       |

```{r}
rawCounts = file2frame('./GSE135902/GSE135902_counts_40.txt')
rawCounts[,1] = geneheader[rawCounts[,1],]$gene_name
rawCounts = rawCounts[!is.na(rawCounts[,1]),]
rawCounts = del_duplicate_genes(rawCounts)

# get clin data
clin_org = getGEO(filename="./GSE135902/GSE135902_series_matrix.txt.gz")
clin = clin_org@phenoData@data
IDs = clin$title
IDs = paste0('X',regmatches(IDs,regexpr('\\d+X.+',IDs)))
clin$ID = IDs
rownames(clin) = IDs
colnames(clin) = gsub(':','_',colnames(clin))
colnames(clin) = gsub(' ','_',colnames(clin))
vid = clin$disease_state_ch1 != "CMML"
clin = clin[vid,]

rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

#write
writetxt_forGPMM(clin,'./GSE135902/clin_GSE135902.txt')
writetxt_forGPMM(fpkm,'./GSE135902/fpkm_GSE135902.txt')
writetxt_forGPMM(tpm,'./GSE135902/tpm_GSE135902.txt')
writetxt_forGPMM(rawCounts,'./GSE135902/rawCounts_GSE135902.txt')

```

## 2.6 GSE148817

**Information GSE148817**

-   **Citation**: Has this study been published? Please login to update or notify GEO.

-   **Summary**: The airway epithelium is a key protective barrier whose integrity is preserved by the self-renewal and differentiation of basal progenitor cells. Epithelial cells are central to the pathogenesis of multiple chronic lung diseases for which age is a principle risk factor. Children are also less susceptible to SARS-CoV-2 infection, suffer less severe symptoms than adults and have a lower rate of mortality. Few studies have addressed differences between airway epithelial cells in children and adults. Here, we perform bulk RNA sequencing studies in laser-captured whole epithelium, FACS-sorted basal cells and cultured basal cells, as well as in vitro cell proliferation experiments, to address the intrinsic molecular differences between paediatric and adult airway basal cells. We find that, while the cellular composition of the paediatric and adult tracheal epithelium is broadly similar, in cell culture, paediatric airway epithelial cells displayed higher colony forming ability, better in vitro growth and outcompeted adult cells in competitive proliferation assays. Although recurring differences between airway epithelial gene expression were seen between samples from children and adults, RNA sequencing showed broad conservation of transcriptional programmes. Genes associated with SARS-CoV-2 infection were not differentially expressed between children and adults, although individuals showed some variability in their expression of viral infection-associated genes. Our results chart important cell intrinsic differences in transcriptional profile and regenerative capacity between tracheal epithelial cells of children and adults.

-   **Submission information**

    |                   |                                                          |
    |-------------------|----------------------------------------------------------|
    | Submission date   | Apr 17, 2020                                             |
    | Last update date  | Apr 21, 2020                                             |
    | Contact name      | Adam Pennycuick                                          |
    | E-mail(s)         | [a.pennycuick\@ucl.ac.uk](mailto:a.pennycuick@ucl.ac.uk) |
    | Organization name | University College London                                |
    | Department        | UCL Respiratory                                          |
    | Lab               | Lungs for Living                                         |
    | Street address    | 5 University Street                                      |
    | City              | London                                                   |
    | ZIP/Postal code   | WC1E 6JF                                                 |
    | Country           | United Kingdom                                           |

```{r}
rawCounts = file2frame('./GSE148817/GSE148817_counts.cultured.txt',row.names = 1)

# get clin data
clin_org = getGEO(filename="./GSE148817/GSE148817_series_matrix.txt.gz")
clin = clin_org@phenoData@data
colnames(clin) = gsub(':','_',colnames(clin))
colnames(clin) = gsub(' ','_',colnames(clin))
rownames(clin) = clin$title

rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

#write
writetxt_forGPMM(clin,'./GSE148817/clin_GGSE148817.txt')
writetxt_forGPMM(fpkm,'./GSE148817/fpkm_GSE148817.txt')
writetxt_forGPMM(tpm,'./GSE148817/tpm_GSE148817.txt')
writetxt_forGPMM(rawCounts,'./GSE148817/rawCounts_GSE148817.txt')
```

## 2.7 GSE132383

**Information GSE132383**

-   **Citation**: Has this study been published? Please login to update or notify GEO.

-   **Summary**: Intrinsic dysfunction of hair follicle dermal progenitors underlies phenotypic changes observed in aged dermis in rodents. The authors confirm similar age-related decline in function is present in adult human DPCs. RNA-Seq is utilized to identify further explanations for this functional decline in aged hDPCs.

-   **Submission information**

    |                   |                        |
    |-------------------|------------------------|
    | Submission date   | Jun 07, 2019           |
    | Last update date  | Jun 06, 2020           |
    | Contact name      | Matthew Workentine     |
    | Organization name | University of Calgary  |
    | Street address    | 3280 Hospital Drive NW |
    | City              | Calgary                |
    | State/province    | Alberta                |
    | ZIP/Postal code   | T2N 4Z6                |
    | Country           | Canada                 |

```{r}
rawCounts = file2frame('./GSE132383/GSE132383_counts.txt')
rawCounts[,1] = geneheader[rawCounts[,1],]$gene_name
rawCounts = rawCounts[!is.na(rawCounts[,1]),]
rawCounts = del_duplicate_genes(rawCounts)

# get clin data
clin_org = getGEO(filename="./GSE132383/GSE132383_series_matrix.txt.gz")
clin = clin_org@phenoData@data
colnames(clin) = gsub(':','_',colnames(clin))
colnames(clin) = gsub(' ','_',colnames(clin))
IDs = paste0('X',clin$title)
IDs = gsub('XS','S',IDs)
rownames(clin) = IDs

rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

#write
writetxt_forGPMM(clin,'./GSE132383/clin_GSE132383.txt')
writetxt_forGPMM(fpkm,'./GSE132383/fpkm_GSE132383.txt')
writetxt_forGPMM(tpm,'./GSE132383/tpm_GSE132383.txt')
writetxt_forGPMM(rawCounts,'./GSE132383/rawCounts_GSE132383.txt')
```

## 2.8 GSE226189

**Information GSE226189**

-   **Citation**: Has this study been published? Please login to update or notify GEO.

-   **Summary**: Changes in the transcriptomes of human tissues with advancing age are poorly cataloged. Here, we set to identify the coding and long noncoding RNAs present in cultured primary skin fibroblasts collected from 82 healthy individuals across a wide age spectrum (22 to 89 years old) who participated in the GESTALT (Genetic and Epigenetic Signatures of Translational Aging Laboratory Testing) study of the National Institute on Aging, NIH. Using high-throughput RNA sequencing and a linear regression model, we identified 1,437 coding RNA (mRNA) and 1,177 linear and circular long noncoding (lncRNA) that were differentially abundant as a function of age. GSEA analysis revealed select transcription factors implicated in coordinating the transcription of subsets of differentially abundant mRNAs, while LncSEA identified RNA-binding proteins predicted to contribute to the age-associated lncRNA profiles. In sum, we report age-associated changes in the global transcriptome, coding and noncoding, from healthy human skin fibroblasts, and propose that they might serve as biomarkers and therapeutic targets in aging skin.

-   **Submission information**

    +-------------------+---------------------------------------+
    | \                 | Feb 27, 2023                          |
    | Submission date   |                                       |
    +-------------------+---------------------------------------+
    | Last update date  | Jul 07, 2023                          |
    +-------------------+---------------------------------------+
    | Contact name      | Supriyo De                            |
    +-------------------+---------------------------------------+
    | Organization name | NIA-IRP, NIH                          |
    +-------------------+---------------------------------------+
    | Department        | Laboratory of Genetics and Genomics   |
    +-------------------+---------------------------------------+
    | Lab               | Computational Biology & Genomics Core |
    +-------------------+---------------------------------------+
    | Street address    | 251 Bayview Blvd                      |
    +-------------------+---------------------------------------+
    | City              | Baltimore                             |
    +-------------------+---------------------------------------+
    | State/province    | Maryland                              |
    +-------------------+---------------------------------------+
    | ZIP/Postal code   | 21224                                 |
    +-------------------+---------------------------------------+
    | Country           | USA                                   |
    +-------------------+---------------------------------------+

```{r}
untar('./GSE226189/GSE226189_RAW.tar',exdir = './GSE226189/GSE226189')
require(rbindlist)
files = list.files(path = './GSE226189/GSE226189',pattern = 'geneCOUNT.txt.gz',full.names = T)
tlist = list()
for(i in 1:length(files)){
  tmp = read.table(gzfile(files[i]))
  tmp  = tmp[-1,]
  tmp$V2 = as.numeric(tmp$V2)
  xx  = as.data.frame(t(tmp[,2]))
  colnames(xx) = tmp[,1]
  tlist[[i]] = xx
}
rawCounts = t(as.matrix(rbindlist(tlist,fill = T)))
ids = regmatches(files,regexpr('GSM\\d+',files))
colnames(rawCounts) = ids
rawCounts = cbind(data.frame(gene = rownames(rawCounts)),as.data.frame(rawCounts))
#
rawCounts[,1] = geneheader[rawCounts[,1],]$gene_name
rawCounts = rawCounts[!is.na(rawCounts[,1]),]
rawCounts = del_duplicate_genes(rawCounts)

# get clin data
clin_org = getGEO(filename="./GSE226189/GSE226189_series_matrix.txt.gz")
clin = clin_org@phenoData@data
colnames(clin) = gsub('[: ]','_',colnames(clin))
colnames(clin) = gsub('[\\(\\))]','',colnames(clin))


rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

#write
writetxt_forGPMM(clin,'./GSE226189/clin_GSE226189.txt')
writetxt_forGPMM(fpkm,'./GSE226189/fpkm_GSE226189.txt')
writetxt_forGPMM(tpm,'./GSE226189/tpm_GSE226189.txt')
writetxt_forGPMM(rawCounts,'./GSE226189/rawCounts_GSE226189.txt')
```

## 2.9 GSE103756

**Information GSE103756**

-   **Citation**: Zhang C, Lee HJ, Shrivastava A, Wang R et al. Long-Term In Vitro Expansion of Epithelial Stem Cells Enabled by Pharmacological Inhibition of PAK1-ROCK-Myosin II and TGF-β Signaling. Cell Rep 2018 Oct 16;25(3):598-610.e5. PMID: [30332641](https://www.ncbi.nlm.nih.gov/pubmed/30332641 "Link to PubMed record").

-   **Summary**: Despite substantial self-renewal capability in vivo, epithelial stem and progenitor cells located in various tissues expand for a few passages in vitro in feeder-free condition before they succumb to growth arrest. Here we describe the EpiX™ method that utilizes small molecules that inhibit PAK1-ROCK-Myosin II and TGF-β signaling to achieve over one trillion-fold expansion of human epithelial stem and progenitor cells from skin, airway, mammary and prostate glands in the absence of feeder cells. Transcriptomic and epigenomic studies show that this condition helps epithelial cells to overcome stresses for continuous proliferation. EpiX-expanded basal epithelial cells differentiate into mature epithelial cells consistent with their tissue origins. Whole genome sequencing reveals that the cells retain remarkable genome integrity after extensive in vitro expansion without acquiring tumorigenicity. EpiX technology provides a solution to exploit the potential of tissue-resident epithelial stem and progenitor cells for regenerative medicine.

-   **Submission information**

    |                   |                                          |
    |-------------------|------------------------------------------|
    | Submission date   | Sep 12, 2017                             |
    | Last update date  | May 15, 2019                             |
    | Contact name      | Hyung Joo Lee                            |
    | Organization name | Washington University School of Medicine |
    | Department        | Department of Genetics                   |
    | Lab               | Ting Wang                                |
    | Street address    | 4515 McKinley Avenue                     |
    | City              | St. Louis                                |
    | State/province    | MO                                       |
    | ZIP/Postal code   | 63108                                    |
    | Country           | USA                                      |

```{r}
rawCounts = file2frame('./GSE103756/RNA_HFKn_featureCounts.txt')
rawCounts[,1] = gsub('\\.\\d+','',rawCounts[,1])
rawCounts[,1] = geneheader[rawCounts[,1],]$gene_name
rawCounts = rawCounts[,-c(2:6)]
rawCounts = rawCounts[!is.na(rawCounts[,1]),]
rawCounts = del_duplicate_genes(rawCounts)

# get clin data
clin_org = getGEO(filename="./GSE103756/GSE103756_series_matrix.txt.gz")
clin = clin_org@phenoData@data
colnames(clin) = gsub('[: ]','_',colnames(clin))
colnames(clin) = gsub('[\\(\\))]','',colnames(clin))
IDs = gsub('RNA_','',clin$title)
clin$ID = IDs
rownames(clin) = IDs


rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

#write
writetxt_forGPMM(clin,'./GSE103756/clin_GSE103756.txt')
writetxt_forGPMM(fpkm,'./GSE103756/fpkm_GSE103756.txt')
writetxt_forGPMM(tpm,'./GSE103756/tpm_GSE103756.txt')
writetxt_forGPMM(rawCounts,'./GSE103756/rawCounts_GSE103756.txt')
```

## 2.10 GSE222639

**Information GSE222639**

-   **Citation**: Saidani M, Darle A, Jarrige M, Polveche H et al. Generating Functional and Highly Proliferative Melanocytes Derived from Human Pluripotent Stem Cells: A Promising Tool for Biotherapeutic Approaches to Treat Skin Pigmentation Disorders. Int J Mol Sci 2023 Mar 29;24(7). PMID: [37047372](https://www.ncbi.nlm.nih.gov/pubmed/37047372 "Link to PubMed record")

-   **Summary**: In this study, we successfully defined a highly efficient standardize method to derive hPSCs into a pure population of functional melanocytes. These cells can be maintain for more than 20 passages without observed any senescence. Transcriptomic anlaysis revealed a differentiation process with a logic gene expression profile following the kinetic of melanocyte embryonic development as previously described in mice. Our findings represent a fruitful ground that can serve as a database tool to decipher molecular signatures in normal and pathological conditions.

-   **Submission information**

    |                   |                                                  |
    |-------------------|--------------------------------------------------|
    | Submission date   | Jan 11, 2023                                     |
    | Last update date  | May 03, 2023                                     |
    | Contact name      | Hélène Polvèche                                  |
    | E-mail(s)         | [hpolveche\@istem.fr](mailto:hpolveche@istem.fr) |
    | Organization name | I-Stem                                           |
    | Street address    | 28, Rue Henri Desbruères                         |
    | City              | Corbeil-Essonnes                                 |
    | ZIP/Postal code   | 91100                                            |
    | Country           | France                                           |

```{r}
rawCounts = file2frame('./GSE222639/GSE222639_counts_v0.csv',sep = ';',row.names = 1)


# get clin data
clin_org = getGEO(filename="./GSE222639/GSE222639_series_matrix.txt.gz")
clin = clin_org@phenoData@data
colnames(clin) = gsub('[: ]','_',colnames(clin))
colnames(clin) = gsub('[\\(\\))]','',colnames(clin))
IDs = regmatches(clin$title,regexpr('\\[\\w+\\]',clin$title))
IDs = gsub('\\[|\\]','',IDs)
clin$ID = IDs
rownames(clin) = IDs


rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

GSEid = 'GSE222639'

#write
writetxt_forGPMM(clin, paste0('./',GSEid,'/clin_',GSEid,'.txt'))
writetxt_forGPMM(fpkm, paste0('./',GSEid,'/fpkm_',GSEid,'.txt'))
writetxt_forGPMM(tpm, paste0('./',GSEid,'/tpm_',GSEid,'.txt'))
writetxt_forGPMM(rawCounts, paste0('./',GSEid,'/rawCounts_',GSEid,'.txt'))
```

## 2.11 GSE180406

**Information GSE180406**

-   **Citation**: Rey-Millet M, Pousse M, Soithong C, Ye J et al. Senescence-associated transcriptional derepression in subtelomeres is determined in a chromosome-end-specific manner. Aging Cell 2023 May;22(5):e13804. PMID: [36924026](https://www.ncbi.nlm.nih.gov/pubmed/36924026 "Link to PubMed record")

-   **Summary**: Telomeres, the nucleoprotein structures located at the end of eukaryotic DNA, protect chromosomal integrity. These structures undergo changes during development and aging, including length shortening and alterations in the levels of the proteins associated to them called shelterin, all this affecting genome stability as the cells age. However, telomeres also behave as transcriptional regulators acting not only on genes present at subtelomeres but also on more distantly located genes presented throughout the genome. This process is referred as Telomere Position Effect (TPE). Here we set to determine whether telomere shortening in senescent cells impact TPE and senescent cells transcriptome. We also investigate whether TRF2 (a shelterin protein) and Lamin B1 (a member of nuclear lamina) are involved in TPE by doing RNA sequencing of young and senescent cells in which we modulated TRF2 and Lamin B1 expression.

-   **Submission information**

    |                   |                                                          |
    |-------------------|----------------------------------------------------------|
    | Submission date   | Jul 19, 2021                                             |
    | Last update date  | Mar 23, 2023                                             |
    | Contact name      | Melanie Pousse                                           |
    | Organization name | IRCAN                                                    |
    | Department        | Bioinformatics                                           |
    | Street address    | Faculty of Medicine - Pasteur tower - 28, av. Valombrose |
    | City              | Nice                                                     |
    | ZIP/Postal code   | 06107                                                    |
    | Country           | France                                                   |

```{r}
rawCounts = read.table(gzfile("./GSE180406/GSE180406_raw_counts.txt.gz"),header = T)
rawCounts[,1] = gsub('\\.\\d+','',rawCounts[,1])
rawCounts[,1] = geneheader[rawCounts[,1],]$gene_name
rawCounts = rawCounts[!is.na(rawCounts[,1]),]
rawCounts = del_duplicate_genes(rawCounts)

# just use young and sens empty
vids = c("young_siCtrl_Empty_rep1","young_siCtrl_Empty_rep2","young_siCtrl_Empty_rep3","sen_siCtrl_Empty_rep1",
         "sen_siCtrl_Empty_rep2" ,"sen_siCtrl_Empty_rep3" )
rawCounts = rawCounts[,vids]


# get clin data
clin = data.frame(ID = colnames(rawCounts),
                  age = c('young','young','young','sen','sen','sen'))
rownames(clin) = clin$ID



# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

GSEid = 'GSE180406'

#write
writetxt_forGPMM(clin, paste0('./',GSEid,'/clin_',GSEid,'.txt'))
writetxt_forGPMM(fpkm, paste0('./',GSEid,'/fpkm_',GSEid,'.txt'))
writetxt_forGPMM(tpm, paste0('./',GSEid,'/tpm_',GSEid,'.txt'))
writetxt_forGPMM(rawCounts, paste0('./',GSEid,'/rawCounts_',GSEid,'.txt'))

```

## 2.12 GSE133281

**Information GSE133281**

-   **Citation**: Has this study been published? Please login to update or notify GEO.

-   **Summary**: RNA-seq analysis comparing the transcriptional landscape of normal young, normal aged and MDS-associated HSCs.

-   **Submission information**

    |                   |                                                  |
    |-------------------|--------------------------------------------------|
    | Submission date   | Jun 25, 2019                                     |
    | Last update date  | Jun 30, 2023                                     |
    | Contact name      | Teresa Ezponda                                   |
    | E-mail(s)         | [tezponda\@gmail.com](mailto:tezponda@gmail.com) |
    | Organization name | Center for Applied Medical Research              |
    | Department        | Hematology/Oncology                              |
    | Lab               | 1.02                                             |
    | Street address    | Pio XII 55                                       |
    | City              | Pamplona                                         |
    | State/province    | Navarra                                          |
    | ZIP/Postal code   | 31008                                            |
    | Country           | Spain                                            |

```{r}
rawCounts = read.table(gzfile("./GSE133281/GSE133281_DataRaw_ReadCounts.txt.gz"),row.names = NULL, header = T)
rawCounts = del_duplicate_genes(rawCounts)


# get clin data
clin_org = getGEO(filename="./GSE133281/GSE133281_series_matrix.txt.gz")
clin = clin_org@phenoData@data
colnames(clin) = gsub('[: ]','_',colnames(clin))
colnames(clin) = gsub('[\\(\\))]','',colnames(clin))
IDs = clin$title
clin$ID = IDs
rownames(clin) = IDs

vid = clin$disease_diagnosis_ch1 == 'Healthy'
clin = clin[vid,]
vid = intersect(rownames(clin),colnames(rawCounts))
clin = clin[vid,]

rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

GSEid = 'GSE133281'

#write
writetxt_forGPMM(clin, paste0('./',GSEid,'/clin_',GSEid,'.txt'))
writetxt_forGPMM(fpkm, paste0('./',GSEid,'/fpkm_',GSEid,'.txt'))
writetxt_forGPMM(tpm, paste0('./',GSEid,'/tpm_',GSEid,'.txt'))
writetxt_forGPMM(rawCounts, paste0('./',GSEid,'/rawCounts_',GSEid,'.txt'))


```

## 2.13 GSE88848

**Information GSE88848**

-   **Citation**: Samuel W, Jaworski C, Postnikova OA, Kutty RK et al. Appropriately differentiated ARPE-19 cells regain phenotype and gene expression profiles similar to those of native RPE cells. Mol Vis 2017;23:60-89. PMID: [28356702](https://www.ncbi.nlm.nih.gov/pubmed/28356702 "Link to PubMed record")

-   **Summary**: The retinal pigment epithelial (RPE) cell line ARPE-19 provides a widely-used alternative to native RPE. However, retention of the native RPE phenotype becomes problematic after multiple passages. We wished to determine if suitable culture conditions and differentiation could restore RPE-appropriate gene expression to ARPE-19. ARPE-19 cells at passages p9 to p12, grown in DMEM containing high glucose and pyruvate with 1% fetal bovine serum, were differentiated for up to 4 months. Using RNA-Seq, we compared the transcriptome of ARPE-19 cells kept in long-term culture with those cultured for 4 days. The 4 month cells developed the classic native RPE phenotype with heavy pigmentation. RNA-Seq analysis provided a comprehensive view of the relative abundance and differential expression of genes in the 4 month cells. Of the 16,757 genes with detectable signals, nearly 2435 genes were upregulated, and 931 genes were down-regulated with a fold change differences of 2 or more. Genes characteristic of RPE, including RPE65, RDH5 and RDH10, were greatly increased in ARPE-19 cells maintained at confluence for 4 months. Comparison with microarray data sets from human primary cell lines revealed important overall similarities in expression of "signature" genes. The results of this study demonstrate that ARPE-19 cells can express genes specific to native human RPE cells when appropriately cultured, and thus, can provide a relevant system to study differentiated cellular functions of RPE in vitro.

-   **Submission information**

    |                   |                  |
    |-------------------|------------------|
    | Submission date   | Oct 17, 2016     |
    | Last update date  | May 15, 2019     |
    | Contact name      | Cynthia Jawoski  |
    | Organization name | NIH              |
    | Department        | NEI              |
    | Lab               | LRCMB            |
    | Street address    | bldg 6, room 136 |
    | City              | Bethesda         |
    | State/province    | MD               |
    | ZIP/Postal code   | 20892            |
    | Country           | USA              |

```{r}
# using salmon
```

## 2.14 GSE220113

**Information GSE220113**

-   **Citation**: Wang X, Campbell MR, Cho HY, Pittman GS et al. Epigenomic profiling of isolated blood cell types reveals highly specific B cell smoking signatures and links to disease risk. Clin Epigenetics 2023 May 25;15(1):90. PMID: [37231515](https://www.ncbi.nlm.nih.gov/pubmed/37231515 "Link to PubMed record")

-   **Summary**: To investigate blood biomarkers of smokers and to provide molecular basis of B cell subtype shift from naïve to memory in smokers, total RNA from peripheral CD19+ B cells isolated from healthy adult smokers and nonsmokers was subjected to RNA-Seq and data were analyzed with covariates including age, race, sex, BMI, estimated CD4+T cell%, CD8+T cell%, natural killer (NK) cell%, monocyte%, granulocyte%, or naive B cell%.

-   **Submission information**

    |                   |                                                     |
    |-------------------|-----------------------------------------------------|
    | Submission date   | Dec 05, 2022                                        |
    | Last update date  | Jun 21, 2023                                        |
    | Contact name      | Douglas A. Bell                                     |
    | E-mail(s)         | [bell1\@niehs.nih.gov](mailto:bell1@niehs.nih.gov)  |
    | Organization name | National Institute of Environmental Health Sciences |
    | Department        | Immunity Inflammation and Disease Lab               |
    | Street address    | PO Box 12233                                        |
    | City              | Research Triangle Park                              |
    | State/province    | NC                                                  |
    | ZIP/Postal code   | 27709                                               |
    | Country           | USA                                                 |

```{r}
untar('./GSE220113/GSE220113_RAW.tar',exdir = './GSE220113/GSE220113')
require(rbindlist)
files = list.files(path = './GSE220113/GSE220113',pattern = 'txt.gz',full.names = T)
tlist = list()
tlist[[length(files)]] = ''
for(i in 1:length(files)){
  tmp = read.table(gzfile(files[i]),header = T)
  xx  = as.data.frame(t(tmp[,2]))
  colnames(xx) = tmp[,1]
  tlist[[i]] = xx
}
rawCounts = t(as.matrix(rbindlist(tlist,fill = T)))
ids = regmatches(files,regexpr('GSM\\d+',files))
colnames(rawCounts) = ids


# get clin data
clin_org = getGEO(filename="./GSE220113/GSE220113_series_matrix.txt.gz")
clin = clin_org@phenoData@data
colnames(clin) = gsub('[: ]','_',colnames(clin))
colnames(clin) = gsub('[\\(\\))]','',colnames(clin))


rawCounts = rawCounts[,rownames(clin)]

# get TPM FPKM
fpkm = get_FPKM(rawCounts)
tpm = get_TPM(rawCounts)

GSEid = 'GSE220113'

#write
writetxt_forGPMM(clin, paste0('./',GSEid,'/clin_',GSEid,'.txt'))
writetxt_forGPMM(fpkm, paste0('./',GSEid,'/fpkm_',GSEid,'.txt'))
writetxt_forGPMM(tpm, paste0('./',GSEid,'/tpm_',GSEid,'.txt'))
writetxt_forGPMM(rawCounts, paste0('./',GSEid,'/rawCounts_',GSEid,'.txt'))
```
